<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Processing and Visualization for Metagenomics: Taxonomic Assignment</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/lab/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/lab/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/lab/favicon-16x16.png"><link rel="manifest" href="favicons/lab/site.webmanifest"><link rel="mask-icon" href="favicons/lab/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav lab"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="lab-logo" alt="Lesson Description" src="assets/images/lab-logo.svg"><abbr class="badge peer-reviewed" title="This lesson has passed peer review.">
          <a href="https://doi.org/10.21105/jose.00209" class="external-link alert-link">
            <i aria-hidden="true" class="icon" data-feather="user-check" style="border-radius: 5px"></i>
            DOI: 10.21105/jose.00209
          </a>
        </abbr>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/06-taxonomic.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav lab" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/lab-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 55%" class="percentage">
    55%
  </div>
  <div class="progress lab">
    <div class="progress-bar lab" role="progressbar" style="width: 55%" aria-valuenow="55" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/06-taxonomic.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-assessing-read-quality.html">2. Assessing Read Quality</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        6. Taxonomic Assignment
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#what-is-a-taxonomic-assignment">What is a taxonomic assignment?</a></li>
<li><a href="#using-kraken-2">Using Kraken 2</a></li>
<li><a href="#visualization-of-taxonomic-assignment-results">Visualization of taxonomic assignment results</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-phyloseq.html">7. Exploring Taxonomy with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-Diversity-tackled-with-R.html">8. Diversity Tackled With R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-abundance-analyses.html">9. Taxonomic Analysis with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-binning.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="07-phyloseq.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-binning.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Metagenome Binning
        </a>
        <a class="chapter-link float-end" href="07-phyloseq.html" rel="next">
          Next: Exploring Taxonomy...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Taxonomic Assignment</h1>
        <p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/06-taxonomic.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I know to which taxa my sequences belong?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand how taxonomic assignment works.</li>
<li>Use Kraken to assign taxonomies to reads and contigs.</li>
<li>Visualize taxonomic assignations in graphics.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-is-a-taxonomic-assignment">What is a taxonomic assignment?<a class="anchor" aria-label="anchor" href="#what-is-a-taxonomic-assignment"></a></h2>
<hr class="half-width"><p>A taxonomic assignment is a process of assigning an Operational
Taxonomic Unit (OTU, that is, groups of related individuals) to
sequences that can be reads or contigs. Sequences are compared against a
database constructed using complete genomes. When a sequence finds a
good enough match in the database, it is assigned to the corresponding
OTU. The comparison can be made in different ways.</p>
<div class="section level3">
<h3 id="strategies-for-taxonomic-assignment">Strategies for taxonomic assignment<a class="anchor" aria-label="anchor" href="#strategies-for-taxonomic-assignment"></a></h3>
<p>There are many programs for doing taxonomic mapping, and almost all
of them follow one of the following strategies:</p>
<ol style="list-style-type: decimal"><li><p>BLAST: Using BLAST or DIAMOND, these mappers search for the most
likely hit for each sequence within a database of genomes (i.e.,
mapping). This strategy is slow.</p></li>
<li><p>Markers: They look for markers of a database made a priori in the
sequences to be classified and assigned the taxonomy depending on the
hits obtained.</p></li>
<li><p>K-mers: A genome database is broken into pieces of length k to be
able to search for unique pieces by taxonomic group, from a lowest
common ancestor (LCA), passing through phylum to species. Then, the
algorithm breaks the query sequence (reads/contigs) into pieces of
length k, looks for where these are placed within the tree and make the
classification with the most probable position.</p></li>
</ol><p><a href="fig/03-06-01.png">
<img src="fig/03-06-01.png" alt="Diagram of a taxonomic tree with four levels of nodes, some nodes have a number from 1 to 3, and some do not. From the most recent nodes, one has a three, and its parent nodes do not have numbers. This node with a three is selected." class="figure"></a> <em> Figure 1. Lowest common ancestor assignment example.<em></em></em></p>
</div>
<div class="section level3">
<h3 id="abundance-bias">Abundance bias<a class="anchor" aria-label="anchor" href="#abundance-bias"></a></h3>
<p>When you do the taxonomic assignment of metagenomes, a key result is
the abundance of each taxon or OTU in your sample. The absolute
abundance of a taxon is the number of sequences (reads or contigs,
depending on what you did) assigned to it. Moreover, its relative
abundance is the proportion of sequences assigned to it. It is essential
to be aware of the many biases that can skew the abundances along the
metagenomics workflow, shown in the figure, and that because of them, we
may not be obtaining the actual abundance of the organisms in the
sample.</p>
<p><a href="fig/03-06-02.png">
<img src="fig/03-06-02.png" alt="Flow diagram that shows how the initial composition of 33% for each of the three taxa in the sample ends up being 4%, 72%, and 24% after the biases imposed by the extraction, PCR, sequencing and bioinformatics steps." class="figure"></a> <em>Figure 2. Abundance biases during a metagenomics protocol.
<em></em></em></p>
<div id="discussion-taxonomic-level-of-assignment" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion-taxonomic-level-of-assignment" class="callout-inner">
<h3 class="callout-title">Discussion: Taxonomic level of assignment</h3>
<div class="callout-content">
<p>What do you think is harder to assign, a species (like <em>E.
coli</em>) or a phylum (like Proteobacteria)?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="using-kraken-2">Using Kraken 2<a class="anchor" aria-label="anchor" href="#using-kraken-2"></a></h2>
<hr class="half-width"><p><a href="https://ccb.jhu.edu/software/kraken2/" class="external-link">Kraken 2</a> is the
newest version of Kraken, a taxonomic classification system using exact
k-mer matches to achieve high accuracy and fast classification speeds.
<code>kraken2</code> is already installed in the <strong>metagenomics
environment</strong>, let us have a look at <code>kraken2</code>
help.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> kraken2  <span class="at">--help</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Need to specify input filenames!
Usage: kraken2 [options] &lt;filename(s)&gt;

Options:
  --db NAME               Name for Kraken 2 DB
                          (default: none)
  --threads NUM           Number of threads (default: 1)
  --quick                 Quick operation (use first hit or hits)
  --unclassified-out FILENAME
                          Print unclassified sequences to filename
  --classified-out FILENAME
                          Print classified sequences to filename
  --output FILENAME       Print output to filename (default: stdout); "-" will
                          suppress normal output
  --confidence FLOAT      Confidence score threshold (default: 0.0); must be
                          in [0, 1].
  --minimum-base-quality NUM
                          Minimum base quality used in classification (def: 0,
                          only effective with FASTQ input).
  --report FILENAME       Print a report with aggregate counts/clade to file
  --use-mpa-style         With --report, format report output like Kraken 1's
                          kraken-mpa-report
  --report-zero-counts    With --report, report counts for ALL taxa, even if
                          counts are zero
  --report-minimizer-data With --report, report minimizer, and distinct minimizer
                          count information in addition to normal Kraken report
  --memory-mapping        Avoids loading database into RAM
  --paired                The filenames provided have paired-end reads
  --use-names             Print scientific names instead of just taxids
  --gzip-compressed       Input files are compressed with gzip
  --bzip2-compressed      Input files are compressed with bzip2
  --minimum-hit-groups NUM
                          Minimum number of hit groups (overlapping k-mers
                          sharing the same minimizer) needed to make a call
                          (default: 2)
  --help                  Print this message
  --version               Print version information

If none of the *-compressed flags are specified, and the filename provided
is a regular file, automatic format detection is attempted.
</code></pre>
</div>
<p>In the help, we can see that in addition to our input files, we also
need a database to compare them. The database you use will determine the
result you get for your data. Imagine you are searching for a recently
discovered lineage that is not part of the available databases. Would
you find it?</p>
<p>There are <a href="https://ccb.jhu.edu/software/kraken2/downloads.shtml" class="external-link">several
databases</a> compatible to be used with kraken2 in the taxonomical
assignment process.</p>
<p>Unfortunately, even the smallest Kraken database Minikraken, which
needs 8Gb of free RAM, is not small enough to be run by the machines we
are using, so <strong>we will not be able to run
<code>kraken2</code></strong>. We can check our available RAM with
<code>free -h</code>to be sure of this.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> free <span class="at">-h</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>              total        used        free      shared  buff/cache   available
Mem:           3.9G        272M        3.3G         48M        251M        3.3G
Swap:            0B          0B          0B</code></pre>
</div>
<div class="section level3">
<h3 id="taxonomic-assignment-of-metagenomic-reads">Taxonomic assignment of metagenomic reads<a class="anchor" aria-label="anchor" href="#taxonomic-assignment-of-metagenomic-reads"></a></h3>
<p>As we have learned, taxonomic assignments can be attempted before the
assembly. In this case, we would use FASTQ files as inputs, which would
be <code>JP4D_R1.trim.fastq.gz</code> and
<code>JP4D_R2.trim.fastq.gz</code>. And the outputs would be two files:
the report <code>JP4D.report</code> and the kraken file
<code>JP4D.kraken</code>.</p>
<p>To run kraken2, we would use a command like this:<br><strong>No need to run this</strong></p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> mkdir TAXONOMY_READS</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">$</span> kraken2 <span class="at">--db</span> kraken-db <span class="at">--threads</span> 8 <span class="at">--paired</span> JP4D_R1.trim.fastq.gz JP4D_R2.trim.fastq.gz <span class="at">--output</span> TAXONOMY_READS/JP4D.kraken <span class="at">--report</span> TAXONOMY_READS/JP4D.report</span></code></pre>
</div>
<p>Since we cannot run <code>kraken2</code> here, we precomputed its
results in a server, i.e., a more powerful machine. In the server we ran
<code>kraken2</code> and obtained<code>JP4D-kraken.kraken</code> and
<code>JP4D.report</code>.</p>
<p>Let us look at the precomputed outputs of <code>kraken2</code> for
our JP4D reads.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">head</span> ~/dc_workshop/taxonomy/JP4D.kraken  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:19691:2037	0	250|251	0:216 |:| 0:217
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14127:2052	0	250|238	0:216 |:| 0:204
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14766:2063	0	251|251	0:217 |:| 0:217
C	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15697:2078	2219696	250|120	0:28 350054:5 1224:2 0:1 2:5 0:77 2219696:5 0:93 |:| 379:4 0:82
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15529:2080	0	250|149	0:216 |:| 0:115
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14172:2086	0	251|250	0:217 |:| 0:216
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:17552:2088	0	251|249	0:217 |:| 0:215
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14217:2104	0	251|227	0:217 |:| 0:193
C	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15110:2108	2109625	136|169	0:51 31989:5 2109625:7 0:39 |:| 0:5 74033:2 31989:5 1077935:1 31989:7 0:7 60890:2 0:105 2109625:1
C	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:19558:2111	119045	251|133	0:18 1224:9 2:5 119045:4 0:181 |:| 0:99</code></pre>
</div>
<p>This information may need to be clarified. Let us take out our
cheatsheet to understand some of its components:</p>
<table class="table"><colgroup><col width="27%"><col width="72%"></colgroup><thead><tr class="header"><th>Column example</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>C</td>
<td>Classified or unclassified</td>
</tr><tr class="even"><td>MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15697:2078</td>
<td>FASTA header of the sequence</td>
</tr><tr class="odd"><td>2219696</td>
<td>Tax ID</td>
</tr><tr class="even"><td>250:120</td>
<td>Read length</td>
</tr><tr class="odd"><td>0:28 350054:5 1224:2 0:1 2:5 0:77 2219696:5 0:93 379:4 0:82</td>
<td>kmers hit to a taxonomic ID <em>e.g.,</em> tax ID 350054 has five
hits, tax ID 1224 has two hits, etc.</td>
</tr></tbody></table><p>The Kraken file could be more readable. So let us look at the report
file:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">head</span> ~/dc_workshop/taxonomy/JP4D.report</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> 78.13	587119	587119	U	0	unclassified
 21.87	164308	1166	R	1	root
 21.64	162584	0	R1	131567	  cellular organisms
 21.64	162584	3225	D	2	    Bacteria
 18.21	136871	3411	P	1224	      Proteobacteria
 14.21	106746	3663	C	28211	        Alphaproteobacteria
  7.71	57950	21	O	204455	          Rhodobacterales
  7.66	57527	6551	F	31989	            Rhodobacteraceae
  1.23	9235	420	G	1060	              Rhodobacter
  0.76	5733	4446	S	1063	                Rhodobacter sphaeroides</code></pre>
</div>
<table class="table"><colgroup><col width="27%"><col width="72%"></colgroup><thead><tr class="header"><th>Column example</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>78.13</td>
<td>Percentage of reads covered by the clade rooted at this taxon</td>
</tr><tr class="even"><td>587119</td>
<td>Number of reads covered by the clade rooted at this taxon</td>
</tr><tr class="odd"><td>587119</td>
<td>Number of reads assigned directly to this taxon</td>
</tr><tr class="even"><td>U</td>
<td>A rank code, indicating (U)nclassified, (D)omain, (K)ingdom,
(P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies. All other
ranks are simply ‘-’.</td>
</tr><tr class="odd"><td>0</td>
<td>NCBI taxonomy ID</td>
</tr><tr class="even"><td>unclassified</td>
<td>Indented scientific name</td>
</tr></tbody></table></div>
<div class="section level3">
<h3 id="taxonomic-assignment-of-the-contigs-of-a-mag">Taxonomic assignment of the contigs of a MAG<a class="anchor" aria-label="anchor" href="#taxonomic-assignment-of-the-contigs-of-a-mag"></a></h3>
<p>We now have the taxonomic identity of the reads of the whole
metagenome, but we need to know to which taxon our MAGs correspond. For
this, we have to make the taxonomic assignment with their contigs
instead of its reads because we do not have the reads corresponding to a
MAG separated from the reads of the entire sample.</p>
<p>For this, the <code>kraken2</code> is a little bit different; here,
we can look at the command for the <code>JP4D.001.fasta</code> MAG:</p>
<p><strong>No need to run this</strong></p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> mkdir TAXONOMY_MAG</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">$</span> kraken2 <span class="at">--db</span> kraken-db <span class="at">--threads</span> 12 <span class="at">-input</span> JP4D.001.fasta <span class="at">--output</span> TAXONOMY_MAG/JP4D.001.kraken <span class="at">--report</span> TAXONOMY_MAG/JP4D.001.report</span></code></pre>
</div>
<p>The results of this are pre-computed in the
<code>~/dc_workshop/taxonomy/mags_taxonomy/</code> directory</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/taxonomy/mags_taxonomy</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">JP4D.001.kraken</span></span>
<span><span class="va">JP4D.001.report</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">more</span> ~/dc_workshop/taxonomy/mags_taxonomy/JP4D.001.report</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> 50.96	955	955	U	0	unclassified
 49.04	919	1	R	1	root
 48.83	915	0	R1	131567	  cellular organisms
 48.83	915	16	D	2	    Bacteria
 44.40	832	52	P	1224	      Proteobacteria
 19.37	363	16	C	28216	        Betaproteobacteria
 16.22	304	17	O	80840	          Burkholderiales
  5.66	106	12	F	506	            Alcaligenaceae
  2.72	51	3	G	517	              Bordetella
  1.12	21	21	S	2163011	                Bordetella sp. HZ20
  .
  .
  .</code></pre>
</div>
<p>Looking at the report, we can see that half of the contigs are
unclassified and that a tiny proportion of contigs have been assigned an
OTU. This result is weird because we expected only one genome in the
bin.</p>
<p>To exemplify how a report of a complete and not contaminated MAG
should look like this; let us look at the report of this MAG from
another study:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>100.00	108	0	R	1	root
100.00	108	0	R1	131567	  cellular organisms
100.00	108	0	D	2	    Bacteria
100.00	108	0	P	1224	      Proteobacteria
100.00	108	0	C	28211	        Alphaproteobacteria
100.00	108	0	O	356	          Rhizobiales
100.00	108	0	F	41294	            Bradyrhizobiaceae
100.00	108	0	G	374	              Bradyrhizobium
100.00	108	108	S	2057741	                Bradyrhizobium sp. SK17</code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="visualization-of-taxonomic-assignment-results">Visualization of taxonomic assignment results<a class="anchor" aria-label="anchor" href="#visualization-of-taxonomic-assignment-results"></a></h2>
<hr class="half-width"><p>After we have the taxonomy assignation, what follows is some
visualization of our results. <a href="https://github.com/marbl/Krona/wiki" class="external-link">Krona</a> is a hierarchical
data visualization software. Krona allows data to be explored with
zooming and multi-layered pie charts and supports several bioinformatics
tools and raw data formats. To use Krona in our results, let us first go
into our taxonomy directory, which contains the pre-calculated Kraken
outputs.</p>
<div class="section level3">
<h3 id="krona">Krona<a class="anchor" aria-label="anchor" href="#krona"></a></h3>
<p>With Krona, we will explore the taxonomy of the JP4D.001 MAG.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/taxonomy/mags_taxonomy</span></code></pre>
</div>
<p>Krona is called with the <code>ktImportTaxonomy</code> command that
needs an input and an output file.<br>
In our case, we will create the input file with columns three and four
from <code>JP4D.001.kraken</code> file.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">$</span> cut <span class="at">-f2,3</span> JP4D.001.kraken <span class="op">&gt;</span> JP4D.001.krona.input</span></code></pre>
</div>
<p>Now we call Krona in our <code>JP4D.001.krona.input</code> file and
save results in <code>JP4D.001.krona.out.html</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">$</span> ktImportTaxonomy JP4D.001.krona.input <span class="at">-o</span> JP4D.001.krona.out.html</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading taxonomy...
Importing JP4D.001.krona.input...
   [ WARNING ]  The following taxonomy IDs were not found in the local database and were set to root
                (if they were recently added to NCBI, use updateTaxonomy.sh to update the local
                database): 1804984 2109625 2259134</code></pre>
</div>
<p>And finally, open another terminal on your local computer, download
the Krona output and open it on a browser.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-3-235-238-92.compute-1.amazonaws.com:~/dc_workshop/taxonomy/JP4D.001.krona.out.html . </span></code></pre>
</div>
<p>You will see a page like this:</p>
<p><a href="fig/03-06-03.png">
<img src="fig/03-06-03.png" alt="Krona displays a circled-shape bacterial taxonomy plot with abundance percentages of each taxon" class="figure"></a></p>
<div id="exercise-1-exploring-krona-visualization" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-exploring-krona-visualization" class="callout-inner">
<h3 class="callout-title">Exercise 1: Exploring Krona visualization</h3>
<div class="callout-content">
<p>Try double-clicking on the pie chart segment representing Bacteria
and see what happens. What percentage of bacteria is represented by the
genus <em>Paracoccus</em>?</p>
<p>Hint: A search box is in the window’s top left corner.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>2% of Bacteria corresponds to the genus Paracoccus in this sample. In
the top right of the window, we see little pie charts that change
whenever we change the visualization to expand certain taxa.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="pavian">Pavian<a class="anchor" aria-label="anchor" href="#pavian"></a></h3>
<p>Pavian is another visualization tool that allows comparison between
multiple samples. Pavian should be locally installed and needs R and
Shiny, but we can try the <a href="https://fbreitwieser.shinyapps.io/pavian/" class="external-link">Pavian demo WebSite</a>
to visualize our results.</p>
<p>First, we need to download the files needed as inputs in Pavian; this
time, we will visualize the assignment of the reads of both samples:
<code>JC1A.report</code> and <code>JP4D.report</code>.<br>
These files correspond to our Kraken reports. Again in our local
machine, let us use the <code>scp</code> command.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-3-235-238-92.compute-1.amazonaws.com:~/dc_workshop/taxonomy/<span class="pp">*</span>report . </span></code></pre>
</div>
<p>We go to the <a href="https://fbreitwieser.shinyapps.io/pavian/" class="external-link">Pavian demo
WebSite</a>, click on Browse, and choose our reports. You need to select
both reports at the same time.</p>
<p><a href="fig/03-06-04.png">
<img src="fig/03-06-04.png" alt="Pavian website showing the upload of two reports" class="figure"></a></p>
<p>We click on the Results Overview tab.</p>
<p><a href="fig/03-06-05.png">
<img src="fig/03-06-05.png" alt="Results Overview tab of the Pavian website where it shows the number of reads classified to several categories for the two samples" class="figure"></a></p>
<p>We click on the Sample tab.</p>
<p><a href="fig/03-06-06.png">
<img src="fig/03-06-06.png" alt="Sankey type visualization that shows the abundance of each taxonomic label in a tree-like manner" class="figure"></a></p>
<p>We can look at the abundance of a specific taxon by clicking on
it.</p>
<p><a href="fig/03-06-07.png">
<img src="fig/03-06-07.png" alt="A bar chart of the abundance of reads of the two samples, showing a segment for the read identified at the specific taxon and another segment for the number of reads identifies at children of the specified taxon" class="figure"></a></p>
<p>We can look at a comparison of both our samples in the Comparison
tab.</p>
<p><a href="fig/03-06-08.png">
<img src="fig/03-06-08.png" alt="A table of the same format as the Kraken report but for both samples at once." class="figure"></a></p>
<div id="discussion-unclassified-reads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="discussion-unclassified-reads" class="callout-inner">
<h3 class="callout-title">Discussion: Unclassified reads</h3>
<div class="callout-content">
<p>As you can see, a percentage of our data could not be assigned to
belong to a specific OTU.<br>
Which factors can affect the taxonomic assignation so that a read is
unclassified?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p><strong>Unclassified reads</strong> can be the result of different
factors that can go from sequencing errors to problems with the
algorithm being used to generate the result. The widely used
Next-generation sequencing (NGS) platforms, showed <a href="https://www.nature.com/articles/s41598-018-29325-6" class="external-link">average error
rate of 0.24±0.06% per base</a>. Besides the sequencing error, we need
to consider the status of the database being used to perform the
taxonomic assignation.<br>
All the characterized genomes obtained by different research groups are
scattered in different repositories, pages, and banks in the cloud. Some
are still unpublished. Incomplete databases can affect the performance
of the taxonomic assignation. Imagine that the dominant OTU in your
sample belongs to a lineage that has never been characterized and does
not have a public genome available to be used as a template for the
database. This possibility makes the assignation an impossible task and
can promote the generation of false positives because the algorithm will
assign a different identity to all those reads.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>A database with previously gathered knowledge (genomes) is needed
for taxonomic assignment.</li>
<li>Taxonomic assignment can be done using Kraken.</li>
<li>Krona and Pavian are web-based tools to visualize the assigned
taxa.</li>
</ul></div>
</div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-binning.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="07-phyloseq.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-binning.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Metagenome Binning
        </a>
        <a class="chapter-link float-end" href="07-phyloseq.html" rel="next">
          Next: Exploring Taxonomy...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/06-taxonomic.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/06-taxonomic.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Taxonomic Assignment",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/06-taxonomic.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/06-taxonomic.html",
  "dateCreated": "2020-08-24",
  "dateModified": "2025-04-06",
  "datePublished": "2025-04-29"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

