<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Processing and Visualization for Metagenomics: Exploring Taxonomy with R</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/lab/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/lab/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/lab/favicon-16x16.png"><link rel="manifest" href="favicons/lab/site.webmanifest"><link rel="mask-icon" href="favicons/lab/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav lab"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="lab-logo" alt="Lesson Description" src="assets/images/lab-logo.svg"><abbr class="badge peer-reviewed" title="This lesson has passed peer review.">
          <a href="https://doi.org/10.21105/jose.00209" class="external-link alert-link">
            <i aria-hidden="true" class="icon" data-feather="user-check" style="border-radius: 5px"></i>
            DOI: 10.21105/jose.00209
          </a>
        </abbr>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/07-phyloseq.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav lab" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/lab-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 65%" class="percentage">
    65%
  </div>
  <div class="progress lab">
    <div class="progress-bar lab" role="progressbar" style="width: 65%" aria-valuenow="65" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/07-phyloseq.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-assessing-read-quality.html">2. Assessing Read Quality</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-taxonomic.html">6. Taxonomic Assignment</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        7. Exploring Taxonomy with R
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#creating-lineage-and-rank-tables">Creating lineage and rank tables</a></li>
<li><a href="#creating-and-manipulating-phyloseq-objects">Creating and manipulating Phyloseq objects</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-Diversity-tackled-with-R.html">8. Diversity Tackled With R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-abundance-analyses.html">9. Taxonomic Analysis with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="06-taxonomic.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="08-Diversity-tackled-with-R.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="06-taxonomic.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Taxonomic Assignment
        </a>
        <a class="chapter-link float-end" href="08-Diversity-tackled-with-R.html" rel="next">
          Next: Diversity Tackled...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Exploring Taxonomy with R</h1>
        <p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/07-phyloseq.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I use my taxonomic assignment results to analyze?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Comprehend which libraries are required for analysis of the taxonomy
of metagenomes.</li>
<li>Create and manage a Phyloseq object.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="creating-lineage-and-rank-tables">Creating lineage and rank tables<a class="anchor" aria-label="anchor" href="#creating-lineage-and-rank-tables"></a></h2>
<hr class="half-width"><p>In this episode, we will use RStudio to analyze our microbial
samples. You do not have to install anything, you already have an
instance on the cloud ready to be used.</p>
<p>Packages like Qiime2, MEGAN, Vegan, or Phyloseq in R allow us to
analyze diversity and abundance by manipulating taxonomic assignment
data. In this lesson, we will use Phyloseq. In order to do so, we need
to generate an abundance matrix from the Kraken output files. One
program widely used for this purpose is <code>kraken-biom</code>.</p>
<p>To do this, we could go to our now familiar Bash terminal, but
RStudio has an integrated terminal that uses the same language as the
one we learned in the Command-line lessons, so let us take advantage of
it. Let us open RStudio and go to the Terminal tab in the bottom left
panel.</p>
<div class="section level3">
<h3 id="kraken-biom">Kraken-biom<a class="anchor" aria-label="anchor" href="#kraken-biom"></a></h3>
<p><a href="https://github.com/smdabdoub/kraken-biom" class="external-link">Kraken-biom</a> is
a program that creates BIOM tables from the Kraken output.</p>
<p>In order to run Kraken-biom, we have to move to the folder where our
taxonomic data files are located:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/taxonomy</span></code></pre>
</div>
<p>First, we will visualize the content of our directory by the
<code>ls</code> command.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A.kraken  JC1A.report  JP41.report  JP4D.kraken  JP4D.report  mags_taxonomy</code></pre>
</div>
<p>The <code>kraken-biom</code> program is installed inside our
<code>metagenomics</code> environment, so let us activate it.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> conda activate metagenomics </span></code></pre>
</div>
<p>Let us take a look at the different flags that
<code>kraken-biom</code> has:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> kraken-biom <span class="at">-h</span>                  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>usage: kraken-biom [-h] [--max {D,P,C,O,F,G,S}] [--min {D,P,C,O,F,G,S}]
                   [-o OUTPUT_FP] [--otu_fp OTU_FP] [--fmt {hdf5,json,tsv}]
                   [--gzip] [--version] [-v]
                   kraken_reports [kraken_reports ...]

Create BIOM-format tables (http://biom-format.org) from Kraken output
(http://ccb.jhu.edu/software/kraken/).
.
.
.</code></pre>
</div>
<p>By a close look at the first output lines, it is noticeable that we
need a specific output from Kraken: the <code>.reports</code>.</p>
<p>With the following command, we will create a table in <a href="https://biom-format.org/" class="external-link">Biom</a> format called
<code>cuatroc.biom</code>. We will include the two samples we have been
working with (<code>JC1A</code> and <code>JP4D</code>) and a third one
(<code>JP41</code>) to be able to perform specific analyses later
on.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">$</span> kraken-biom JC1A.report JP4D.report JP41.report <span class="at">--fmt</span> json <span class="at">-o</span> cuatroc.biom</span></code></pre>
</div>
<p>If we inspect our folder, we will see that the
<code>cuatroc.biom</code> file has been created. This <code>biom</code>
object contains both the abundance and the ID (a number) of each
OTU.<br>
With this result, we are ready to return to RStudio’s console and begin
to manipulate our taxonomic-data.</p>
<div id="command-line-prompts" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="command-line-prompts" class="callout-inner">
<h3 class="callout-title">Command line prompts</h3>
<div class="callout-content">
<p>Note that you can distinguish the Bash terminal from the R console by
looking at the prompt. In Bash is the <code>$</code> sign, and in R is
the <code>&gt;</code> sign.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="creating-and-manipulating-phyloseq-objects">Creating and manipulating Phyloseq objects<a class="anchor" aria-label="anchor" href="#creating-and-manipulating-phyloseq-objects"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="load-required-packages">Load required packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a></h3>
<p>Phyloseq is a library with tools to analyze and plot your
metagenomics samples’ taxonomic assignment and abundance information.
Let us install <a href="https://joey711.github.io/phyloseq/" class="external-link">phyloseq</a> (This
instruction might not work on specific versions of R) and other
libraries required for its execution:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="sc">&gt;</span> BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"phyloseq"</span>) <span class="co"># Install phyloseq</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"RColorBrewer"</span>, <span class="st">"patchwork"</span>)) <span class="co">#install patchwork to chart publication-quality plots and readr to read rectangular datasets.</span></span></code></pre>
</div>
<p>Once the libraries are installed, we must make them available for
this R session. Now load the libraries (a process needed every time we
begin a new work session in R):</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"phyloseq"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"RColorBrewer"</span>)</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"patchwork"</span>)</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="creating-the-phyloseq-object">Creating the phyloseq object<a class="anchor" aria-label="anchor" href="#creating-the-phyloseq-object"></a></h3>
<p>First, we tell R in which directory we are working.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">setwd</span>(<span class="st">"~/dc_workshop/taxonomy/"</span>)</span></code></pre>
</div>
<p>Let us proceed to create the phyloseq object with the
<code>import_biom</code> command:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes <span class="ot">&lt;-</span> <span class="fu">import_biom</span>(<span class="st">"cuatroc.biom"</span>)</span></code></pre>
</div>
<p>Now, we can inspect the result by asking the class of the object
created and doing a close inspection of some of its content:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">class</span>(merged_metagenomes)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "phyloseq"
attr("package")
[1] "phyloseq"</code></pre>
</div>
<p>The “class” command indicates that we already have our phyloseq
object.</p>
</div>
<div class="section level3">
<h3 id="exploring-the-taxonomic-labels">Exploring the taxonomic labels<a class="anchor" aria-label="anchor" href="#exploring-the-taxonomic-labels"></a></h3>
<p>Let us try to access the data that is stored inside our
<code>merged_metagenomes</code> object. Since a phyloseq object is a
special object in R, we need to use the operator <code>@</code> to
explore the subsections of data inside <code>merged_metagenomes</code>.
If we type <code>merged_metagenomes@</code>, five options are displayed;
<code>tax_table</code> and <code>otu_table</code> are the ones we will
use. After writing <code>merged_metagenomes@otu_table</code> or
<code>merged_metagenomes@tax_table</code>, an option of
<code>.Data</code> will be the one chosen in both cases. Let us see what
is inside our <code>tax_table</code>:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p><a href="fig/03-07-01.png">
<img src="fig/03-07-01.png" alt="A table where the taxonomic
  identification information of all OTUs is displayed. Each row represents one
  OTU and the columns represent its identification at different levels in the taxonomic classification ranks, begging with Kingdom until we reach Species
  in the seventh column." class="figure"></a> <em> Figure 1. Table of the taxonomic
labels from our <code>merged_metagenomes</code> object. <em></em></em></p>
<p>Here we can see that the <code>tax_table</code> inside our phyloseq
object stores all the taxonomic labels corresponding to each OTU.
Numbers in the row names of the table identify OTUs.</p>
<p>Next, let us get rid of some of the unnecessary characters in the
OTUs id and put names to the taxonomic ranks:</p>
<p>To remove unnecessary characters in <code>.Data</code> (matrix), we
will use the command <code>substring()</code>. This command helps
extract or replace characters in a vector. To use the command, we have
to indicate the vector (x) followed by the first element to replace or
extract (first) and the last element to be replaced (last). For
instance: <code>substring (x, first, last)</code>.
<code>substring()</code> is a “flexible” command, especially to select
characters of different lengths, as in our case. Therefore, it is not
necessary to indicate “last”, so it will take the last position of the
character by default. Since a matrix is an arrangement of vectors, we
can use this command. Each character in <code>.Data</code> is preceded
by three spaces occupied by a letter and two underscores, for example:
<code>o__Rhodobacterales</code>. In this case, “Rodobacterales” starts
at position 4 with an R. So, to remove the unnecessary characters, we
will use the following code:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data <span class="ot">&lt;-</span> <span class="fu">substring</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data, <span class="dv">4</span>)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">colnames</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span>)</span></code></pre>
</div>
<p><a href="fig/03-07-02.png">
<img src="fig/03-07-02.png" alt="The same table we saw in Figure 3 but with informative names in each of the columns. Now, we can see which of
  the columns are associated with which taxonomic classification rank" class="figure"></a> <em> Figure 2. Table of the taxonomic labels from our
<code>merged_metagenomes</code> object with corrections. <em></em></em></p>
<p>We will use a command named <code>unique()</code> to explore how many
phyla we have. Let us see the result we obtain from the following
code:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="st">"Phylum"</span>])</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Proteobacteria"              "Actinobacteria"              "Firmicutes"
 [4] "Cyanobacteria"               "Deinococcus-Thermus"         "Chloroflexi"
 [7] "Armatimonadetes"             "Bacteroidetes"               "Chlorobi"
[10] "Gemmatimonadetes"            "Planctomycetes"              "Verrucomicrobia"
[13] "Lentisphaerae"               "Kiritimatiellaeota"          "Chlamydiae"
[16] "Acidobacteria"               "Spirochaetes"                "Synergistetes"
[19] "Nitrospirae"                 "Tenericutes"                 "Coprothermobacterota"
[22] "Ignavibacteriae"             "Candidatus Cloacimonetes"    "Fibrobacteres"
[25] "Fusobacteria"                "Thermotogae"                 "Aquificae"
[28] "Thermodesulfobacteria"       "Deferribacteres"             "Chrysiogenetes"
[31] "Calditrichaeota"             "Elusimicrobia"               "Caldiserica"
[34] "Candidatus Saccharibacteria" "Dictyoglomi" </code></pre>
</div>
<p>Knowing phyla is helpful, but what we need to know is how many of our
OTUs have been assigned to the phylum Firmicutes?. Let´s use the command
<code>sum()</code> to ask R:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">sum</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="st">"Phylum"</span>] <span class="sc">==</span> <span class="st">"Firmicutes"</span>)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 580</code></pre>
</div>
<p>Now, to know for that phylum in particular which taxa there are in a
certain rank, we can also ask it to phyloseq.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="st">"Phylum"</span>] <span class="sc">==</span> <span class="st">"Firmicutes"</span>, <span class="st">"Class"</span>])</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Bacilli"          "Clostridia"       "Negativicutes"    "Limnochordia"     "Erysipelotrichia" "Tissierellia" </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="exploring-the-abundance-table">Exploring the abundance table<a class="anchor" aria-label="anchor" href="#exploring-the-abundance-table"></a></h3>
<p>Until now, we have looked at the part of the phyloseq object that
stores the information about the taxonomy (at all the possible levels)
of each OTU found in our samples. However, there is also a part of the
phyloseq object that stores the information about how many sequenced
reads corresponding to a certain OTU are in each sample. This table is
the <code>otu_table</code>.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p><a href="fig/03-07-03.png">
<img src="fig/03-07-03.png" alt="A table where the abundance of each OTU in each sample is shown. Each row represents one
  OTU and the columns represent the samples. In the intersection, a number indicates how many sequenced reads of that OTU are present in that sample." class="figure"></a> <em> Figure 3. Table of the abundance of reads in the
<code>merged_metagenomes</code> object. <em></em></em></p>
<p>We will take advantage of this information later on in our
analyses.</p>
<div id="phyloseq-objects" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="phyloseq-objects" class="callout-inner">
<h3 class="callout-title">Phyloseq objects</h3>
<div class="callout-content">
<p>Finally, we can review our object and see that all datasets (i.e.,
JC1A, JP4D, and JP41) are in the object. If you look at our Phyloseq
object, you will see that there are more data types that we can use to
build our object(<code>?phyloseq()</code>), such as a phylogenetic tree
and metadata concerning our samples. These are optional, so we will use
our basic phyloseq object, composed of the abundances of specific OTUs
and the names of those OTUs.</p>
</div>
</div>
</div>
<div id="exercise-1-explore-a-phylum" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-explore-a-phylum" class="callout-inner">
<h3 class="callout-title">Exercise 1: Explore a phylum</h3>
<div class="callout-content">
<p>Go into groups and choose one phylum that is interesting for your
group, and use the learned code to find out how many OTUs have been
assigned to your chosen phylum and what are the unique names of the
genera inside it. がんばって! (ganbatte; <em>good luck</em>):</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Change the name of a new phylum wherever needed and the name of the
rank we are asking for to get the result. As an example, here is the
solution for Proteobacteria:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="va">merged_metagenomes</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span>,<span class="st">"Phylum"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"Proteobacteria"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">unique</span><span class="op">(</span><span class="va">merged_metagenomes</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span><span class="va">merged_metagenomes</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span>,<span class="st">"Phylum"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"Proteobacteria"</span>, <span class="st">"Genus"</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-2-searching-for-the-read-counts" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-searching-for-the-read-counts" class="callout-inner">
<h3 class="callout-title">Exercise 2: Searching for the read counts</h3>
<div class="callout-content">
<p>Using the information from both the <code>tax_table</code> and the
<code>otu_table</code>, find how many reads there are for any species of
your interest (one that can be found in the
<code>tax_table</code>).<br><strong>Hint</strong>: Remember that you can access the contents of a
data frame with the <code>["row_name", "column_name"]</code>
syntax.<br>
がんばって! (ganbatte; <em>good luck</em>):</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Go to the <code>tax_table</code>:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p>Take note of the OTU number for some species:
<a href="%7B%7B%20page.root%20%7D%7D/fig/03-07-04.png">
<img src="%7B%7B%20page.root%20%7D%7D/fig/03-07-04.png" alt="The OTU number is in the leftmost space of the table as a row name for the searched species." class="figure"></a> <em> Figure 4. The row of the <code>tax_table</code> corresponds to
the species <em>Paracoccus zhejiangensis</em>. <em></em></em></p>
<p>Search for the row of the <code>otu_table</code> with the row name
you chose.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data[<span class="st">"1077935"</span>,]</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>JC1A JP4D JP41 </span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>  <span class="dv">42</span>  <span class="dv">782</span>  <span class="dv">257</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>kraken-biom formats Kraken output-files of several samples into the
single <code>.biom</code> file that will be phyloseq input.</li>
<li>The library <code>phyloseq</code> manages metagenomics objects and
computes analyses.</li>
<li>A phyloseq object stores a table with the taxonomic information of
each OTU and a table with the abundance of each OTU.</li>
</ul></div>
</div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="06-taxonomic.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="08-Diversity-tackled-with-R.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="06-taxonomic.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Taxonomic Assignment
        </a>
        <a class="chapter-link float-end" href="08-Diversity-tackled-with-R.html" rel="next">
          Next: Diversity Tackled...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/07-phyloseq.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/07-phyloseq.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Exploring Taxonomy with R",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/07-phyloseq.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/07-phyloseq.html",
  "dateCreated": "2020-08-24",
  "dateModified": "2025-04-06",
  "datePublished": "2025-04-06"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

