<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Data Processing and Visualization for Metagenomics: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/lab/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/lab/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/lab/favicon-16x16.png">
<link rel="manifest" href="favicons/lab/site.webmanifest">
<link rel="mask-icon" href="favicons/lab/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav lab"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="lab-logo" alt="Lesson Description" src="assets/images/lab-logo.svg"><abbr class="badge peer-reviewed" title="This lesson has passed peer review.">
          <a href="https://doi.org/10.21105/jose.00209" class="external-link alert-link">
            <i aria-hidden="true" class="icon" data-feather="user-check" style="border-radius: 5px"></i>
            DOI: 10.21105/jose.00209
          </a>
        </abbr>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav lab" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/lab-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress lab">
    <div class="progress-bar lab" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-assessing-read-quality.html">2. Assessing Read Quality</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-taxonomic.html">6. Taxonomic Assignment</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-phyloseq.html">7. Exploring Taxonomy with R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-Diversity-tackled-with-R.html">8. Diversity Tackled With R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-abundance-analyses.html">9. Taxonomic Analysis with R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-background-metadata"><p>Content from <a href="01-background-metadata.html">Starting a Metagenomics Project</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/01-background-metadata.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you plan a metagenomics experiment?</li>
<li>How does a metagenomics project look like?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn the differences between shotgun and metabarcoding (amplicon
metagenomics) techniques.</li>
<li>Understand the importance of metadata.</li>
<li>Familiarize yourself with the Cuatro Ciénegas experiment.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="metagenomics">Metagenomics<a class="anchor" aria-label="anchor" href="#metagenomics"></a>
</h2>
<hr class="half-width">
<p>Metagenomes are collections of genomic sequences from various
(micro)organisms that coexist in any given space. They are like
snapshots that can give us information about the taxonomic and even
metabolic or functional composition of the communities we decide to
study. Thus, metagenomes are usually employed to investigate the ecology
of defining characteristics of niches (* e.g.,*, the human gut or the
ocean floor).</p>
<p>Since metagenomes are mixtures of sequences that belong to different
species, a metagenomic workflow is designed to answer two questions:</p>
<ol style="list-style-type: decimal">
<li>What species are represented in the sample?</li>
<li>What are they capable of doing?</li>
</ol>
<p>To find which species are present in a niche, we must do a taxonomic
assignation of the obtained sequences. To find out their capabilities,
we can look at the genes directly encoded in the metagenome or find the
genes associated with the species that we found. In order to know which
methodology we should use, it is essential to know what questions we
want to answer.</p>
</section><section><h2 class="section-heading" id="shotgun-and-amplicon">Shotgun and amplicon<a class="anchor" aria-label="anchor" href="#shotgun-and-amplicon"></a>
</h2>
<hr class="half-width">
<p>There are two paths to obtain information from a complex sample:</p>
<ol style="list-style-type: decimal">
<li><strong>Shotgun Metagenomics</strong></li>
<li>
<strong>Metabarcoding</strong>.</li>
</ol>
<p>Each is named after the sequencing methodology employed Moreover,
have particular use cases with inherent advantages and
disadvantages.</p>
<p>With <strong>Shotgun Metagenomics</strong>, we sequence random parts
(ideally all of them) of the genomes present in a sample. We can search
the origin of these pieces (<em>i.e.,</em> their taxonomy) and also try
to find to what part of the genome they correspond. Given enough pieces,
it is possible to obtain complete individual genomes from a shotgun
metagenome (MAGs), which could give us a bunch of information about the
species in our study. MAGs assembly, however, requires a lot of genomic
sequences from one organism. Since the sequencing is done at random, it
needs a high depth of community sequencing to ensure that we obtain
enough pieces of a given genome. Required depth gets exponentially
challenging when our species of interest is not very abundant. It also
requires that we have enough DNA to work with, which can be challenging
to obtain in some instances. Finally, sequencing is expensive, and
because of this, making technical and biological replicates can be
prohibitively costly.</p>
<p>On the contrary, <strong>Metabarcoding</strong> tends to be cheaper,
which makes it easier to duplicate and even triplicate them without
taking a big financial hit. The lower cost is because Metabarcoding is
the collection of small genomic fragments present in the community and
amplified through PCR. Ideally, if the amplified region is present only
once in every genome, we would not need to sequence the amplicon
metagenome so thoroughly because one sequence is all we need to get the
information about that genome, and by extension, about that species. On
the other hand, if a genome in the community lacks the region targeted
by the PCR primers, then no amount of sequencing can give us information
about that genome. Conservation across species is why the most popular
amplicon used for this methodology are 16S amplicons for Bacteria since
every known bacterium has this particular region. Other regions can be
chosen, but they are used for specific cases. However, even 16S
amplicons are limited to, well, the 16S region, so amplicon metagenomes
cannot directly tell us a lot about the metabolic functions found in
each genome, although educated guesses can be made by knowing which
genes are commonly found in every identified species.</p>
<p><a href="fig/03-01-01.png">
<img src="fig/03-01-01.png" alt="Flow chart that shows the steps of a metagenomics project: Experimental design, Sampling, DNA extraction, Sequencing, Read quality, Assembly, Binning, Bin quality and Data analysis " class="figure"></a></p>
</section><section><h2 class="section-heading" id="on-metadata">On Metadata<a class="anchor" aria-label="anchor" href="#on-metadata"></a>
</h2>
<hr class="half-width">
<p>Once we have chosen an adequate methodology for our study, we must
take extensive notes on the origin of our samples and how we treated
them. These notes constitute the <strong>metadata</strong>, or data
about our data, and they are crucial to understanding and interpreting
the results we will obtain later in our metagenomic analysis. Most of
the time, the differences that we observe when comparing metagenomes can
be correlated to the metadata, which is why we must devote a whole
section of our experimental design to the metadata we expect to collect
and record carefully.</p>
<div id="discussion-1-choosing-amplicon-or-shotgun-sequencing" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="discussion-1-choosing-amplicon-or-shotgun-sequencing" class="callout-inner">
<h3 class="callout-title">Discussion #1: Choosing amplicon or shotgun sequencing?</h3>
<div class="callout-content">
<p>Suppose you want to find the source of a nasty gut infection in
people. Which type of sequencing methodology would you choose?<br>
Which type of metadata would be helpful to record?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>For a first exploration, 16S is a better idea since you could detect
known pathogens by knowing the taxons in the community. Nevertheless, if
the disease is the consequence of a viral infection, the pathogen can
only be discovered with shotgun metagenomics (that was the case of
SARS-CoV 2). Also, metabarcoding does not provide insights into the
genetic basis of the pathogenic phenotypes. Metadata will depend on the
type of experiment. For this case, some helpful metadata could be
sampling methodology, date, place (country, state, region, city, etc.),
patient’s sex and age, the anatomical origin of the sample, symptoms,
medical history, diet, lifestyle, and environment.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="cuatro-ciénegas">Cuatro Ciénegas<a class="anchor" aria-label="anchor" href="#cuatro-ci%C3%A9negas"></a>
</h2>
<hr class="half-width">
<p><a href="fig/03-01-02.jpeg">
<img src="fig/03-01-02.jpeg" alt="Photography of a pond in Cuatro Ciénegas" class="figure"></a></p>
<p>During this lesson, we will work with actual metagenomic information,
so we should be familiarized with it. The metagenomes that we will use
were collected in Cuatro Ciénegas, a region that has been extensively
studied by <a href="https://es.wikipedia.org/wiki/Valeria_Souza_Saldivar" class="external-link">Valeria
Souza</a>. Cuatro Ciénegas is an oasis in the Mexican desert whose
environmental conditions are often linked to the ones present in <a href="https://elifesciences.org/articles/38278" class="external-link">ancient seas</a>, due to
a higher-than-average content of sulfur and magnesium but a lower
concentrations of phosphorus and other nutrients. Because of these
particular conditions, the Cuatro Ciénegas basin is a fascinating place
to conduct a metagenomic study to learn more about the bacterial
diversity that is capable to survive and thrive in that environment.</p>
<p>The particular metagenomic study that we are going to work with was
collected in a <a href="https://elifesciences.org/articles/49816" class="external-link">study
about the response of the Cuatro Cienegas’ bacterial community to
nutrient enrichment.</a> In this study, authors compared the differences
between the microbial community in its natural, oligotrophic,
phosphorus-deficient environment, a pond from the Cuatro Ciénegas Basin
(CCB), and the same microbial community under a fertilization treatment.
The comparison between bacterial communities showed that many genomic
traits, such as mean bacterial genome size, GC content, total number of
tRNA genes, total number of rRNA genes, and codon usage bias were
significantly changed when the bacterial community underwent the
treatment.</p>
<div id="exercise-1-reviewing-metadata" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-reviewing-metadata" class="callout-inner">
<h3 class="callout-title">Exercise 1: Reviewing metadata</h3>
<div class="callout-content">
<p>According to the results described for this CCB study.</p>
<ol style="list-style-type: decimal">
<li>What kind of sequencing method do you think they used, and why do
you think so?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Metabarcoding<br>
</li>
<li>Shotgun metagenomics<br>
</li>
<li>Genomics of axenic cultures</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>In the table <a href="https://github.com/carpentries-incubator/metagenomics/blob/gh-pages/files/Samples_treatment_information.tsv" class="external-link">samples
treatment information</a>, what was the most critical piece of metadata
that the authors took?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha">
<li>Metabarcoding. False. With this technique, usually, only one region
of the genome is amplified.<br>
</li>
<li>Shotgun Metagenomics. True. Only shotgun metagenomics could have
been used to investigate the total number of tRNA genes.<br>
</li>
<li>Genomics of axenic cultures. False. Information on the microbial
community cannot be fully obtained with axenic cultures.</li>
</ol>
<p>The most crucial thing to know about our data is which community was
and was not supplemented with fertilizers.<br>
However, any differences in the technical parts of the study, such as
the DNA extraction protocol, could have affected the results, so
tracking those is also essential.</p>
</div>
</div>
</div>
</div>
<div id="exercise-2-differentiate-between-ids-and-sample-names" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-differentiate-between-ids-and-sample-names" class="callout-inner">
<h3 class="callout-title">Exercise 2: Differentiate between IDs and sample names</h3>
<div class="callout-content">
<p>Depending on the database, several IDs can be used for the same
sample. Please open the document where the <a href="https://github.com/carpentries-incubator/metagenomics/blob/gh-pages/files/Samples_treatment_information.tsv" class="external-link">metadata
information is stored</a>. Here, inspect the IDs and find out which of
them correspond to sample <strong>JP4110514WATERRESIZE</strong></p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>ERS1949771 is the SRA ID corresponding to JP4110514WATERRESIZE</p>
</div>
</div>
</div>
</div>
<div id="exercise-3-discuss-the-importance-of-metadata" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3-discuss-the-importance-of-metadata" class="callout-inner">
<h3 class="callout-title">Exercise 3: Discuss the importance of metadata</h3>
<div class="callout-content">
<p>Which other information could you recommend to add in the
metadata?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>Metadata will depend on the type of the experiment, but some examples
are the properties of the water before and after fertilization,
sampling, and processing methodology, date and time, place (country,
state, region, city, etc.).</p>
</div>
</div>
</div>
</div>
<p>Throughout the lesson, we will use the first four characters of the
<code>File names (alias)</code> to identify the data files corresponding
to a sample. We are going to use the first two sapmples for most of the
lesson and the third one for one exercise at the end.</p>
<table class="table">
<colgroup>
<col width="13%">
<col width="55%">
<col width="18%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th>SRA Accession</th>
<th>File name (alias)</th>
<th>Sample name in the lesson</th>
<th>Treatment</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ERS1949784</td>
<td>JC1ASEDIMENT120627</td>
<td>JC1A</td>
<td>Control mesocosm</td>
</tr>
<tr class="even">
<td>ERS1949801</td>
<td>JP4DASH2120627WATERAMPRESIZED</td>
<td>JP4D</td>
<td>Fertilized pond</td>
</tr>
<tr class="odd">
<td>ERS1949771</td>
<td>JP4110514WATERRESIZE</td>
<td>JP41</td>
<td>Unenriched pond</td>
</tr>
</tbody>
</table>
<p>The results of this study, raw sequences, and metadata have been
submitted to the NCBI Sequence Read Archive (SRA) and stored in the
BioProject <a href="https://www.ncbi.nlm.nih.gov/sra/?term=PRJEB22811" class="external-link">PRJEB22811</a>.</p>
<div id="other-metagenomic-databases" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="other-metagenomic-databases" class="callout-inner">
<h3 class="callout-title">Other metagenomic databases</h3>
<div class="callout-content">
<p>The NCBI SRA is not the only repository for metagenomic information.
There are other public metagenomic databases such as <a href="https://www.mg-rast.org/index.html?stay=1" class="external-link">MG-RAST</a>, <a href="https://www.ebi.ac.uk/metagenomics/" class="external-link">MGnify</a>, <a href="https://mmp.sfb.uit.no/" class="external-link">Marine Metagenomics Portal</a>, <a href="https://webapp.ufz.de/tmdb/" class="external-link">Terrestrial Metagenome DB</a> and the
<a href="https://gmrepo.humangut.info/home" class="external-link">GM Repo</a>.</p>
</div>
</div>
</div>
<p>Each database requires certain metadata linked with the data. As an
example, when <code>JP4D.fasta</code> is uploaded to mg-RAST the
associated metadata looks like this:</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="80%">
</colgroup>
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>file_name</td>
<td>JP4D.fasta</td>
</tr>
<tr class="even">
<td>investigation_type</td>
<td>metagenome</td>
</tr>
<tr class="odd">
<td>seq_meth</td>
<td>Illumina</td>
</tr>
<tr class="even">
<td>project_description</td>
<td>This project is a teaching project and uses data from Okie et
al. Elife 2020</td>
</tr>
<tr class="odd">
<td>collection_date</td>
<td>2012-06-27</td>
</tr>
<tr class="even">
<td>country</td>
<td>Mexico</td>
</tr>
<tr class="odd">
<td>feature</td>
<td>pond water</td>
</tr>
<tr class="even">
<td>latitude</td>
<td>26.8717055555556</td>
</tr>
<tr class="odd">
<td>longitude</td>
<td>-102.14</td>
</tr>
<tr class="even">
<td>env_package</td>
<td>water</td>
</tr>
<tr class="odd">
<td>depth</td>
<td>0.165</td>
</tr>
</tbody>
</table>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Shotgun metagenomics can be used for taxonomic and functional
studies.</li>
<li>Metabarcoding can be used for taxonomic studies.</li>
<li>Collecting metadata beforehand is fundamental for downstream
analysis.</li>
<li>We will use data from a Cuatro Ciénegas project to learn about
shotgun metagenomics.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-02-assessing-read-quality"><p>Content from <a href="02-assessing-read-quality.html">Assessing Read Quality</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/02-assessing-read-quality.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I describe the quality of my data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how a FASTQ file encodes per-base quality scores.</li>
<li>Interpret a FastQC plot summarizing per-base quality across all
reads.</li>
<li>Use <code>for</code> loops to automate operations on multiple
files.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="bioinformatic-workflows">Bioinformatic workflows<a class="anchor" aria-label="anchor" href="#bioinformatic-workflows"></a>
</h2>
<hr class="half-width">
<p>When working with high-throughput sequencing data, the raw reads you
get off the sequencer must pass through several different tools to
generate your final desired output. The execution of this set of tools
in a specified order is commonly referred to as a <em>workflow</em> or a
<em>pipeline</em>.</p>
<p>An example of the workflow we will be using for our analysis is
provided below, with a brief description of each step.</p>
<p><a href="fig/03-02-01.png">
<img src="fig/03-02-01.png" alt="Flow diagram that shows the steps: Sequence reads, Quality control, Assembly, Binning and Taxonomy" class="figure"></a></p>
<ol style="list-style-type: decimal">
<li>Quality control - Assessing quality using FastQC and Trimming and/or
filtering reads (if necessary)</li>
<li>Assembly of metagenome</li>
<li>Binning</li>
<li>Taxonomic assignation</li>
</ol>
<p>These workflows in bioinformatics adopt a plug-and-play approach in
that the output of one tool can be easily used as input to another tool
without any extensive configuration. Having standards for data formats
is what makes this feasible. Standards ensure that data is stored in a
way that is generally accepted and agreed upon within the community.
Therefore, the tools used to analyze data at different workflow stages
are<br>
built, assuming that the data will be provided in a specific format.</p>
</section><section><h2 class="section-heading" id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<hr class="half-width">
<p>We will now assess the quality of the sequence reads contained in our
FASTQ files.</p>
<p><a href="fig/03-02-02.png">
<img src="fig/03-02-02.png" alt="Flow diagram that shows the steps: Sequence reads and Quality control." class="figure"></a></p>
<div class="section level3">
<h3 id="details-on-the-fastq-format">Details on the FASTQ format<a class="anchor" aria-label="anchor" href="#details-on-the-fastq-format"></a>
</h3>
<p>Although it looks complicated (and it is), we can understand the <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">FASTQ</a> format with
a little decoding. Some rules about the format include the
following:</p>
<table class="table">
<colgroup>
<col width="3%">
<col width="96%">
</colgroup>
<thead><tr class="header">
<th>Line</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Always begins with ‘@’ followed by the information about the
read</td>
</tr>
<tr class="even">
<td>2</td>
<td>The actual DNA sequence</td>
</tr>
<tr class="odd">
<td>3</td>
<td>Always begins with a ‘+’ and sometimes contains the same info as in
line 1</td>
</tr>
<tr class="even">
<td>4</td>
<td>Has a string of characters which represent the quality scores; must
have same number of characters as line 2</td>
</tr>
</tbody>
</table>
<p>We can view the first complete read in one of the files from our
dataset using <code>head</code> to look at the first four lines. But we
have to decompress one of the files first.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq/</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">$</span> gunzip JP4D_R1.fastq.gz</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n</span> 4 JP4D_R1.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:12622:2006 1:N:0:CTCAGA
CCCGTTCCTCGGGCGTGCAGTCGGGCTTGCGGTCTGCCATGTCGTGTTCGGCGTCGGTGGTGCCGATCAGGGTGAAATCCGTCTCGTAGGGGATCGCGAAGATGATCCGCCCGTCCGTGCCCTGAAAGAAATAGCACTTGTCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCTCAGAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAAGCAAACCTCTCACTCCCTCTACTCTACTCCCTT
+
A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### </code></pre>
</div>
<p>Line 4 shows the quality of each nucleotide in the read. Quality is
interpreted as the probability of an incorrect base call (e.g., 1 in 10)
or, equivalently, the base call accuracy (e.g., 90%). Each nucleotide’s
numerical score’s value is converted into a character code where every
single character represents a quality score for an individual
nucleotide. This conversion allows the alignment of each individual
nucleotide with its quality score. For example, in the line above, the
quality score line is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### </code></pre>
</div>
<p>The numerical value assigned to each character depends on the
sequencing platform that generated the reads. The sequencing machine
used to generate our data uses the standard Sanger quality PHRED score
encoding, using Illumina version 1.8 onwards. Each character is assigned
a quality score between 0 and 41, as shown in the chart below.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Quality encoding: !"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJ
                   |         |         |         |         |
Quality score:    01........11........21........31........41                                </code></pre>
</div>
<p>Each quality score represents the probability that the corresponding
nucleotide call is incorrect. These probability values are the results
of the base calling algorithm and depend on how much signal was captured
for the base incorporation. This quality score is logarithmically based,
so a quality score of 10 reflects a base call accuracy of 90%, but a
quality score of 20 reflects a base call accuracy of 99%. In this <a href="https://drive5.com/usearch/manual/quality_score.html" class="external-link">link</a> you
can find more information about quality scores.</p>
<p>Looking back at our read:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:12622:2006 1:N:0:CTCAGA
CCCGTTCCTCGGGCGTGCAGTCGGGCTTGCGGTCTGCCATGTCGTGTTCGGCGTCGGTGGTGCCGATCAGGGTGAAATCCGTCTCGTAGGGGATCGCGAAGATGATCCGCCCGTCCGTGCCCTGAAAGAAATAGCACTTGTCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCTCAGAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAAGCAAACCTCTCACTCCCTCTACTCTACTCCCTT
+
A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### </code></pre>
</div>
<p>We can now see that there is a range of quality scores but that the
end of the sequence is very poor (<code>#</code> = a quality score of
2).</p>
<div id="exercise-1-looking-at-specific-reads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-looking-at-specific-reads" class="callout-inner">
<h3 class="callout-title">Exercise 1: Looking at specific reads</h3>
<div class="callout-content">
<p>In the terminal, how would you show the ID and quality of the last
read <code>JP4D_R1.fastq</code>?<br>
a) <code>tail JP4D_R1.fastq</code><br>
b) <code>head -n 4 JP4D_R1.fastq</code><br>
c) <code>more JP4D_R1.fastq</code><br>
d) <code>tail -n4 JP4D_R1.fastq</code><br>
e) <code>tail -n4 JP4D_R1.fastq | head -n2</code></p>
<p>Do you trust the sequence in this read?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>  </span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">a</span><span class="er">)</span> <span class="ex">It</span> shows the ID and quality of the last read but also unnecessary lines from previous reads.  </span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="ex">b</span><span class="er">)</span> <span class="ex">No.</span> It shows the first read<span class="st">'s info.  </span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="st">c) It shows the text of the entire file.  </span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="st">d) This option is the best answer as it only shows the last read'</span>s information.  </span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="ex">e</span><span class="er">)</span> <span class="ex">It</span> does show the ID of the last read but not the quality.  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@MISEQ-LAB244-W7:156:000000000-A80CV:1:2114:17866:28868 1:N:0:CTCAGA

CCCGTTCTCCACCTCGGCGCGCGCCAGCTGCGGCTCGTCCTTCCACAGGAACTTCCACGTCGCCGTCAGCCGCGACACGTTCTCCCCCCTCGCATGCTCGTCCTGTCTCTCGTGCTTGGCCGACGCCTGCGCCTCGCACTGCGCCCGCTCGGTGTCGTTCATGTTGATCTTCACCGTGGCGTGCATGAAGCGGTTCCCGGCCTCGTCGCCACCCACGCCATCCGCGTCGGCCAGCCACTCTCACTGCTCGC

+

AA11AC1&gt;3@DC1F1111000A0/A///BB#############################################################################################################################################################################################################################          </code></pre>
</div>
<p>This read has more consistent quality at its first than at the end
but still has a range of quality scores, most of them are low. We will
look at variations in position-based quality in just a moment.</p>
</div>
</div>
</div>
</div>
<p>In real life, you won’t be assessing the quality of your reads by
visually inspecting your FASTQ files. Instead, you’ll use a software
program to assess read quality and filter out poor reads. We’ll first
use a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
to visualize the quality of our reads. Later in our workflow, we’ll use
another program to filter out poor-quality reads.</p>
<p>First, let’s make available our metagenomics software:</p>
</div>
</section><section><h2 class="section-heading" id="activating-an-environment">Activating an environment<a class="anchor" aria-label="anchor" href="#activating-an-environment"></a>
</h2>
<hr class="half-width">
<p>Environments are part of a bioinformatic tendency to do reproducible
research; they are a way to share and maintain our programs in their
needed versions used for a pipeline with our colleagues and our future
self. FastQC has not been activated in the (base) environment, but this
AWS instance came with an environment called metagenomics. We need to
activate it in order to start using FastQC.</p>
<p>We will use <a href="https://docs.conda.io/en/latest/" class="external-link">Conda</a> as
our environment manager. Conda is an open-source package and environment
management system that runs on Windows, macOS and Linux. Conda
environments are activated with the <code>conda activate</code>
direction:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> conda activate metagenomics  </span></code></pre>
</div>
<p>After the environment has been activated, a label is shown before the
<code>$</code> sign.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>(metagenomics) $</code></pre>
</div>
<p>Now, if we call FastQC, a long help page will be displayed on our
screen.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="at">-h</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

        fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam]
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of many different modules, each one of
    which will help to identify a different potential type of problem in your
    data.
.
.
.</code></pre>
</div>
<p>If FastQC is not installed, then you would expect to see an error
like</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>The program 'fastqc' is currently not installed. You can install it by typing:
sudo apt-get install fastqc</code></pre>
</div>
<p>If this happens, check with your instructor before trying to install
it.</p>
</section><section><h2 class="section-heading" id="assessing-quality-using-fastqc">Assessing quality using FastQC<a class="anchor" aria-label="anchor" href="#assessing-quality-using-fastqc"></a>
</h2>
<hr class="half-width">
<p>FastQC has several features that can give you a quick impression of
any problems your data may have, so you can consider these issues before
moving forward with your analyses. Rather than looking at quality scores
for each read, FastQC looks at quality collectively across all reads
within a sample. The image below shows one FastQC-generated plot that
indicates a very high-quality sample:</p>
<p><a href="fig/03-02-03.png">
<img src="fig/03-02-03.png" alt="Quality graph showing a boxplot for each sequence position in the x-axis, and the Phred score in the y-axis. The background is colored red for the Phred scores 0 to 20, yellow for the scores 20 to 28, and green for the scores 28 to 38. All of the boxes for each position are in the green area." class="figure"></a></p>
<p>The x-axis displays the base position in the read, and the y-axis
shows quality scores. In this example, the sample contains reads that
are 40 bp long. This length is much shorter than the reads we are
working on within our workflow. For each position, there is a
box-and-whisker plot showing the distribution of quality scores for all
reads at that position. The horizontal red line indicates the median
quality score, and the yellow box shows the 1st to 3rd quartile range.
This range means that 50% of reads have a quality score that falls
within the range of the yellow box at that position. The whiskers show
the whole range covering the lowest (0th quartile) to highest (4th
quartile) values.</p>
<p>The quality values for each position in this sample do not drop much
lower than 32, which is a high-quality score. The plot background is
also color-coded to identify good (green), acceptable (yellow) and bad
(red) quality scores.</p>
<p>Now let’s look at a quality plot on the other end of the
spectrum.</p>
<p><a href="fig/03-02-04.png">
<img src="fig/03-02-04.png" alt="Graphic of boxplots, where the first ones are in the good range of scores of the y-axis and extend to the acceptable and bad ranges of scores toward the right of the x-axis" class="figure"></a></p>
<p>The FastQC tool produces several other diagnostic plots to assess
sample quality and the one plotted above. Here, we see positions within
the read in which the boxes span a much more comprehensive range. Also,
quality scores drop pretty low into the “bad” range, particularly on the
tail end of the reads.</p>
</section><section><h2 class="section-heading" id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a>
</h2>
<hr class="half-width">
<p>We will now assess the quality of the reads that we downloaded.
First, make sure you’re still in the <code>untrimmed_fastq</code>
directory.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq/ </span></code></pre>
</div>
<div id="exercise-2-looking-at-metadata-about-the-untrimmed-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-looking-at-metadata-about-the-untrimmed-files" class="callout-inner">
<h3 class="callout-title">Exercise 2: Looking at metadata about the untrimmed-files</h3>
<div class="callout-content">
<p>To know which files have more data, you need to see metadata about
untrimmed files. In files, metadata includes owners of the file, state
of the write, read, and execute permissions, size, and modification
date. Using the <code>ls</code> command, how would you get the size of
the files in the <code>untrimmed_fastq\</code> directory?<br>
(Hint: Look at the options for the <code>ls</code> command to see how to
show file sizes.)<br>
a) <code>ls -a</code><br>
b) <code>ls -S</code><br>
c) <code>ls -l</code><br>
d) <code>ls -lh</code><br>
e) <code>ls -ahlS</code></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol style="list-style-type: lower-alpha">
<li>No. The flag <code>-a</code> shows all the contents, including
hidden files and directories, but not the sizes.<br>
</li>
<li>No. The flag <code>-S</code> shows the content Sorted by size,
starting with the most extensive file, but not the sizes.<br>
</li>
<li>Yes. The flag <code>-l</code> shows the contents with metadata,
including file size. Other metadata are permissions, owners, and
modification dates.<br>
</li>
<li>Yes. The flag <code>-lh</code> shows the content with metadata in a
human-readable manner.<br>
</li>
<li>Yes. The combination of all the flags shows all the contents with
metadata, including hidden files, sorted by size.</li>
</ol>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-ahls</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-rw-r--r-- 1 dcuser dcuser  24M Nov 26 21:34 JC1A_R1.fastq.gz
-rw-r--r-- 1 dcuser dcuser  24M Nov 26 21:34 JC1A_R2.fastq.gz
-rw-r--r-- 1 dcuser dcuser 616M Nov 26 21:34 JP4D_R1.fastq
-rw-r--r-- 1 dcuser dcuser 203M Nov 26 21:35 JP4D_R2.fastq.gz   </code></pre>
</div>
<p>Four FASTQ files oscillate between 24M (24MB) to 616M. The largest
file is JP4D_R1.fastq with 616M.</p>
</div>
</div>
</div>
</div>
<p>FastQC can accept multiple file names as input, and on both zipped
and unzipped files, so we can use the <code>\*.fastq*</code>wildcard to
run FastQC on all FASTQ files in this directory.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="pp">*</span>.fastq<span class="pp">*</span> </span></code></pre>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Started analysis of JC1A_R1.fastq.gz
Approx 5% complete for JC1A_R1.fastq.gz
Approx 10% complete for JC1A_R1.fastq.gz
Approx 15% complete for JC1A_R1.fastq.gz
Approx 20% complete for JC1A_R1.fastq.gz
Approx 25% complete for JC1A_R1.fastq.gz
Approx 30% complete for JC1A_R1.fastq.gz
Approx 35% complete for JC1A_R1.fastq.gz  </code></pre>
</div>
<p>It should take around five minutes for FastQC to run on all four of
our FASTQ files. When the analysis completes, your prompt will return.
So your screen will look something like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Approx 80% complete for JP4D_R2.fastq.gz
Approx 85% complete for JP4D_R2.fastq.gz
Approx 90% complete for JP4D_R2.fastq.gz
Approx 95% complete for JP4D_R2.fastq.gz
Analysis complete for JP4D_R2.fastq.gz
$</code></pre>
</div>
<p>The FastQC program has created several new files within our
<code>data/untrimmed_fastq/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">$</span> ls </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc.html  JC1A_R2_fastqc.html  JP4D_R1.fastq        JP4D_R2_fastqc.html  TruSeq3-PE.fa
JC1A_R1_fastqc.zip   JC1A_R2_fastqc.zip   JP4D_R1_fastqc.html  JP4D_R2_fastqc.zip
JC1A_R1.fastq.gz     JC1A_R2.fastq.gz     JP4D_R1_fastqc.zip   JP4D_R2.fastq.gz    </code></pre>
</div>
<p>For each input FASTQ file, FastQC has created a <code>.zip</code>
file and a <code>.html</code> file. The <code>.zip</code> file extension
indicates that this is a compressed set of multiple output files. We’ll
be working with these output files soon. The <code>.html</code> file is
a stable webpage displaying the summary report for each of our
samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into a new directory within our
<code>results/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/dc_workshop/results/fastqc_untrimmed_reads </span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.zip ~/dc_workshop/results/fastqc_untrimmed_reads/ </span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.html ~/dc_workshop/results/fastqc_untrimmed_reads/ </span></code></pre>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/results/fastqc_untrimmed_reads/ </span></code></pre>
</div>
</section><section><h2 class="section-heading" id="viewing-the-fastqc-results">Viewing the FastQC results<a class="anchor" aria-label="anchor" href="#viewing-the-fastqc-results"></a>
</h2>
<hr class="half-width">
<p>If we were working on our local computers, we’d be able to look at
each of these HTML files by opening them in a web browser. However,
these files are currently sitting on our remote AWS instance, where our
local computer can’t see them. If we are working with the terminal
provided by r-studio we can either: select the html files and with the
secondary click chose the option open in a browser or export the files
to our local computer as we learned in the Introduction to the Command
Line lesson.</p>
<div id="exercise-3-downloading-files-by-scp-optional" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3-downloading-files-by-scp-optional" class="callout-inner">
<h3 class="callout-title">Exercise 3: Downloading files by scp (optional)</h3>
<div class="callout-content">
<p>Suppose you only have access to a terminal and there is not a web
browser available at the AWS remote machine. The following are the steps
needed to download your files to your computer. Observe the third step
and answer false/true for each question about the <code>scp</code>
command.</p>
<ol style="list-style-type: decimal">
<li>Open a new terminal on your local computer.</li>
<li>Make a new local directory on our computer to store the HTML files
<code>$ mkdir -p ~/Desktop/fastqc_html</code>
</li>
<li>Transfer our HTML files to our local computer using
<code>scp</code>.
<code>$ scp dcuser@ec2-34-238-162-94.compute-1.amazonaws.com:~/dc_workshop/results/fastqc_untrimmed_reads/*.html ~/Desktop/fastqc_html</code><br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>
<code>dcuser</code> is your local user<br>
</li>
<li>
<code>ec2-34-238-162-94.compute-1.amazonaws.com</code> is the
address of your remote machine<br>
</li>
<li>the current adress of the file goes after the second space in the
<code>scp</code> command.<br>
</li>
<li>
<code>~/dc_workshop/results/fastqc_untrimmed_reads/*.html</code> is
the path of the file you want to download<br>
</li>
<li>
<code>~/Desktop/fastqc_html</code> is a remote path.<br>
</li>
<li>
<code>:</code> Divides the host name of your local computer and the
path of the file.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha">
<li>False. <code>dcuser</code> is your remote user.<br>
</li>
<li>True. <code>ec2-34-238-162-94.compute-1.amazonaws.com</code> is the
adress of your remote machine<br>
</li>
<li>False. The current adress of the file goes after the first space in
the <code>scp</code> command.<br>
</li>
<li>True.
<code>~/dc_workshop/results/fastqc_untrimmed_reads/*.html</code> is the
path of the file you want to download in the remote machine.</li>
<li>False. <code>~/Desktop/fastqc_html</code> is a local path where your
file will be downloaded.</li>
<li>False. <code>:</code> Divides the host name of a <em>remote</em>
computer and the path of the file on the remote computer.</li>
</ol>
<p>You should see a status output like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc.html     100%  253KB 320.0KB/s   00:00
JC1A_R2_fastqc.html     100%  262KB 390.1KB/s   00:00
JP4D_R1_fastqc.html     100%  237KB 360.8KB/s   00:00
JP4D_R2_fastqc.html     100%  244KB 385.2KB/s   00:00</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Now we can open the 4 HTML files.</p>
<p>Depending on your system, you should be able to select and open them
all at once via a right-click menu in your file browser.</p>
<div id="exercise-4-discuss-the-quality-of-sequencing-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4-discuss-the-quality-of-sequencing-files" class="callout-inner">
<h3 class="callout-title">Exercise 4: Discuss the quality of sequencing files</h3>
<div class="callout-content">
<p>Discuss your results with a neighbor. Which sample(s) looks the best
per base sequence quality? Which sample(s) look the worst?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>All of the reads contain usable data, but the quality decreases
toward the end of the reads. File JC1A_R2_fastqc shows the lowest
quality.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="decoding-the-other-fastqc-outputs">Decoding the other FastQC outputs<a class="anchor" aria-label="anchor" href="#decoding-the-other-fastqc-outputs"></a>
</h2>
<hr class="half-width">
<p>We’ve now looked at quite a few “Per base sequence quality” FastQC
graphs, but there are nine other graphs that we haven’t talked about!
Below we have provided a brief overview of interpretations for each
plot. For more information, please see the FastQC documentation <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/" class="external-link">here</a></p>
<ul>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html" class="external-link"><strong>Per
tile sequence quality</strong></a>: the machines that perform sequencing
are divided into tiles. This plot displays patterns in base quality
along these tiles. Consistently low scores are often found around the
edges, but hot spots could also occur in the middle if an air bubble was
introduced during the run.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/3%20Per%20Sequence%20Quality%20Scores.html" class="external-link"><strong>Per
sequence quality scores</strong></a>: a density plot of quality for all
reads at all positions. This plot shows what quality scores are most
common.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html" class="external-link"><strong>Per
base sequence content</strong></a>: plots the proportion of each base
position over all of the reads. Typically, we expect to see each base
roughly 25% of the time at each position, but this often fails at the
beginning or end of the read due to quality or adapter content.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/5%20Per%20Sequence%20GC%20Content.html" class="external-link"><strong>Per
sequence GC content</strong></a>: a density plot of average GC content
in each of the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/6%20Per%20Base%20N%20Content.html" class="external-link"><strong>Per
base N content</strong></a>: the percent of times that ‘N’ occurs at a
position in all reads. If there is an increase at a particular position,
this might indicate that something went wrong during sequencing.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html" class="external-link"><strong>Sequence
Length Distribution</strong></a>: the distribution of sequence lengths
of all reads in the file. If the data is raw, there is often a sharp
peak; however, if the reads have been trimmed, there may be a
distribution of shorter lengths.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/8%20Duplicate%20Sequences.html" class="external-link"><strong>Sequence
Duplication Levels</strong></a>: a distribution of duplicated sequences.
In sequencing, we expect most reads to only occur once. If some
sequences are occurring more than once, it might indicate enrichment
bias (e.g. from PCR). This might not be true if the samples are high
coverage (or RNA-seq or amplicon).</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/9%20Overrepresented%20Sequences.html" class="external-link"><strong>Overrepresented
sequences</strong></a>: a list of sequences that occur more frequently
than would be expected by chance.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/10%20Adapter%20Content.html" class="external-link"><strong>Adapter
Content</strong></a>: a graph indicating where adapter sequences occur
in the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/11%20Kmer%20Content.html" class="external-link"><strong>K-mer
Content</strong></a>: a graph showing any sequences which may show a
positional bias within the reads.</li>
</ul></section><section><h2 class="section-heading" id="working-with-the-fastqc-text-output">Working with the FastQC text output<a class="anchor" aria-label="anchor" href="#working-with-the-fastqc-text-output"></a>
</h2>
<hr class="half-width">
<p>Now that we’ve looked at our HTML reports getting a feel for the
data, let’s look more closely at the other output files. Go back to the
tab in your terminal program that is connected to your AWS instance (the
tab label will start with <code>dcuser@ip</code>) and make sure you’re
in our results subdirectory.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/results/fastqc_untrimmed_reads/ </span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="ex">$</span> ls </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc.html           JP4D_R1_fastqc.html
JC1A_R1_fastqc.zip            JP4D_R1_fastqc.zip
JC1A_R2_fastqc.html           JP4D_R2_fastqc.html
JC1A_R2_fastqc.zip            JP4D_R2_fastqc.zip </code></pre>
</div>
<p>Our <code>.zip</code> files are compressed files. Each contains
multiple different types of output files for a single input FASTQ file.
To view the contents of a <code>.zip</code> file, we can use the program
<code>unzip</code> to decompress these files. Let’s try doing them all
at once using a wildcard.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">$</span> unzip <span class="pp">*</span>.zip </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Archive:  JC1A_R1_fastqc.zip
caution: filename not matched:  JC1A_R2_fastqc.zip
caution: filename not matched:  JP4D_R1_fastqc.zip
caution: filename not matched:  JP4D_R2_fastqc.zip  </code></pre>
</div>
<p>This decompression didn’t work. It identified the first file and got
a warning message for the other <code>.zip</code> files. This is because
<code>unzip</code> expects to get only one zip file as input. We could
go through and unzip each file one at a time, but this is very
time-consuming and error-prone. Someday you may have 500 files to
unzip!</p>
<p>A more efficient way is to use a <code>for</code> loop like we
learned in the Command Line lesson to iterate through all of our
<code>.zip</code> files. Let’s see what that looks like, and then we’ll
discuss what we’re doing with each line of our loop.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="ex">$</span> for filename in <span class="pp">*</span>.zip</span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="op">&gt;</span> unzip <span class="va">$filename</span></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>In this example, the input is the four filenames (one filename for
each of our <code>.zip</code> files). Each time the loop iterates, it
will assign a file name to the variable <code>filename</code> and run
the <code>unzip</code> command. The first time through the loop,
<code>$filename</code> is <code>JC1A_R1_fastqc.zip</code>. The
interpreter runs the command <code>unzip</code> on
<code>JC1A_R1_fastqc.zip</code>. For the second iteration,
<code>$filename</code> becomes <code>JC1A_R2_fastqc.zip</code>. This
time, the shell runs <code>unzip</code> on
<code>JC1A_R2_fastqc.zip</code>. It then repeats this process for the
other <code>.zip</code> files in our directory.</p>
<p>When we run the <code>for</code> loop, you will see an output that
starts like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">Archive</span><span class="op">:</span>  <span class="va">JC1A_R1_fastqc.zip</span></span>
<span><span class="va">creating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span></span>
<span><span class="va">creating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span></span>
<span><span class="va">creating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">fastqc_icon.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">warning.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">error.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">tick.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">summary.txt</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_quality.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_tile_quality.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_quality.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_sequence_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_gc_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_n_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">sequence_length_distribution.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">duplication_levels.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">adapter_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">fastqc_report.html</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">fastqc_data.txt</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">fastqc.fo</span>  </span></code></pre>
</div>
<p>The <code>unzip</code> program is decompressing the <code>.zip</code>
files and creates a new directory (with subdirectories) for each of our
samples, to store all of the different output that is produced by
FastQC. There are a lot of files here. We’re going to focus on the
<code>summary.txt</code> file.</p>
<p>If you list the files in our directory, now you will see the
following:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">$</span> ls </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc                  JP4D_R1_fastqc
JC1A_R1_fastqc.html             JP4D_R1_fastqc.html
JC1A_R1_fastqc.zip              JP4D_R1_fastqc.zip
JC1A_R2_fastqc                  JP4D_R2_fastqc
JC1A_R2_fastqc.html             JP4D_R2_fastqc.html
JC1A_R2_fastqc.zip              JP4D_R2_fastqc.zip                                         </code></pre>
</div>
<p>The <code>.html</code> files and the uncompressed <code>.zip</code>
files are still present, but now we also have a new directory for each
sample. We can see that it’s a directory if we use the <code>-F</code>
flag for <code>ls</code>.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc/                  JP4D_R1_fastqc/
JC1A_R1_fastqc.html             JP4D_R1_fastqc.html
JC1A_R1_fastqc.zip              JP4D_R1_fastqc.zip
JC1A_R2_fastqc/                  JP4D_R2_fastqc/
JC1A_R2_fastqc.html             JP4D_R2_fastqc.html
JC1A_R2_fastqc.zip              JP4D_R2_fastqc.zip                                         </code></pre>
</div>
<p>Let’s see what files are present within one of these output
directories.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span> JC1A_R1_fastqc/ </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fastqc_data.txt  fastqc.fo  fastqc_report.html	Icons/	Images/  summary.txt</code></pre>
</div>
<p>Use <code>less</code> to preview the <code>summary.txt</code> file
for this sample.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="ex">$</span> less JC1A_R1_fastqc/summary.txt </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>PASS    Basic Statistics        JC1A_R1.fastq.gz
FAIL    Per base sequence quality       JC1A_R1.fastq.gz
PASS    Per tile sequence quality       JC1A_R1.fastq.gz
PASS    Per sequence quality scores     JC1A_R1.fastq.gz
WARN    Per base sequence content       JC1A_R1.fastq.gz
FAIL    Per sequence GC content JC1A_R1.fastq.gz
PASS    Per base N content      JC1A_R1.fastq.gz
PASS    Sequence Length Distribution    JC1A_R1.fastq.gz
FAIL    Sequence Duplication Levels     JC1A_R1.fastq.gz
PASS    Overrepresented sequences       JC1A_R1.fastq.gz
FAIL    Adapter Content JC1A_R1.fastq.gz  </code></pre>
</div>
<p>The summary file gives us a list of tests that FastQC ran and tells
us whether this sample passed, failed, or is borderline
(<code>WARN</code>). Remember, to quit from <code>less</code>, you must
type <code>q</code>.</p>
</section><section><h2 class="section-heading" id="documenting-our-work">Documenting our work<a class="anchor" aria-label="anchor" href="#documenting-our-work"></a>
</h2>
<hr class="half-width">
<p>We can make a record of the results we obtained for all our samples
by concatenating all of our <code>summary.txt</code> files into a single
file using the <code>cat</code> command. We’ll call this
<code>fastqc_summaries.txt</code> and store it to
<code>~/dc_workshop/docs</code>.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/dc_workshop/docs</span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a><span class="ex">$</span> cat <span class="pp">*</span>/summary.txt <span class="op">&gt;</span> ~/dc_workshop/docs/fastqc_summaries.txt</span></code></pre>
</div>
<div id="exercise-4-quality-tests" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4-quality-tests" class="callout-inner">
<h3 class="callout-title">Exercise 4: Quality tests</h3>
<div class="callout-content">
<p>Which samples failed at least one of FastQC’s quality tests? What
test(s) did those samples failed</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>We can get the list of all failed tests using <code>grep</code>.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/docs</span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a><span class="ex">$</span> grep FAIL fastqc_summaries.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>FAIL    Per base sequence quality       JC1A_R1.fastq.gz
FAIL    Per sequence GC content JC1A_R1.fastq.gz
FAIL    Sequence Duplication Levels     JC1A_R1.fastq.gz
FAIL    Adapter Content JC1A_R1.fastq.gz
FAIL    Per base sequence quality       JC1A_R2.fastq.gz
FAIL    Per sequence GC content JC1A_R2.fastq.gz
FAIL    Sequence Duplication Levels     JC1A_R2.fastq.gz
FAIL    Adapter Content JC1A_R2.fastq.gz
FAIL    Per base sequence content       JP4D_R1.fastq
FAIL    Adapter Content JP4D_R1.fastq
FAIL    Per base sequence quality       JP4D_R2.fastq.gz
FAIL    Per base sequence content       JP4D_R2.fastq.gz
FAIL    Adapter Content JP4D_R2.fastq.gz</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="quality-of-large-datasets" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="quality-of-large-datasets" class="callout-inner">
<h3 class="callout-title">Quality of large datasets</h3>
<div class="callout-content">
<p>Explore <a href="https://multiqc.info/" class="external-link">MultiQC</a> if you want a
tool that can show the quality of many samples at once.</p>
</div>
</div>
</div>
<div id="quality-encodings-vary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="quality-encodings-vary" class="callout-inner">
<h3 class="callout-title">Quality Encodings Vary</h3>
<div class="callout-content">
<p>Although we’ve used a particular quality encoding system to
demonstrate the interpretation of read quality, different sequencing
machines use different encoding systems. This means that depending on
which sequencer you use to generate your data, a <code>#</code> may not
indicate a poor quality base call.</p>
<p>This mainly relates to older Solexa/Illumina data. However, it’s
essential that you know which sequencing platform was used to generate
your data to tell your quality control program which encoding to use. If
you choose the wrong encoding, you run the risk of throwing away good
reads or (even worse) not throwing away bad reads!</p>
</div>
</div>
</div>
<div id="bonus-exercise-automating-a-quality-control-workflow" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="bonus-exercise-automating-a-quality-control-workflow" class="callout-inner">
<h3 class="callout-title">Bonus Exercise: Automating a quality control workflow</h3>
<div class="callout-content">
<p>If you lose your FastQC analysis results. How would you do it again
but faster than the first time? As we have seen in a previous lesson,
making scripts for repetitive tasks is a very efficient practice during
bioinformatic pipelines.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<p>Make a new script with nano</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="fu">nano</span> quality_control.sh</span></code></pre>
</div>
<p>Paste inside the commands that we used along with <code>echo</code>
commands that shows you how the script is running.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="bu">set</span> <span class="at">-e</span> <span class="co"># This will ensure that our script will exit if an error occurs</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a><span class="bu">cd</span> ~/dc_workshop/data/untrimmed_fastq/</span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Running FastQC ..."</span></span>
<span id="cb42-5"><a href="#cb42-5" tabindex="-1"></a><span class="ex">fastqc</span> <span class="pp">*</span>.fastq<span class="pp">*</span></span>
<span id="cb42-6"><a href="#cb42-6" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/dc_workshop/results/fastqc_untrimmed_reads</span>
<span id="cb42-8"><a href="#cb42-8" tabindex="-1"></a></span>
<span id="cb42-9"><a href="#cb42-9" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Saving FastQC results..."</span></span>
<span id="cb42-10"><a href="#cb42-10" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.zip ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb42-11"><a href="#cb42-11" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.html ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb42-12"><a href="#cb42-12" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" tabindex="-1"></a><span class="bu">cd</span> ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb42-14"><a href="#cb42-14" tabindex="-1"></a></span>
<span id="cb42-15"><a href="#cb42-15" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Unzipping..."</span></span>
<span id="cb42-16"><a href="#cb42-16" tabindex="-1"></a><span class="cf">for</span> filename <span class="kw">in</span> <span class="pp">*</span>.zip</span>
<span id="cb42-17"><a href="#cb42-17" tabindex="-1"></a>    <span class="cf">do</span></span>
<span id="cb42-18"><a href="#cb42-18" tabindex="-1"></a>    <span class="fu">unzip</span> <span class="va">$filename</span></span>
<span id="cb42-19"><a href="#cb42-19" tabindex="-1"></a>    <span class="cf">done</span></span>
<span id="cb42-20"><a href="#cb42-20" tabindex="-1"></a></span>
<span id="cb42-21"><a href="#cb42-21" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Saving summary..."</span></span>
<span id="cb42-22"><a href="#cb42-22" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/dc_workshop/docs</span>
<span id="cb42-23"><a href="#cb42-23" tabindex="-1"></a><span class="fu">cat</span> <span class="pp">*</span>/summary.txt <span class="op">&gt;</span> ~/dc_workshop/docs/fastqc_summaries.txt</span></code></pre>
</div>
<p>If we were to run this script, it would ask us for confirmation to
redo several steps because we already did all of them. If you want to,
you can run it to check that it works, but it is not necessary if you
already completed every step of the previous episode.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>It is important to know the quality of our data to make decisions in
the subsequent steps.</li>
<li>FastQC is a program that allows us to know the quality of FASTQ
files.</li>
<li>
<code>for</code> loops let you perform the same operations on
multiple files with a single command.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-trimming-filtering"><p>Content from <a href="03-trimming-filtering.html">Trimming and Filtering</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/03-trimming-filtering.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we get rid of sequence data that does not meet our quality
standards?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Clean FASTQ reads using Trimmomatic.</li>
<li>Select and set multiple options for command line bioinformatic
tools.</li>
<li>Write <code>for</code> loops with two variables.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="cleaning-reads">Cleaning reads<a class="anchor" aria-label="anchor" href="#cleaning-reads"></a>
</h2>
<hr class="half-width">
<p>In the last episode, we took a high-level look at the quality of each
of our samples using <code>FastQC</code>. We visualized per-base quality
graphs showing the distribution of the quality at each base across all
the reads from our sample. This information helps us to determine the
quality threshold we will accept, and thus, we saw information about
which samples fail which quality checks. Some of our samples failed
quite a few quality metrics used by FastQC. However, this does not mean
that our samples should be thrown out! It is common to have some quality
metrics fail, which may or may not be a problem for your downstream
application. For our workflow, we will remove some low-quality sequences
to reduce our false-positive rate due to sequencing errors.</p>
<p>To accomplish this, we will use a program called <a href="https://www.usadellab.org/cms/?page=trimmomatic" class="external-link">Trimmomatic</a>.
This useful tool filters poor quality reads and trims poor quality bases
from the specified samples.</p>
</section><section><h2 class="section-heading" id="trimmomatic-options">Trimmomatic options<a class="anchor" aria-label="anchor" href="#trimmomatic-options"></a>
</h2>
<hr class="half-width">
<p>Trimmomatic has a variety of options to accomplish its task. If we
run the following command, we can see some of its options:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> trimmomatic</span></code></pre>
</div>
<p>Which will give you the following output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Usage:
       PE [-version] [-threads &lt;threads&gt;] [-phred33|-phred64] [-trimlog &lt;trimLogFile&gt;] [-summary &lt;statsSummaryFile&gt;] [-quiet] [-validatePairs] [-basein &lt;inputBase&gt; | &lt;inputFile1&gt; &lt;inputFile2&gt;] [-baseout &lt;outputBase&gt; | &lt;outputFile1P&gt; &lt;outputFile1U&gt; &lt;outputFile2P&gt; &lt;outputFile2U&gt;] &lt;trimmer1&gt;...
   or:
       SE [-version] [-threads &lt;threads&gt;] [-phred33|-phred64] [-trimlog &lt;trimLogFile&gt;] [-summary &lt;statsSummaryFile&gt;] [-quiet] &lt;inputFile&gt; &lt;outputFile&gt; &lt;trimmer1&gt;...
   or:
       -version</code></pre>
</div>
<p>This output shows that we must first specify whether we have
paired-end (<code>PE</code>) or single-end (<code>SE</code>) reads.
Next, we will specify with which flags we want to run Trimmomatic. For
example, you can specify <code>threads</code> to indicate the number of
processors on your computer that you want Trimmomatic to use. In most
cases, using multiple threads(processors) can help to run the trimming
faster. These flags are unnecessary, but they can give you more control
over the command. The flags are followed by <strong>positional
arguments</strong>, meaning the order in which you specify them is
essential. In paired-end mode, Trimmomatic expects the two input files
and then the names of the output files. These files are described below.
While in single-end mode, Trimmomatic will expect one file as input,
after which you can enter the optional settings and, lastly, the name of
the output file.</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>Option</th>
<th>Meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>&lt;inputFile1&gt;</td>
<td>input forward reads to be trimmed. Typically the file name will
contain an <code>_1</code> or <code>_R1</code> in the name.</td>
</tr>
<tr class="even">
<td>&lt;inputFile2&gt;</td>
<td>Input reverse reads to be trimmed. Typically the file name will
contain an <code>_2</code> or <code>_R2</code> in the name.</td>
</tr>
<tr class="odd">
<td>&lt;outputFile1P&gt;</td>
<td>Output file that contains surviving pairs from the <code>_1</code>
file.</td>
</tr>
<tr class="even">
<td>&lt;outputFile1U&gt;</td>
<td>Output file that contains orphaned reads from the <code>_1</code>
file.</td>
</tr>
<tr class="odd">
<td>&lt;outputFile2P&gt;</td>
<td>Output file that contains surviving pairs from the <code>_2</code>
file.</td>
</tr>
<tr class="even">
<td>&lt;outputFile2U&gt;</td>
<td>Output file that contains orphaned reads from the <code>_2</code>
file.</td>
</tr>
</tbody>
</table>
<p>The last thing Trimmomatic expects to see is the trimming
parameters:</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>step</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ILLUMINACLIP</code></td>
<td>Perform adapter removal.</td>
</tr>
<tr class="even">
<td><code>SLIDINGWINDOW</code></td>
<td>Perform sliding window trimming, cutting once the average quality
within the window falls below a threshold.</td>
</tr>
<tr class="odd">
<td><code>LEADING</code></td>
<td>Cut bases off the start of a read if below a threshold quality.</td>
</tr>
<tr class="even">
<td><code>TRAILING</code></td>
<td>Cut bases off the end of a read if below a threshold quality.</td>
</tr>
<tr class="odd">
<td><code>CROP</code></td>
<td>Cut the read to a specified length.</td>
</tr>
<tr class="even">
<td><code>HEADCROP</code></td>
<td>Cut the specified number of bases from the start of the read.</td>
</tr>
<tr class="odd">
<td><code>MINLEN</code></td>
<td>Drop an entire read if it is below a specified length.</td>
</tr>
<tr class="even">
<td><code>TOPHRED33</code></td>
<td>Convert quality scores to Phred-33.</td>
</tr>
<tr class="odd">
<td><code>TOPHRED64</code></td>
<td>Convert quality scores to Phred-64.</td>
</tr>
</tbody>
</table>
<p>Understanding the steps you are using to clean your data is
essential. We will use only a few options and trimming steps in our
analysis. For more information about the Trimmomatic arguments and
options, see <a href="https://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf" class="external-link">the
Trimmomatic manual</a>.</p>
<p><a href="fig/03-03-01.png">
<img src="fig/03-03-01.png" alt="Diagram showing the parts of the sequence that are reviewed by each parameter and the parts that are maintained or discarded at the end of the process. The Illuminaclip parameter removes the adapters, and the SlidingWindow scans the read by sections and removes a part of the read below the quality threshold. We remain with a trimmed read that has a valid quality." class="figure"></a></p>
<p>However, a complete command for Trimmomatic will look something like
the command below. This command is an example and will not work, as we
do not have the files it refers to:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> trimmomatic PE <span class="at">-threads</span> 4 SRR_1056_1.fastq SRR_1056_2.fastq <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>              SRR_1056_1.trimmed.fastq SRR_1056_1un.trimmed.fastq <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>              SRR_1056_2.trimmed.fastq SRR_1056_2un.trimmed.fastq <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>              ILLUMINACLIP:SRR_adapters.fa SLIDINGWINDOW:4:20</span></code></pre>
</div>
<p>In this example, we have told Trimmomatic:</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>code</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>PE</code></td>
<td>that it will be taking a paired-end file as input</td>
</tr>
<tr class="even">
<td><code>-threads 4</code></td>
<td>to use four computing threads to run (this will speed up our
run)</td>
</tr>
<tr class="odd">
<td><code>SRR_1056_1.fastq</code></td>
<td>the first input file name. Forward</td>
</tr>
<tr class="even">
<td><code>SRR_1056_2.fastq</code></td>
<td>the second input file name. Reverse</td>
</tr>
<tr class="odd">
<td><code>SRR_1056_1.trimmed.fastq</code></td>
<td>the output file for surviving pairs from the <code>_1</code>
file</td>
</tr>
<tr class="even">
<td><code>SRR_1056_1un.trimmed.fastq</code></td>
<td>the output file for orphaned reads from the <code>_1</code>
file</td>
</tr>
<tr class="odd">
<td><code>SRR_1056_2.trimmed.fastq</code></td>
<td>the output file for surviving pairs from the <code>_2</code>
file</td>
</tr>
<tr class="even">
<td><code>SRR_1056_2un.trimmed.fastq</code></td>
<td>the output file for orphaned reads from the <code>_2</code>
file</td>
</tr>
<tr class="odd">
<td><code>ILLUMINACLIP:SRR_adapters.fa</code></td>
<td>to clip the Illumina adapters from the input file using the adapter
sequences listed in <code>SRR_adapters.fa</code>
</td>
</tr>
<tr class="even">
<td><code>SLIDINGWINDOW:4:20</code></td>
<td>to use a sliding window of size 4 that will remove bases if their
Phred score is below 20</td>
</tr>
</tbody>
</table>
<div id="multi-line-for-long-commands" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="multi-line-for-long-commands" class="callout-inner">
<h3 class="callout-title">Multi-line for long commands</h3>
<div class="callout-content">
<p>Some of the commands we ran in this lesson are long! To separate code
chunks onto separate lines When typing into your terminal one command
with long input or many modifying parameters, you can use the
<code>\</code> character to make your code more readable. For example,
let us use multi lines with the echo command. With <code>\</code> it is
possible to use several lines to print “hello world” on your
terminal.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> echo he<span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>$ llo<span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>$ world</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ hello world</code></pre>
</div>
<p>Note: Some terminals only wait a few seconds for you to keep typing.
In that case, you may write down the full command in a text file and
then copy it to your terminal.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="running-trimmomatic">Running Trimmomatic<a class="anchor" aria-label="anchor" href="#running-trimmomatic"></a>
</h2>
<hr class="half-width">
<p>Now, we will run Trimmomatic on our data. Navigate to your
<code>untrimmed_fastq</code> data directory and verify that you are
located in the <code>untrimmed_fastq/</code> directory:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">$</span> pwd</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ /home/dcuser/dc_workshop/data/untrimmed_fastq</code></pre>
</div>
<p>You should have only four files in this directory. Those files
correspond to the files of forward and reverse reads from samples JC1A
and JP4D.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ JC1A_R1.fastq.gz  JC1A_R2.fastq.gz  JP4D_R1.fastq  JP4D_R2.fastq.gz  TruSeq3-PE.fa   </code></pre>
</div>
<p>We are going to run Trimmomatic on one of our paired-end samples.
While using FastQC, we saw that Universal adapters were present in our
samples. The adapter sequences came with the installation of Trimmomatic
and it is located in our current directory in the file
<code>TruSeq3-PE.fa</code>.</p>
<p>We will also use a sliding window of size 4 that will remove bases if
their Phred score is below 20 (like in our example above). We will also
discard any reads that do not have at least 25 bases remaining after
this trimming step. This command will take a few minutes to run.</p>
<p>Before, we unzipped one of our files to work with it. Let us compress
the file corresponding to the sample <code>JP4D</code> again before we
run Trimmomatic.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">gzip</span> JP4D_R1.fastq</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> trimmomatic PE JP4D_R1.fastq.gz JP4D_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>JP4D_R1.trim.fastq.gz  JP4D_R1un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>JP4D_R2.trim.fastq.gz  JP4D_R2un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>SLIDINGWINDOW:4:20 MINLEN:35 ILLUMINACLIP:TruSeq3-PE.fa:2:40:15 </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>TrimmomaticPE: Started with arguments:
 JP4D_R1.fastq.gz JP4D_R2.fastq.gz JP4D_R1.trim.fastq.gz JP4D_R1un.trim.fastq.gz JP4D_R2.trim.fastq.gz JP4D_R2un.trim.fastq.gz SLIDINGWINDOW:4:20 MINLEN:35 ILLUMINACLIP:TruSeq3-PE.fa:2:40:15
Multiple cores found: Using 2 threads
Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'
ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
Quality encoding detected as phred33
Input Read Pairs: 1123987 Both Surviving: 751427 (66.85%) Forward Only Surviving: 341434 (30.38%) Reverse Only Surviving: 11303 (1.01%) Dropped: 19823 (1.76%)
TrimmomaticPE: Completed successfully</code></pre>
</div>
<div id="exercise-1-what-did-trimmomatic-do" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-what-did-trimmomatic-do" class="callout-inner">
<h3 class="callout-title">Exercise 1: What did Trimmomatic do?</h3>
<div class="callout-content">
<p>Use the output from your Trimmomatic command to answer the following
questions.</p>
<ol style="list-style-type: decimal">
<li>What percent of reads did we discard from our sample?</li>
<li>What percent of reads did we keep both pairs?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>1.76%</li>
<li>66.85%</li>
</ol>
</div>
</div>
</div>
</div>
<p>You may have noticed that Trimmomatic automatically detected the
quality encoding of our sample. It is always a good idea to double-check
this or manually enter the quality encoding.</p>
<p>We can confirm that we have our output files:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> ls JP4D<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JP4D_R1.fastq.gz       JP4D_R1un.trim.fastq.gz	JP4D_R2.trim.fastq.gz
JP4D_R1.trim.fastq.gz  JP4D_R2.fastq.gz		JP4D_R2un.trim.fastq.gz</code></pre>
</div>
<p>The output files are also FASTQ files. It should be smaller than our
input file because we have removed reads. We can confirm this with this
command:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">$</span> ls JP4D<span class="pp">*</span> <span class="at">-l</span> <span class="at">-h</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-rw-r--r-- 1 dcuser dcuser 179M Nov 26 12:44 JP4D_R1.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 107M Mar 11 23:05 JP4D_R1.trim.fastq.gz
-rw-rw-r-- 1 dcuser dcuser  43M Mar 11 23:05 JP4D_R1un.trim.fastq.gz
-rw-r--r-- 1 dcuser dcuser 203M Nov 26 12:51 JP4D_R2.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 109M Mar 11 23:05 JP4D_R2.trim.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 1.3M Mar 11 23:05 JP4D_R2un.trim.fastq.gz</code></pre>
</div>
<p>We have just successfully run Trimmomatic on one of our FASTQ files!
However, there is some bad news. Trimmomatic can only operate on one
sample at a time, and we have more than one sample. The good news is
that we can use a <code>for</code> loop to iterate through our sample
files quickly!</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">$</span> for infile in <span class="pp">*</span>_R1.fastq.gz</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="va">base</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">${infile}</span> _R1.fastq.gz<span class="va">)</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="ex">trimmomatic</span> PE <span class="va">${infile}</span> <span class="va">${base}</span>_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="va">${base}</span>_R1.trim.fastq.gz <span class="va">${base}</span>_R1un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a><span class="va">${base}</span>_R2.trim.fastq.gz <span class="va">${base}</span>_R2un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>SLIDINGWINDOW:4:20 MINLEN:35 ILLUMINACLIP:TruSeq3-PE.fa:2:40:15</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>Go ahead and run the <code>for</code> loop. It should take a few
minutes for Trimmomatic to run for each of our four input files. Once it
is done, take a look at your directory contents. You will notice that
even though we ran Trimmomatic on file <code>JP4D</code> before running
the for loop, there is only one set of files for it. Because we matched
the ending <code>_R1.fastq.gz</code>, we re-ran Trimmomatic on this
file, overwriting our first results. That is ok, but it is good to be
aware that it happened.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1.fastq.gz                     JP4D_R1.fastq.gz
JC1A_R1.trim.fastq.gz                JP4D_R1.trim.fastq.gz
JC1A_R1un.trim.fastq.gz              JP4D_R1un.trim.fastq.gz
JC1A_R2.fastq.gz                     JP4D_R2.fastq.gz
JC1A_R2.trim.fastq.gz                JP4D_R2.trim.fastq.gz
JC1A_R2un.trim.fastq.gz              JP4D_R2un.trim.fastq.gz
TruSeq3-PE.fa   </code></pre>
</div>
<p>We have completed the trimming and filtering steps of our quality
control process! Before we move on, let us move our trimmed FASTQ files
to a new subdirectory within our <code>data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="ex">$</span> mkdir ../trimmed_fastq</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.trim<span class="pp">*</span> ../trimmed_fastq</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="ex">$</span> cd ../trimmed_fastq</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1.trim.fastq.gz    JP4D_R1.trim.fastq.gz
JC1A_R1un.trim.fastq.gz  JP4D_R1un.trim.fastq.gz
JC1A_R2.trim.fastq.gz    JP4D_R2.trim.fastq.gz
JC1A_R2un.trim.fastq.gz  JP4D_R2un.trim.fastq.gz   </code></pre>
</div>
<div id="bonus-exercise-advanced-quality-test-after-trimming" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="bonus-exercise-advanced-quality-test-after-trimming" class="callout-inner">
<h3 class="callout-title">Bonus Exercise (Advanced): Quality test after trimming</h3>
<div class="callout-content">
<p>Now that our samples have gone through quality control, they should
perform better on the quality tests run by FastQC.</p>
<p>Sort the following chunks of code to re-run FastQC on your trimmed
FASTQ files and visualize the HTML files to see whether your per base
sequence quality is higher after trimming.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-203-203-131.compute-1.amazonaws.com:~/dc_workshop/data/trimmed_fastq/<span class="pp">*</span>.html ~/Desktop/fastqc_html/trimmed</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">$</span> fastqc ~/dc_workshop/data/trimmed_fastq/<span class="pp">*</span>.fastq<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">$</span> mkdir ~/Desktop/fastqc_html/trimmed</span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>In your AWS terminal window, do the following:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">$</span> fastqc ~/dc_workshop/data/trimmed_fastq/<span class="pp">*</span>.fastq<span class="pp">*</span></span></code></pre>
</div>
<p>In a terminal standing on your local computer, do:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">$</span> mkdir ~/Desktop/fastqc_html/trimmed</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-203-203-131.compute-1.amazonaws.com:~/dc_workshop/data/trimmed_fastq/<span class="pp">*</span>.html ~/Desktop/fastqc_html/trimmed</span></code></pre>
</div>
<p>Then take a look at the html files in your browser.</p>
<p>Remember to replace everything between the <code>@</code> and
<code>:</code> in your scp command with your AWS instance number.</p>
<p>After trimming and filtering, our overall quality is much higher, we
have a distribution of sequence lengths, and more samples pass adapter
content. However, quality trimming is not perfect, and some programs are
better at removing some sequences than others. Trimmomatic did pretty
well, though, and its performance is good enough for our workflow.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The options you set for the command-line tools you use are
important!</li>
<li>Data cleaning is essential at the beginning of metagenomics
workflows.</li>
<li>Use Trimmomatic to get rid of adapters and low-quality bases or
reads.</li>
<li>Carefully fill in the parameters and options required to call a
function in the bash shell.</li>
<li>Automate repetitive workflows using for loops.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-assembly"><p>Content from <a href="04-assembly.html">Metagenome Assembly</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/04-assembly.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why should genomic data be assembled?</li>
<li>What is the difference between reads and contigs?</li>
<li>How can we assemble a metagenome?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand what an assembly is.</li>
<li>Run a metagenomics assembly workflow.</li>
<li>Use an environment in a bioinformatic pipeline.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="assembling-reads">Assembling reads<a class="anchor" aria-label="anchor" href="#assembling-reads"></a>
</h2>
<hr class="half-width">
<p>The assembly process groups reads into contigs and contigs into
scaffolds to obtain (ideally) the sequence of a whole chromosome. There
are many programs devoted to <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2874646/" class="external-link">genome</a>
and metagenome assembly, some of the main strategies they use are Greedy
extension, OLC, and De Bruijn charts. Contrary to metabarcoding, shotgun
metagenomics needs an assembly step, which does not mean that
metabarcoding never uses an assembly step but sometimes is
unnecessary.</p>
<p><a href="fig/03-04-01.png"></a></p>
<figure><img src="fig/03-04-01.png" width="868" height="777" alt="Three diagrams depicting the three assembly algorithms: The Greedy extension starts with any read, extends it whit the reads that make a match to make a contig, it continues with a different read when the previous contig can not be extended anymore. The Overlap Layout consensus finds every pairwise overlap, makes a layout graph with all the overlaps and chooses consensus sequences to make the contigs. The De Bruijn Graphs divides the reads in k-mers, makes a k-mer graph that shows all the overlapping k-mers, and chooses paths from the graph to make the contigs. " class="figure mx-auto d-block"></figure><p></p>
<p><a href="https://github.com/ablab/spades" class="external-link">MetaSPAdes</a> is an NGS de
novo assembler for assembling large and complex metagenomics data, and
it is one of the most used and recommended. It is part of the SPAdes
toolkit, which contains several assembly pipelines.</p>
<p>Some of the problems faced by metagenomics assembly are:</p>
<ul>
<li>Differences in coverage between the genomes due to the differences
in abundance in the sample.</li>
<li>The fact that different species often share conserved regions.</li>
<li>The presence of several strains of a single species in the
community.</li>
</ul>
<p>SPAdes already deals with the non-uniform coverage problem in its
algorithm, so it is helpful for the assembly of simple communities, but
the <a href="https://pubmed.ncbi.nlm.nih.gov/28298430/" class="external-link">metaSPAdes</a>
algorithm deals with the other problems as well, allowing it to assemble
metagenomes from complex communities.</p>
<p>The process of (metagenomics) assembly can take a long time, and if
the connection to the server drops, the process is killed, and the
process needs to restart. To avoid this, we can create a screen
session.</p>
</section><section><h2 class="section-heading" id="screen-sessions">Screen sessions<a class="anchor" aria-label="anchor" href="#screen-sessions"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="starting-a-new-session">Starting a new session<a class="anchor" aria-label="anchor" href="#starting-a-new-session"></a>
</h3>
<p>A ‘session’ can be considered a new window or screen: you might open
a terminal to do one thing on the computer and then open a new terminal
to work on another task at the command line. You can start a session and
give it a descriptive name:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> screen <span class="at">-S</span> assembly</span></code></pre>
</div>
<p>This steps creates a session with the name ‘assembly’.</p>
<p>As you work, this session will stay active until you close it. Even
if you log out or work on something else, the jobs you start in this
session will run until completion.</p>
</div>
<div class="section level3">
<h3 id="detach-session-process-keeps-running-in-the-background">Detach session (process keeps running in the background)<a class="anchor" aria-label="anchor" href="#detach-session-process-keeps-running-in-the-background"></a>
</h3>
<p>You can detach from a session by pressing <code>control + a</code>
followed by <code>d</code> (for detach) on your keyboard. If you
reconnect to your machine, you will have to reconnect to your session to
see how it went.</p>
<div id="additional-session-commands" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="additional-session-commands" class="callout-inner">
<h3 class="callout-title">Additional session commands</h3>
<div class="callout-content">
<p><strong>Seeing active sessions</strong> If you disconnect from your
session or from your ssh, you will need to reconnect to an existing
<code>screen</code> session. You can see a list of existing
sessions:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> screen <span class="at">-ls</span></span></code></pre>
</div>
<p><strong>Reconnecting to a session</strong> To reconnect to an
existing session:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> screen <span class="at">-r</span> session_name</span></code></pre>
</div>
<p>The <code>-r</code> option = ‘resume a detached screen session’</p>
<p><strong>Kill a session</strong> To end a session, type
<code>exit</code> after reconnecting to the session:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> screen <span class="at">-r</span> session_name</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">$</span> exit</span></code></pre>
</div>
</div>
</div>
</div>
<p>Let’s see if our program is installed correctly:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> metaspades.py</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>SPAdes genome assembler v3.15.0 [metaSPAdes mode]

Usage: spades.py [options] -o &lt;output_dir&gt;

Basic options:
  -o &lt;output_dir&gt;             directory to store all the resulting files (required)
  --iontorrent                this flag is required for IonTorrent data
  --test                      runs SPAdes on a toy dataset
  -h, --help                  prints this usage message
  -v, --version               prints version

Input data:
  --12 &lt;filename&gt;             file with interlaced forward and reverse paired-end reads
  -1 &lt;filename&gt;               file with forward paired-end reads
  -2 &lt;filename&gt;               file with reverse paired-end reads    </code></pre>
</div>
<div id="activate-your-environment" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="activate-your-environment" class="callout-inner">
<h3 class="callout-title">Activate your environment</h3>
<div class="callout-content">
<p>If you do not have the metagenomics environment activated, the
previous command should have given you an error. Before you proceed,
activate the environment:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">conda</span> activate metagenomics</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="metaspades-is-a-metagenomics-assembler">MetaSPAdes is a metagenomics assembler<a class="anchor" aria-label="anchor" href="#metaspades-is-a-metagenomics-assembler"></a>
</h2>
<hr class="half-width">
<p>The help we just saw tells us how to run <code>metaspades.py</code>.
We are going to use the most straightforward options, just specifying
our forward paired-end reads with <code>-1</code> and reverse paired-end
reads with <code>-2</code>, and the output directory where we want our
results to be stored.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/trimmed_fastq</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="ex">$</span> metaspades.py <span class="at">-1</span> JC1A_R1.trim.fastq.gz <span class="at">-2</span> JC1A_R2.trim.fastq.gz <span class="at">-o</span> ../../results/assembly_JC1A</span></code></pre>
</div>
<p>Now that it is running we should detach our screen with
<code>control + a</code> <code>d</code> and wait for a few minutes while
it running. And then attach the screen with
<code>screen -r assembly</code> to see how it went.</p>
<p>When the run is finished, it shows this message:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">=======</span> SPAdes pipeline finished.</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="ex">SPAdes</span> log can be found here: /home/dcuser/dc_workshop/results/assembly_JC1A/spades.log</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="ex">Thank</span> you for using SPAdes!</span></code></pre>
</div>
<p>Now we can kill the screen with <code>exit</code> and look at our
results in the main screen.</p>
<p>Now, let’s go to the output files:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> cd ../../results/assembly_JC1A</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>assembly_graph_after_simplification.gfa  corrected/              K55/             scaffolds.fasta
assembly_graph.fastg                     dataset.info            misc/            scaffolds.paths
assembly_graph_with_scaffolds.gfa        first_pe_contigs.fasta  params.txt       spades.log
before_rr.fasta                          input_dataset.yaml      pipeline_state/  strain_graph.gfa
contigs.fasta                            K21/                    run_spades.sh    tmp/
contigs.paths                            K33/                    run_spades.yaml   </code></pre>
</div>
<p>As we can see, MetaSPAdes gave us a lot of files. The ones with the
assembly are the <code>contigs.fasta</code> and the
<code>scaffolds.fasta</code>. Also, we found three <code>K</code>
folders: <em>K21, K33, and K55</em>; this contains the individual result
files for an assembly with k-mers equal to those numbers: 21, 33, and
55. The best-assembled results are the ones that are displayed outside
these k-folders. The folder <code>corrected</code> hold the corrected
reads with the SPAdes algorithm. Moreover, the file
<code>assembly_graph_with_scaffolds.gfa</code> have the information
needed to visualize our assembly by different means, like programs such
as <a href="https://rrwick.github.io/Bandage/" class="external-link">Bandage</a>.</p>
<p>The contigs are just made from assembled reads, but the scaffolds are
the result from a subsequent process in which the contigs are ordered,
oriented, and connected with Ns.</p>
<p>We can recognize which sample our assembly outputs corresponds to
because they are inside the assembly results folder:
<code>assembly_JC1A/</code>. However, the files within it do not have
the sample ID. If we need the files out of their folder, it is
beneficial to rename them.</p>
<div id="exercise-1-rename-all-files-in-a-folder-needed-in-the-next-episode" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-rename-all-files-in-a-folder-needed-in-the-next-episode" class="callout-inner">
<h3 class="callout-title">Exercise 1: Rename all files in a folder (needed in the next episode)</h3>
<div class="callout-content">
<p>Add the prefix <code>JC1A</code> (the sample ID) separated by a
<code>_</code> to the beginning of the names of all the contents in the
<code>assembly_JC1A/</code> directory. Remember that many solutions are
possible.</p>
<ol style="list-style-type: upper-alpha">
<li>
<code>$ mv * JC1A_</code><br>
</li>
<li>
<code>$ mv * JC1A_*</code><br>
</li>
<li>
<code>$ for name in *; do mv $name JC1A_; done</code><br>
</li>
<li><code>$ for name in *; do mv $name JC1A_$name; done</code></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha">
<li>No, this option is going to give you as error
<code>mv: target 'JC1A_' is not a directory</code> This is because
<code>mv</code> has two options:<br><code>mv file_1 file_2</code><br><code>mv file_1, file_2, ..... file_n directory</code><br>
When a list of files is passed to <code>mv</code>, the <code>mv</code>
expects the last parameters to be a directory.<br>
Here, <code>*</code> gives you a list of all the files in the directory.
The last parameter is <code>JC1A_</code> (which <code>mv</code> expects
to be a directory).<br>
</li>
<li>No. Again, every file is sent to the same file.<br>
</li>
<li>No, every file is sent to the same file JC1A_<br>
</li>
<li>Yes, this is one of the possible solutions.</li>
</ol>
<p>¿Do you have another solution?</p>
</div>
</div>
</div>
</div>
<div id="exercise-2-compare-two-fasta-files-from-the-assembly-output" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-compare-two-fasta-files-from-the-assembly-output" class="callout-inner">
<h3 class="callout-title">Exercise 2: Compare two fasta files from the assembly output</h3>
<div class="callout-content">
<p>You want to know how many contigs and scaffolds result from the
assembly. Use <code>contigs.fasta</code> and
<code>scaffolds.fasta</code> files and sort the commands to create
correct code lines.<br>
Do they have the same number of lines? Why?<br><strong>Hint</strong>: You can use the following commands:
<code>grep</code>, <code>|</code> (pipe), <code>-l</code>,
<code>"&gt;"</code>, <code>wc</code>, <code>filename.fasta</code></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">$</span> grep <span class="st">'&gt;'</span> contigs.fasta <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="ex">$</span> grep <span class="st">'&gt;'</span> scaffolds.fasta <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre>
</div>
<p>A contig is created from reads and then a scaffold from a group of
contigs, so we expect fewer lines in the
<code>scaffolds.fasta</code>.</p>
</div>
</div>
</div>
</div>
<div id="quality-of-assemblies" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="quality-of-assemblies" class="callout-inner">
<h3 class="callout-title">Quality of assemblies</h3>
<div class="callout-content">
<p>You can use several metrics to know the quality of your assemblies.
<a href="https://quast.sourceforge.net/metaquast.html" class="external-link">MetaQuast</a> is
a program that gives you these metrics for metagenome assemblies in an
interactive report and text files and plots.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Assembly groups reads into contigs.</li>
<li>De Bruijn Graphs use Kmers to assembly cleaned reads.</li>
<li>Program screen allows you to keep open remote sessions.</li>
<li>MetaSPAdes is a metagenomes assembler.</li>
<li>Assemblers take FastQ files as input and produce a Fasta file as
output.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-05-binning"><p>Content from <a href="05-binning.html">Metagenome Binning</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/05-binning.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we obtain the original genomes from a metagenome?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Obtain Metagenome-Assembled Genomes from the metagenomic
assembly.</li>
<li>Check the quality of the Metagenome-Assembled genomes.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="metagenomic-binning">Metagenomic binning<a class="anchor" aria-label="anchor" href="#metagenomic-binning"></a>
</h2>
<hr class="half-width">
<p>Original genomes in the sample can be separated with a process called
binning. This process allows separate analysis of each species contained
in the metagenome with enough reads to reconstruct a genome. Genomes
reconstructed from metagenomic assemblies are called MAGs
(Metagenome-Assembled Genomes). In this process, the assembled contigs
from the metagenome will be assigned to different bins (FASTA files that
contain certain contigs). Ideally, each bin corresponds to only one
original genome (a MAG).</p>
<p><a href="fig/03-05-01.png">
<img src="fig/03-05-01.png" width="435" height="631" alt="Diagram depicts the DNA sequences in the original sample as circular chromosomes of three different taxa. After sequencing, the DNA sequences of the three different taxa are mixed as small linear reads; after the assembly, we have contigs, each corresponding to a single taxon, except for the ones with a bad assembly that has sequences of different taxa in the same contig, after the binning taxa separate the contigs." class="figure"></a></p>
<p>Although an obvious way to separate contigs that correspond to a
different species is by their taxonomic assignation, there are more
reliable methods that do the binning using characteristics of the
contigs, such as their GC content, the use of tetranucleotides
(composition), or their coverage (abundance).</p>
<p><a href="https://sourceforge.net/projects/maxbin/files/" class="external-link">Maxbin</a>
is a binning algorithm that distinguishes between contigs that belong to
different bins according to their coverage levels and the
tetranucleotide frequencies they have.</p>
<div id="discussion-1-relation-between-mags-and-depth" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="discussion-1-relation-between-mags-and-depth" class="callout-inner">
<h3 class="callout-title">Discussion 1: Relation between MAGs and depth</h3>
<div class="callout-content">
<p>The sequencing center has returned you a file with 18,412 reads.
Given that the bacterial genome size range between 4Mbp and 13Mbp
(Mb=10^6 bp) and that the size of the reads in this run is 150bp. With
these data, how many complete bacterial genomes can you reconstruct?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>None, because 18,412 reads of 150bp give a total count of 2,761,800bp
(~2Mbp). Even if no read maps to the same region, the amount of base
pairs is inferior to the size of a bacterial genome.</p>
</div>
</div>
</div>
</div>
<p>Let us bin the sample we just assembled. The command for running
MaxBin is <code>run_MaxBin.pl</code>, and the arguments it needs are the
FASTA file of the assembly, the FASTQ with the forward and reverse
reads, the output directory, and the name.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/results/assembly_JC1A</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">$</span> mkdir MAXBIN</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">$</span> run_MaxBin.pl <span class="at">-thread</span> 8 <span class="at">-contig</span> JC1A_contigs.fasta <span class="at">-reads</span> ../../data/trimmed_fastq/JC1A_R1.trim.fastq.gz <span class="at">-reads2</span> ../../data/trimmed_fastq/JC1A_R2.trim.fastq.gz <span class="at">-out</span> MAXBIN/JC1A</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>MaxBin 2.2.7
Thread: 12
Input contig: JC1A_contigs.fasta
Located reads file [../../data/trimmed_fastq/JC1A_R1.trim.fastq.gz]
Located reads file [../../data/trimmed_fastq/JC1A_R2.trim.fastq.gz]
out header: MAXBIN/JC1A
Running Bowtie2 on reads file [../../data/trimmed_fastq/JC1A_R1.trim.fastq.gz]...this may take a while...
Reading SAM file to estimate abundance values...
Running Bowtie2 on reads file [../../data/trimmed_fastq/JC1A_R2.trim.fastq.gz]...this may take a while...
Reading SAM file to estimate abundance values...
Searching against 107 marker genes to find starting seed contigs for [JC1A_contigs.fasta]...
Running FragGeneScan....
Running HMMER hmmsearch....
Try harder to dig out marker genes from contigs.
Marker gene search reveals that the dataset cannot be binned (the medium of marker gene number &lt;= 1). Program stop.</code></pre>
</div>
<p>It seems impossible to bin our assembly because the number of marker
genes is less than 1. We could have expected this as we know it is a
small sample.</p>
<p>We will perform the binning process with the other sample from the
same study that is larger. We have the assembly precomputed in the
<code>~/dc-workshop/mags/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/mags/</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">$</span> mkdir MAXBIN</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">$</span> run_MaxBin.pl <span class="at">-thread</span> 8 <span class="at">-contig</span> JP4D_contigs.fasta <span class="at">-reads</span> ../data/trimmed_fastq/JP4D_R1.trim.fastq.gz <span class="at">-reads2</span> ../data/trimmed_fastq/JP4D_R2.trim.fastq.gz <span class="at">-out</span> MAXBIN/JP4D</span></code></pre>
</div>
<p>It will take a few minutes to run. Moreover, it will finish with an
output like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>========== Job finished ==========
Yielded 4 bins for contig (scaffold) file JP4D_contigs.fasta

Here are the output files for this run.
Please refer to the README file for further details.

Summary file: MAXBIN/JP4D.summary
Genome abundance info file: MAXBIN/JP4D.abundance
Marker counts: MAXBIN/JP4D.marker
Marker genes for each bin: MAXBIN/JP4D.marker_of_each_gene.tar.gz
Bin files: MAXBIN/JP4D.001.fasta - MAXBIN/JP4D.004.fasta
Unbinned sequences: MAXBIN/JP4D.noclass

Store abundance information of reads file [../data/trimmed_fastq/JP4D_R1.trim.fastq.gz] in [MAXBIN/JP4D.abund1].
Store abundance information of reads file [../data/trimmed_fastq/JP4D_R2.trim.fastq.gz] in [MAXBIN/JP4D.abund2].


========== Elapsed Time ==========
0 hours 6 minutes and 56 seconds.
</code></pre>
</div>
<p>With the <code>.summary</code> file, we can quickly look at the bins
that MaxBin produced.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> cat MAXBIN/JP4D.summary</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Bin name	Completeness	Genome size	GC content
JP4D.001.fasta	57.9%	3141556	55.5
JP4D.002.fasta	87.9%	6186438	67.3
JP4D.003.fasta	51.4%	3289972	48.1
JP4D.004.fasta	77.6%	5692657	38.9</code></pre>
</div>
<div id="discussion-2-the-quality-of-mags" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="discussion-2-the-quality-of-mags" class="callout-inner">
<h3 class="callout-title">Discussion 2: The quality of MAGs</h3>
<div class="callout-content">
<p>Can we trust the quality of our bins only with the given information?
What else do we want to know about our MAGs to use for further analysis
confidently?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p><strong>completeness</strong> is fundamental to know which data you
are working with. If the MAG is incomplete, you can hypothesize that if
you did not find something, it is because you do not have a complete
genome. <strong>Genome size</strong> and <strong>GC content</strong> are
like genomic fingerprints of taxa, so you can know if you have the taxa
you are looking for. Since we are working with the mixed genomes of a
community when we try to separate them with binning, we want to know if
we were able to separate them correctly. So we need to measure
<strong>contamination</strong> to know if we have only one genome in our
bin.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="quality-check">Quality check<a class="anchor" aria-label="anchor" href="#quality-check"></a>
</h2>
<hr class="half-width">
<p>The quality of a MAG is highly dependent on the size of the genome of
the species, its abundance in the community and the depth at which we
sequenced it. Two important things that can be measured to know its
quality are completeness (is the MAG a complete genome?) and if it is
contaminated (does the MAG contain only one genome?).</p>
<p><a href="https://github.com/Ecogenomics/CheckM" class="external-link">CheckM</a> is an
excellent program to see the quality of our MAGs. It measures
completeness and contamination by counting marker genes in the MAGs. The
lineage workflow that is a part of CheckM places your bins in a
reference tree to know to which lineage it corresponds and to use the
appropriate marker genes to estimate the quality parameters.
Unfortunately, the lineage workflow uses much memory, so it cannot run
on our machines, but we can tell CheckM to use marker genes from
Bacteria only to spend less memory. This is a less accurate approach,
but it can also be advantageous if you want all of your bins analyzed
with the same markers.</p>
<p>We will run the taxonomy workflow specifying the use of markers at
the domain level, specific for the rank Bacteria, we will specify that
our bins are in FASTA format, that they are located in the
<code>MAXBIN</code> directory and that we want our output in the
<code>CHECKM/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">$</span> mkdir CHECKM</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="ex">$</span> checkm taxonomy_wf domain Bacteria <span class="at">-x</span> fasta MAXBIN/ CHECKM/ </span></code></pre>
</div>
<p>The run will end with our results printed in the console.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>--------------------------------------------------------------------------------------------------------------------------------------------------------
  Bin Id     Marker lineage   # genomes   # markers   # marker sets   0    1    2    3    4   5+   Completeness   Contamination   Strain heterogeneity
--------------------------------------------------------------------------------------------------------------------------------------------------------
  JP4D.002      Bacteria         5449        104            58        3    34   40   21   5   1       94.83           76.99              11.19
  JP4D.004      Bacteria         5449        104            58        12   40   46   6    0   0       87.30           51.64               3.12
  JP4D.001      Bacteria         5449        104            58        24   65   11   3    1   0       70.48           13.09               0.00
  JP4D.003      Bacteria         5449        104            58        44   49   11   0    0   0       64.44           10.27               0.00
--------------------------------------------------------------------------------------------------------------------------------------------------------
</code></pre>
</div>
<p>To have these values in an output that is more usable and shearable,
we can now run the quality step of CheckM <code>checkm qa</code> and
make it print the output in a <code>TSV</code> table instead of the
console. In this step, we can ask CheckM to give us more parameters,
like contig number and length.</p>
<p>Ideally, we would like to get only one contig per bin, with a length
similar to the genome size of the corresponding taxa. Since this
scenario is complicated to obtain, we can use parameters showing how
good our assembly is. Here are some of the most common metrics: If we
arrange our contigs by size, from larger to smaller, and divide the
whole sequence in half, N50 is the size of the smallest contig in the
half that has the larger contigs; and L50 is the number of contigs in
this half of the sequence. So we want big N50 and small L50 values for
our genomes. Read <a href="https://www.molecularecologist.com/2017/03/29/whats-n50/" class="external-link">What is
N50?</a>.</p>
<p>To get the table with these extra parameters, we need to specify the
file of the markers that CheckM used in the previous step,
<code>Bacteria.ms</code>, the name of the output file we want,
<code>quality_JP4D.tsv</code>, that we want a table
<code>--tab_table</code>, and the option number 2 <code>-o 2</code> is
to ask for the extra parameters printed on the table.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">$</span>  checkm qa CHECKM/Bacteria.ms CHECKM/ <span class="at">--file</span> CHECKM/quality_JP4D.tsv <span class="at">--tab_table</span> <span class="at">-o</span> 2</span></code></pre>
</div>
<p>The table we just made looks like <a href="https://github.com/carpentries-incubator/metagenomics/blob/gh-pages/files/quality_JP4D.tsv" class="external-link">this</a>.
This will be very useful when you need to document or communicate your
work.</p>
<p>The question of how much contamination we can tolerate and how much
completeness we need certainly depends on the scientific question being
tackled, but in the <a href="https://genome.cshlp.org/content/25/7/1043" class="external-link">CheckM</a> paper,
there are some parameters that we can follow.</p>
<div id="exercise-1-discuss-the-quality-of-the-obtained-mags" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-discuss-the-quality-of-the-obtained-mags" class="callout-inner">
<h3 class="callout-title">Exercise 1: Discuss the quality of the obtained MAGs</h3>
<div class="callout-content">
<p>Fill in the blanks to complete the code you need to download the
<code>quality_JP4D.tsv</code> to your local computer:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">____</span> dcuser____ec2-18-207-132-236.compute-1.amazonaws.com____/home/dcuser/dc_workshop/mags/CHECKM/quality_JP4D.tsv ____</span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>In a terminal that is standing on your local computer, do:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-18-207-132-236.compute-1.amazonaws.com:/home/dcuser/dc_workshop/mags/CHECKM/quality_JP4D.tsv <span class="op">&lt;</span>the destination directory of your choice<span class="op">&gt;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-1-discuss-the-quality-of-the-obtained-mags" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-1-discuss-the-quality-of-the-obtained-mags" class="callout-inner">
<h3 class="callout-title">Exercise 1: Discuss the quality of the
obtained MAGs<em> (continued)</em>
</h3>
<div class="callout-content">
<p>Then open the table in a spreadsheet and discuss with your team which
of the parameters in the table you find useful.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Metagenome-Assembled Genomes (MAGs) sometimes are obtained from
curated contigs grouped into bins.</li>
<li>Use MAXBIN to assign the contigs to bins of different taxa.</li>
<li>Use CheckM to evaluate the quality of each Metagenomics-Assembled
Genome.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-06-taxonomic"><p>Content from <a href="06-taxonomic.html">Taxonomic Assignment</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/06-taxonomic.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I know to which taxa my sequences belong?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand how taxonomic assignment works.</li>
<li>Use Kraken to assign taxonomies to reads and contigs.</li>
<li>Visualize taxonomic assignations in graphics.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-is-a-taxonomic-assignment">What is a taxonomic assignment?<a class="anchor" aria-label="anchor" href="#what-is-a-taxonomic-assignment"></a>
</h2>
<hr class="half-width">
<p>A taxonomic assignment is a process of assigning an Operational
Taxonomic Unit (OTU, that is, groups of related individuals) to
sequences that can be reads or contigs. Sequences are compared against a
database constructed using complete genomes. When a sequence finds a
good enough match in the database, it is assigned to the corresponding
OTU. The comparison can be made in different ways.</p>
<div class="section level3">
<h3 id="strategies-for-taxonomic-assignment">Strategies for taxonomic assignment<a class="anchor" aria-label="anchor" href="#strategies-for-taxonomic-assignment"></a>
</h3>
<p>There are many programs for doing taxonomic mapping, and almost all
of them follow one of the following strategies:</p>
<ol style="list-style-type: decimal">
<li><p>BLAST: Using BLAST or DIAMOND, these mappers search for the most
likely hit for each sequence within a database of genomes (i.e.,
mapping). This strategy is slow.</p></li>
<li><p>Markers: They look for markers of a database made a priori in the
sequences to be classified and assigned the taxonomy depending on the
hits obtained.</p></li>
<li><p>K-mers: A genome database is broken into pieces of length k to be
able to search for unique pieces by taxonomic group, from a lowest
common ancestor (LCA), passing through phylum to species. Then, the
algorithm breaks the query sequence (reads/contigs) into pieces of
length k, looks for where these are placed within the tree and make the
classification with the most probable position.</p></li>
</ol>
<p><a href="fig/03-06-01.png">
<img src="fig/03-06-01.png" alt="Diagram of a taxonomic tree with four levels of nodes, some nodes have a number from 1 to 3, and some do not. From the most recent nodes, one has a three, and its parent nodes do not have numbers. This node with a three is selected." class="figure"></a> <em> Figure 1. Lowest common ancestor assignment example.<em></em></em></p>
</div>
<div class="section level3">
<h3 id="abundance-bias">Abundance bias<a class="anchor" aria-label="anchor" href="#abundance-bias"></a>
</h3>
<p>When you do the taxonomic assignment of metagenomes, a key result is
the abundance of each taxon or OTU in your sample. The absolute
abundance of a taxon is the number of sequences (reads or contigs,
depending on what you did) assigned to it. Moreover, its relative
abundance is the proportion of sequences assigned to it. It is essential
to be aware of the many biases that can skew the abundances along the
metagenomics workflow, shown in the figure, and that because of them, we
may not be obtaining the actual abundance of the organisms in the
sample.</p>
<p><a href="fig/03-06-02.png">
<img src="fig/03-06-02.png" alt="Flow diagram that shows how the initial composition of 33% for each of the three taxa in the sample ends up being 4%, 72%, and 24% after the biases imposed by the extraction, PCR, sequencing and bioinformatics steps." class="figure"></a> <em>Figure 2. Abundance biases during a metagenomics protocol.
<em></em></em></p>
<div id="discussion-taxonomic-level-of-assignment" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion-taxonomic-level-of-assignment" class="callout-inner">
<h3 class="callout-title">Discussion: Taxonomic level of assignment</h3>
<div class="callout-content">
<p>What do you think is harder to assign, a species (like <em>E.
coli</em>) or a phylum (like Proteobacteria)?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="using-kraken-2">Using Kraken 2<a class="anchor" aria-label="anchor" href="#using-kraken-2"></a>
</h2>
<hr class="half-width">
<p><a href="https://ccb.jhu.edu/software/kraken2/" class="external-link">Kraken 2</a> is the
newest version of Kraken, a taxonomic classification system using exact
k-mer matches to achieve high accuracy and fast classification speeds.
<code>kraken2</code> is already installed in the <strong>metagenomics
environment</strong>, let us have a look at <code>kraken2</code>
help.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> kraken2  <span class="at">--help</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Need to specify input filenames!
Usage: kraken2 [options] &lt;filename(s)&gt;

Options:
  --db NAME               Name for Kraken 2 DB
                          (default: none)
  --threads NUM           Number of threads (default: 1)
  --quick                 Quick operation (use first hit or hits)
  --unclassified-out FILENAME
                          Print unclassified sequences to filename
  --classified-out FILENAME
                          Print classified sequences to filename
  --output FILENAME       Print output to filename (default: stdout); "-" will
                          suppress normal output
  --confidence FLOAT      Confidence score threshold (default: 0.0); must be
                          in [0, 1].
  --minimum-base-quality NUM
                          Minimum base quality used in classification (def: 0,
                          only effective with FASTQ input).
  --report FILENAME       Print a report with aggregate counts/clade to file
  --use-mpa-style         With --report, format report output like Kraken 1's
                          kraken-mpa-report
  --report-zero-counts    With --report, report counts for ALL taxa, even if
                          counts are zero
  --report-minimizer-data With --report, report minimizer, and distinct minimizer
                          count information in addition to normal Kraken report
  --memory-mapping        Avoids loading database into RAM
  --paired                The filenames provided have paired-end reads
  --use-names             Print scientific names instead of just taxids
  --gzip-compressed       Input files are compressed with gzip
  --bzip2-compressed      Input files are compressed with bzip2
  --minimum-hit-groups NUM
                          Minimum number of hit groups (overlapping k-mers
                          sharing the same minimizer) needed to make a call
                          (default: 2)
  --help                  Print this message
  --version               Print version information

If none of the *-compressed flags are specified, and the filename provided
is a regular file, automatic format detection is attempted.
</code></pre>
</div>
<p>In the help, we can see that in addition to our input files, we also
need a database to compare them. The database you use will determine the
result you get for your data. Imagine you are searching for a recently
discovered lineage that is not part of the available databases. Would
you find it?</p>
<p>There are <a href="https://ccb.jhu.edu/software/kraken2/downloads.shtml" class="external-link">several
databases</a> compatible to be used with kraken2 in the taxonomical
assignment process.</p>
<p>Unfortunately, even the smallest Kraken database Minikraken, which
needs 8Gb of free RAM, is not small enough to be run by the machines we
are using, so <strong>we will not be able to run
<code>kraken2</code></strong>. We can check our available RAM with
<code>free -h</code>to be sure of this.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> free <span class="at">-h</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>              total        used        free      shared  buff/cache   available
Mem:           3.9G        272M        3.3G         48M        251M        3.3G
Swap:            0B          0B          0B</code></pre>
</div>
<div class="section level3">
<h3 id="taxonomic-assignment-of-metagenomic-reads">Taxonomic assignment of metagenomic reads<a class="anchor" aria-label="anchor" href="#taxonomic-assignment-of-metagenomic-reads"></a>
</h3>
<p>As we have learned, taxonomic assignments can be attempted before the
assembly. In this case, we would use FASTQ files as inputs, which would
be <code>JP4D_R1.trim.fastq.gz</code> and
<code>JP4D_R2.trim.fastq.gz</code>. And the outputs would be two files:
the report <code>JP4D.report</code> and the kraken file
<code>JP4D.kraken</code>.</p>
<p>To run kraken2, we would use a command like this:<br><strong>No need to run this</strong></p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> mkdir TAXONOMY_READS</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">$</span> kraken2 <span class="at">--db</span> kraken-db <span class="at">--threads</span> 8 <span class="at">--paired</span> JP4D_R1.trim.fastq.gz JP4D_R2.trim.fastq.gz <span class="at">--output</span> TAXONOMY_READS/JP4D.kraken <span class="at">--report</span> TAXONOMY_READS/JP4D.report</span></code></pre>
</div>
<p>Since we cannot run <code>kraken2</code> here, we precomputed its
results in a server, i.e., a more powerful machine. In the server we ran
<code>kraken2</code> and obtained<code>JP4D-kraken.kraken</code> and
<code>JP4D.report</code>.</p>
<p>Let us look at the precomputed outputs of <code>kraken2</code> for
our JP4D reads.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">head</span> ~/dc_workshop/taxonomy/JP4D.kraken  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:19691:2037	0	250|251	0:216 |:| 0:217
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14127:2052	0	250|238	0:216 |:| 0:204
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14766:2063	0	251|251	0:217 |:| 0:217
C	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15697:2078	2219696	250|120	0:28 350054:5 1224:2 0:1 2:5 0:77 2219696:5 0:93 |:| 379:4 0:82
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15529:2080	0	250|149	0:216 |:| 0:115
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14172:2086	0	251|250	0:217 |:| 0:216
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:17552:2088	0	251|249	0:217 |:| 0:215
U	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:14217:2104	0	251|227	0:217 |:| 0:193
C	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15110:2108	2109625	136|169	0:51 31989:5 2109625:7 0:39 |:| 0:5 74033:2 31989:5 1077935:1 31989:7 0:7 60890:2 0:105 2109625:1
C	MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:19558:2111	119045	251|133	0:18 1224:9 2:5 119045:4 0:181 |:| 0:99</code></pre>
</div>
<p>This information may need to be clarified. Let us take out our
cheatsheet to understand some of its components:</p>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Column example</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>C</td>
<td>Classified or unclassified</td>
</tr>
<tr class="even">
<td>MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:15697:2078</td>
<td>FASTA header of the sequence</td>
</tr>
<tr class="odd">
<td>2219696</td>
<td>Tax ID</td>
</tr>
<tr class="even">
<td>250:120</td>
<td>Read length</td>
</tr>
<tr class="odd">
<td>0:28 350054:5 1224:2 0:1 2:5 0:77 2219696:5 0:93 379:4 0:82</td>
<td>kmers hit to a taxonomic ID <em>e.g.,</em> tax ID 350054 has five
hits, tax ID 1224 has two hits, etc.</td>
</tr>
</tbody>
</table>
<p>The Kraken file could be more readable. So let us look at the report
file:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">head</span> ~/dc_workshop/taxonomy/JP4D.report</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> 78.13	587119	587119	U	0	unclassified
 21.87	164308	1166	R	1	root
 21.64	162584	0	R1	131567	  cellular organisms
 21.64	162584	3225	D	2	    Bacteria
 18.21	136871	3411	P	1224	      Proteobacteria
 14.21	106746	3663	C	28211	        Alphaproteobacteria
  7.71	57950	21	O	204455	          Rhodobacterales
  7.66	57527	6551	F	31989	            Rhodobacteraceae
  1.23	9235	420	G	1060	              Rhodobacter
  0.76	5733	4446	S	1063	                Rhodobacter sphaeroides</code></pre>
</div>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Column example</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>78.13</td>
<td>Percentage of reads covered by the clade rooted at this taxon</td>
</tr>
<tr class="even">
<td>587119</td>
<td>Number of reads covered by the clade rooted at this taxon</td>
</tr>
<tr class="odd">
<td>587119</td>
<td>Number of reads assigned directly to this taxon</td>
</tr>
<tr class="even">
<td>U</td>
<td>A rank code, indicating (U)nclassified, (D)omain, (K)ingdom,
(P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies. All other
ranks are simply ‘-’.</td>
</tr>
<tr class="odd">
<td>0</td>
<td>NCBI taxonomy ID</td>
</tr>
<tr class="even">
<td>unclassified</td>
<td>Indented scientific name</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="taxonomic-assignment-of-the-contigs-of-a-mag">Taxonomic assignment of the contigs of a MAG<a class="anchor" aria-label="anchor" href="#taxonomic-assignment-of-the-contigs-of-a-mag"></a>
</h3>
<p>We now have the taxonomic identity of the reads of the whole
metagenome, but we need to know to which taxon our MAGs correspond. For
this, we have to make the taxonomic assignment with their contigs
instead of its reads because we do not have the reads corresponding to a
MAG separated from the reads of the entire sample.</p>
<p>For this, the <code>kraken2</code> is a little bit different; here,
we can look at the command for the <code>JP4D.001.fasta</code> MAG:</p>
<p><strong>No need to run this</strong></p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> mkdir TAXONOMY_MAG</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">$</span> kraken2 <span class="at">--db</span> kraken-db <span class="at">--threads</span> 12 <span class="at">-input</span> JP4D.001.fasta <span class="at">--output</span> TAXONOMY_MAG/JP4D.001.kraken <span class="at">--report</span> TAXONOMY_MAG/JP4D.001.report</span></code></pre>
</div>
<p>The results of this are pre-computed in the
<code>~/dc_workshop/taxonomy/mags_taxonomy/</code> directory</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/taxonomy/mags_taxonomy</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">JP4D.001.kraken</span></span>
<span><span class="va">JP4D.001.report</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">more</span> ~/dc_workshop/taxonomy/mags_taxonomy/JP4D.001.report</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> 50.96	955	955	U	0	unclassified
 49.04	919	1	R	1	root
 48.83	915	0	R1	131567	  cellular organisms
 48.83	915	16	D	2	    Bacteria
 44.40	832	52	P	1224	      Proteobacteria
 19.37	363	16	C	28216	        Betaproteobacteria
 16.22	304	17	O	80840	          Burkholderiales
  5.66	106	12	F	506	            Alcaligenaceae
  2.72	51	3	G	517	              Bordetella
  1.12	21	21	S	2163011	                Bordetella sp. HZ20
  .
  .
  .</code></pre>
</div>
<p>Looking at the report, we can see that half of the contigs are
unclassified and that a tiny proportion of contigs have been assigned an
OTU. This result is weird because we expected only one genome in the
bin.</p>
<p>To exemplify how a report of a complete and not contaminated MAG
should look like this; let us look at the report of this MAG from
another study:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>100.00	108	0	R	1	root
100.00	108	0	R1	131567	  cellular organisms
100.00	108	0	D	2	    Bacteria
100.00	108	0	P	1224	      Proteobacteria
100.00	108	0	C	28211	        Alphaproteobacteria
100.00	108	0	O	356	          Rhizobiales
100.00	108	0	F	41294	            Bradyrhizobiaceae
100.00	108	0	G	374	              Bradyrhizobium
100.00	108	108	S	2057741	                Bradyrhizobium sp. SK17</code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="visualization-of-taxonomic-assignment-results">Visualization of taxonomic assignment results<a class="anchor" aria-label="anchor" href="#visualization-of-taxonomic-assignment-results"></a>
</h2>
<hr class="half-width">
<p>After we have the taxonomy assignation, what follows is some
visualization of our results. <a href="https://github.com/marbl/Krona/wiki" class="external-link">Krona</a> is a hierarchical
data visualization software. Krona allows data to be explored with
zooming and multi-layered pie charts and supports several bioinformatics
tools and raw data formats. To use Krona in our results, let us first go
into our taxonomy directory, which contains the pre-calculated Kraken
outputs.</p>
<div class="section level3">
<h3 id="krona">Krona<a class="anchor" aria-label="anchor" href="#krona"></a>
</h3>
<p>With Krona, we will explore the taxonomy of the JP4D.001 MAG.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/taxonomy/mags_taxonomy</span></code></pre>
</div>
<p>Krona is called with the <code>ktImportTaxonomy</code> command that
needs an input and an output file.<br>
In our case, we will create the input file with columns three and four
from <code>JP4D.001.kraken</code> file.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">$</span> cut <span class="at">-f2,3</span> JP4D.001.kraken <span class="op">&gt;</span> JP4D.001.krona.input</span></code></pre>
</div>
<p>Now we call Krona in our <code>JP4D.001.krona.input</code> file and
save results in <code>JP4D.001.krona.out.html</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">$</span> ktImportTaxonomy JP4D.001.krona.input <span class="at">-o</span> JP4D.001.krona.out.html</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading taxonomy...
Importing JP4D.001.krona.input...
   [ WARNING ]  The following taxonomy IDs were not found in the local database and were set to root
                (if they were recently added to NCBI, use updateTaxonomy.sh to update the local
                database): 1804984 2109625 2259134</code></pre>
</div>
<p>And finally, open another terminal on your local computer, download
the Krona output and open it on a browser.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-3-235-238-92.compute-1.amazonaws.com:~/dc_workshop/taxonomy/JP4D.001.krona.out.html . </span></code></pre>
</div>
<p>You will see a page like this:</p>
<p><a href="fig/03-06-03.png">
<img src="fig/03-06-03.png" alt="Krona displays a circled-shape bacterial taxonomy plot with abundance percentages of each taxon" class="figure"></a></p>
<div id="exercise-1-exploring-krona-visualization" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-exploring-krona-visualization" class="callout-inner">
<h3 class="callout-title">Exercise 1: Exploring Krona visualization</h3>
<div class="callout-content">
<p>Try double-clicking on the pie chart segment representing Bacteria
and see what happens. What percentage of bacteria is represented by the
genus <em>Paracoccus</em>?</p>
<p>Hint: A search box is in the window’s top left corner.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>2% of Bacteria corresponds to the genus Paracoccus in this sample. In
the top right of the window, we see little pie charts that change
whenever we change the visualization to expand certain taxa.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="pavian">Pavian<a class="anchor" aria-label="anchor" href="#pavian"></a>
</h3>
<p>Pavian is another visualization tool that allows comparison between
multiple samples. Pavian should be locally installed and needs R and
Shiny, but we can try the <a href="https://fbreitwieser.shinyapps.io/pavian/" class="external-link">Pavian demo WebSite</a>
to visualize our results.</p>
<p>First, we need to download the files needed as inputs in Pavian; this
time, we will visualize the assignment of the reads of both samples:
<code>JC1A.report</code> and <code>JP4D.report</code>.<br>
These files correspond to our Kraken reports. Again in our local
machine, let us use the <code>scp</code> command.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-3-235-238-92.compute-1.amazonaws.com:~/dc_workshop/taxonomy/<span class="pp">*</span>report . </span></code></pre>
</div>
<p>We go to the <a href="https://fbreitwieser.shinyapps.io/pavian/" class="external-link">Pavian demo
WebSite</a>, click on Browse, and choose our reports. You need to select
both reports at the same time.</p>
<p><a href="fig/03-06-04.png">
<img src="fig/03-06-04.png" alt="Pavian website showing the upload of two reports" class="figure"></a></p>
<p>We click on the Results Overview tab.</p>
<p><a href="fig/03-06-05.png">
<img src="fig/03-06-05.png" alt="Results Overview tab of the Pavian website where it shows the number of reads classified to several categories for the two samples" class="figure"></a></p>
<p>We click on the Sample tab.</p>
<p><a href="fig/03-06-06.png">
<img src="fig/03-06-06.png" alt="Sankey type visualization that shows the abundance of each taxonomic label in a tree-like manner" class="figure"></a></p>
<p>We can look at the abundance of a specific taxon by clicking on
it.</p>
<p><a href="fig/03-06-07.png">
<img src="fig/03-06-07.png" alt="A bar chart of the abundance of reads of the two samples, showing a segment for the read identified at the specific taxon and another segment for the number of reads identifies at children of the specified taxon" class="figure"></a></p>
<p>We can look at a comparison of both our samples in the Comparison
tab.</p>
<p><a href="fig/03-06-08.png">
<img src="fig/03-06-08.png" alt="A table of the same format as the Kraken report but for both samples at once." class="figure"></a></p>
<div id="discussion-unclassified-reads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="discussion-unclassified-reads" class="callout-inner">
<h3 class="callout-title">Discussion: Unclassified reads</h3>
<div class="callout-content">
<p>As you can see, a percentage of our data could not be assigned to
belong to a specific OTU.<br>
Which factors can affect the taxonomic assignation so that a read is
unclassified?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p><strong>Unclassified reads</strong> can be the result of different
factors that can go from sequencing errors to problems with the
algorithm being used to generate the result. The widely used
Next-generation sequencing (NGS) platforms, showed <a href="https://www.nature.com/articles/s41598-018-29325-6" class="external-link">average error
rate of 0.24±0.06% per base</a>. Besides the sequencing error, we need
to consider the status of the database being used to perform the
taxonomic assignation.<br>
All the characterized genomes obtained by different research groups are
scattered in different repositories, pages, and banks in the cloud. Some
are still unpublished. Incomplete databases can affect the performance
of the taxonomic assignation. Imagine that the dominant OTU in your
sample belongs to a lineage that has never been characterized and does
not have a public genome available to be used as a template for the
database. This possibility makes the assignation an impossible task and
can promote the generation of false positives because the algorithm will
assign a different identity to all those reads.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>A database with previously gathered knowledge (genomes) is needed
for taxonomic assignment.</li>
<li>Taxonomic assignment can be done using Kraken.</li>
<li>Krona and Pavian are web-based tools to visualize the assigned
taxa.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-07-phyloseq"><p>Content from <a href="07-phyloseq.html">Exploring Taxonomy with R</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/07-phyloseq.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I use my taxonomic assignment results to analyze?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Comprehend which libraries are required for analysis of the taxonomy
of metagenomes.</li>
<li>Create and manage a Phyloseq object.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="creating-lineage-and-rank-tables">Creating lineage and rank tables<a class="anchor" aria-label="anchor" href="#creating-lineage-and-rank-tables"></a>
</h2>
<hr class="half-width">
<p>In this episode, we will use RStudio to analyze our microbial
samples. You do not have to install anything, you already have an
instance on the cloud ready to be used.</p>
<p>Packages like Qiime2, MEGAN, Vegan, or Phyloseq in R allow us to
analyze diversity and abundance by manipulating taxonomic assignment
data. In this lesson, we will use Phyloseq. In order to do so, we need
to generate an abundance matrix from the Kraken output files. One
program widely used for this purpose is <code>kraken-biom</code>.</p>
<p>To do this, we could go to our now familiar Bash terminal, but
RStudio has an integrated terminal that uses the same language as the
one we learned in the Command-line lessons, so let us take advantage of
it. Let us open RStudio and go to the Terminal tab in the bottom left
panel.</p>
<div class="section level3">
<h3 id="kraken-biom">Kraken-biom<a class="anchor" aria-label="anchor" href="#kraken-biom"></a>
</h3>
<p><a href="https://github.com/smdabdoub/kraken-biom" class="external-link">Kraken-biom</a> is
a program that creates BIOM tables from the Kraken output.</p>
<p>In order to run Kraken-biom, we have to move to the folder where our
taxonomic data files are located:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/taxonomy</span></code></pre>
</div>
<p>First, we will visualize the content of our directory by the
<code>ls</code> command.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A.kraken  JC1A.report  JP41.report  JP4D.kraken  JP4D.report  mags_taxonomy</code></pre>
</div>
<p>The <code>kraken-biom</code> program is installed inside our
<code>metagenomics</code> environment, so let us activate it.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> conda activate metagenomics </span></code></pre>
</div>
<p>Let us take a look at the different flags that
<code>kraken-biom</code> has:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> kraken-biom <span class="at">-h</span>                  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>usage: kraken-biom [-h] [--max {D,P,C,O,F,G,S}] [--min {D,P,C,O,F,G,S}]
                   [-o OUTPUT_FP] [--otu_fp OTU_FP] [--fmt {hdf5,json,tsv}]
                   [--gzip] [--version] [-v]
                   kraken_reports [kraken_reports ...]

Create BIOM-format tables (http://biom-format.org) from Kraken output
(http://ccb.jhu.edu/software/kraken/).
.
.
.</code></pre>
</div>
<p>By a close look at the first output lines, it is noticeable that we
need a specific output from Kraken: the <code>.reports</code>.</p>
<p>With the following command, we will create a table in <a href="https://biom-format.org/" class="external-link">Biom</a> format called
<code>cuatroc.biom</code>. We will include the two samples we have been
working with (<code>JC1A</code> and <code>JP4D</code>) and a third one
(<code>JP41</code>) to be able to perform specific analyses later
on.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">$</span> kraken-biom JC1A.report JP4D.report JP41.report <span class="at">--fmt</span> json <span class="at">-o</span> cuatroc.biom</span></code></pre>
</div>
<p>If we inspect our folder, we will see that the
<code>cuatroc.biom</code> file has been created. This <code>biom</code>
object contains both the abundance and the ID (a number) of each
OTU.<br>
With this result, we are ready to return to RStudio’s console and begin
to manipulate our taxonomic-data.</p>
<div id="command-line-prompts" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="command-line-prompts" class="callout-inner">
<h3 class="callout-title">Command line prompts</h3>
<div class="callout-content">
<p>Note that you can distinguish the Bash terminal from the R console by
looking at the prompt. In Bash is the <code>$</code> sign, and in R is
the <code>&gt;</code> sign.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="creating-and-manipulating-phyloseq-objects">Creating and manipulating Phyloseq objects<a class="anchor" aria-label="anchor" href="#creating-and-manipulating-phyloseq-objects"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="load-required-packages">Load required packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h3>
<p>Phyloseq is a library with tools to analyze and plot your
metagenomics samples’ taxonomic assignment and abundance information.
Let us install <a href="https://joey711.github.io/phyloseq/" class="external-link">phyloseq</a> (This
instruction might not work on specific versions of R) and other
libraries required for its execution:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="sc">&gt;</span> BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"phyloseq"</span>) <span class="co"># Install phyloseq</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"RColorBrewer"</span>, <span class="st">"patchwork"</span>)) <span class="co">#install patchwork to chart publication-quality plots and readr to read rectangular datasets.</span></span></code></pre>
</div>
<p>Once the libraries are installed, we must make them available for
this R session. Now load the libraries (a process needed every time we
begin a new work session in R):</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"phyloseq"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"RColorBrewer"</span>)</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">"patchwork"</span>)</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="creating-the-phyloseq-object">Creating the phyloseq object<a class="anchor" aria-label="anchor" href="#creating-the-phyloseq-object"></a>
</h3>
<p>First, we tell R in which directory we are working.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">setwd</span>(<span class="st">"~/dc_workshop/taxonomy/"</span>)</span></code></pre>
</div>
<p>Let us proceed to create the phyloseq object with the
<code>import_biom</code> command:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes <span class="ot">&lt;-</span> <span class="fu">import_biom</span>(<span class="st">"cuatroc.biom"</span>)</span></code></pre>
</div>
<p>Now, we can inspect the result by asking the class of the object
created and doing a close inspection of some of its content:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">class</span>(merged_metagenomes)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "phyloseq"
attr("package")
[1] "phyloseq"</code></pre>
</div>
<p>The “class” command indicates that we already have our phyloseq
object.</p>
</div>
<div class="section level3">
<h3 id="exploring-the-taxonomic-labels">Exploring the taxonomic labels<a class="anchor" aria-label="anchor" href="#exploring-the-taxonomic-labels"></a>
</h3>
<p>Let us try to access the data that is stored inside our
<code>merged_metagenomes</code> object. Since a phyloseq object is a
special object in R, we need to use the operator <code>@</code> to
explore the subsections of data inside <code>merged_metagenomes</code>.
If we type <code>merged_metagenomes@</code>, five options are displayed;
<code>tax_table</code> and <code>otu_table</code> are the ones we will
use. After writing <code>merged_metagenomes@otu_table</code> or
<code>merged_metagenomes@tax_table</code>, an option of
<code>.Data</code> will be the one chosen in both cases. Let us see what
is inside our <code>tax_table</code>:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p><a href="fig/03-07-01.png">
<img src="fig/03-07-01.png" alt="A table where the taxonomic
  identification information of all OTUs is displayed. Each row represents one
  OTU and the columns represent its identification at different levels in the taxonomic classification ranks, begging with Kingdom until we reach Species
  in the seventh column." class="figure"></a> <em> Figure 1. Table of the taxonomic
labels from our <code>merged_metagenomes</code> object. <em></em></em></p>
<p>Here we can see that the <code>tax_table</code> inside our phyloseq
object stores all the taxonomic labels corresponding to each OTU.
Numbers in the row names of the table identify OTUs.</p>
<p>Next, let us get rid of some of the unnecessary characters in the
OTUs id and put names to the taxonomic ranks:</p>
<p>To remove unnecessary characters in <code>.Data</code> (matrix), we
will use the command <code>substring()</code>. This command helps
extract or replace characters in a vector. To use the command, we have
to indicate the vector (x) followed by the first element to replace or
extract (first) and the last element to be replaced (last). For
instance: <code>substring (x, first, last)</code>.
<code>substring()</code> is a “flexible” command, especially to select
characters of different lengths, as in our case. Therefore, it is not
necessary to indicate “last”, so it will take the last position of the
character by default. Since a matrix is an arrangement of vectors, we
can use this command. Each character in <code>.Data</code> is preceded
by three spaces occupied by a letter and two underscores, for example:
<code>o__Rhodobacterales</code>. In this case, “Rodobacterales” starts
at position 4 with an R. So, to remove the unnecessary characters, we
will use the following code:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data <span class="ot">&lt;-</span> <span class="fu">substring</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data, <span class="dv">4</span>)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">colnames</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span>)</span></code></pre>
</div>
<p><a href="fig/03-07-02.png">
<img src="fig/03-07-02.png" alt="The same table we saw in Figure 3 but with informative names in each of the columns. Now, we can see which of
  the columns are associated with which taxonomic classification rank" class="figure"></a> <em> Figure 2. Table of the taxonomic labels from our
<code>merged_metagenomes</code> object with corrections. <em></em></em></p>
<p>We will use a command named <code>unique()</code> to explore how many
phyla we have. Let us see the result we obtain from the following
code:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="st">"Phylum"</span>])</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Proteobacteria"              "Actinobacteria"              "Firmicutes"
 [4] "Cyanobacteria"               "Deinococcus-Thermus"         "Chloroflexi"
 [7] "Armatimonadetes"             "Bacteroidetes"               "Chlorobi"
[10] "Gemmatimonadetes"            "Planctomycetes"              "Verrucomicrobia"
[13] "Lentisphaerae"               "Kiritimatiellaeota"          "Chlamydiae"
[16] "Acidobacteria"               "Spirochaetes"                "Synergistetes"
[19] "Nitrospirae"                 "Tenericutes"                 "Coprothermobacterota"
[22] "Ignavibacteriae"             "Candidatus Cloacimonetes"    "Fibrobacteres"
[25] "Fusobacteria"                "Thermotogae"                 "Aquificae"
[28] "Thermodesulfobacteria"       "Deferribacteres"             "Chrysiogenetes"
[31] "Calditrichaeota"             "Elusimicrobia"               "Caldiserica"
[34] "Candidatus Saccharibacteria" "Dictyoglomi" </code></pre>
</div>
<p>Knowing phyla is helpful, but what we need to know is how many of our
OTUs have been assigned to the phylum Firmicutes?. Let´s use the command
<code>sum()</code> to ask R:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">sum</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="st">"Phylum"</span>] <span class="sc">==</span> <span class="st">"Firmicutes"</span>)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 580</code></pre>
</div>
<p>Now, to know for that phylum in particular which taxa there are in a
certain rank, we can also ask it to phyloseq.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="st">"Phylum"</span>] <span class="sc">==</span> <span class="st">"Firmicutes"</span>, <span class="st">"Class"</span>])</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Bacilli"          "Clostridia"       "Negativicutes"    "Limnochordia"     "Erysipelotrichia" "Tissierellia" </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="exploring-the-abundance-table">Exploring the abundance table<a class="anchor" aria-label="anchor" href="#exploring-the-abundance-table"></a>
</h3>
<p>Until now, we have looked at the part of the phyloseq object that
stores the information about the taxonomy (at all the possible levels)
of each OTU found in our samples. However, there is also a part of the
phyloseq object that stores the information about how many sequenced
reads corresponding to a certain OTU are in each sample. This table is
the <code>otu_table</code>.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p><a href="fig/03-07-03.png">
<img src="fig/03-07-03.png" alt="A table where the abundance of each OTU in each sample is shown. Each row represents one
  OTU and the columns represent the samples. In the intersection, a number indicates how many sequenced reads of that OTU are present in that sample." class="figure"></a> <em> Figure 3. Table of the abundance of reads in the
<code>merged_metagenomes</code> object. <em></em></em></p>
<p>We will take advantage of this information later on in our
analyses.</p>
<div id="phyloseq-objects" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="phyloseq-objects" class="callout-inner">
<h3 class="callout-title">Phyloseq objects</h3>
<div class="callout-content">
<p>Finally, we can review our object and see that all datasets (i.e.,
JC1A, JP4D, and JP41) are in the object. If you look at our Phyloseq
object, you will see that there are more data types that we can use to
build our object(<code>?phyloseq()</code>), such as a phylogenetic tree
and metadata concerning our samples. These are optional, so we will use
our basic phyloseq object, composed of the abundances of specific OTUs
and the names of those OTUs.</p>
</div>
</div>
</div>
<div id="exercise-1-explore-a-phylum" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-explore-a-phylum" class="callout-inner">
<h3 class="callout-title">Exercise 1: Explore a phylum</h3>
<div class="callout-content">
<p>Go into groups and choose one phylum that is interesting for your
group, and use the learned code to find out how many OTUs have been
assigned to your chosen phylum and what are the unique names of the
genera inside it. がんばって! (ganbatte; <em>good luck</em>):</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Change the name of a new phylum wherever needed and the name of the
rank we are asking for to get the result. As an example, here is the
solution for Proteobacteria:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="va">merged_metagenomes</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span>,<span class="st">"Phylum"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"Proteobacteria"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">unique</span><span class="op">(</span><span class="va">merged_metagenomes</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span><span class="va">merged_metagenomes</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span>,<span class="st">"Phylum"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"Proteobacteria"</span>, <span class="st">"Genus"</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-2-searching-for-the-read-counts" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-searching-for-the-read-counts" class="callout-inner">
<h3 class="callout-title">Exercise 2: Searching for the read counts</h3>
<div class="callout-content">
<p>Using the information from both the <code>tax_table</code> and the
<code>otu_table</code>, find how many reads there are for any species of
your interest (one that can be found in the
<code>tax_table</code>).<br><strong>Hint</strong>: Remember that you can access the contents of a
data frame with the <code>["row_name", "column_name"]</code>
syntax.<br>
がんばって! (ganbatte; <em>good luck</em>):</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Go to the <code>tax_table</code>:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p>Take note of the OTU number for some species:
<a href="%7B%7B%20page.root%20%7D%7D/fig/03-07-04.png">
<img src="%7B%7B%20page.root%20%7D%7D/fig/03-07-04.png" alt="The OTU number is in the leftmost space of the table as a row name for the searched species." class="figure"></a> <em> Figure 4. The row of the <code>tax_table</code> corresponds to
the species <em>Paracoccus zhejiangensis</em>. <em></em></em></p>
<p>Search for the row of the <code>otu_table</code> with the row name
you chose.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data[<span class="st">"1077935"</span>,]</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>JC1A JP4D JP41 </span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>  <span class="dv">42</span>  <span class="dv">782</span>  <span class="dv">257</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>kraken-biom formats Kraken output-files of several samples into the
single <code>.biom</code> file that will be phyloseq input.</li>
<li>The library <code>phyloseq</code> manages metagenomics objects and
computes analyses.</li>
<li>A phyloseq object stores a table with the taxonomic information of
each OTU and a table with the abundance of each OTU.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-08-Diversity-tackled-with-R"><p>Content from <a href="08-Diversity-tackled-with-R.html">Diversity Tackled With R</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/08-Diversity-tackled-with-R.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we measure diversity?</li>
<li>How can I use R to analyze diversity?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Plot alpha and beta diversity.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>Look at your fingers; controlled by the mind can do great things.
However, imagine if each one has a little brain of its own, with
different ideas, desires, and fears ¡How wonderful things will be made
out of an artist with such hands!</em> -Ode to multidisciplinarity</p>
<section><h2 class="section-heading" id="first-plunge-into-diversity">First plunge into diversity<a class="anchor" aria-label="anchor" href="#first-plunge-into-diversity"></a>
</h2>
<hr class="half-width">
<p>Species diversity, in its simplest definition, is the number of
species in a particular area and their relative abundance (evenness).
Once we know the taxonomic composition of our metagenomes, we can do
diversity analyses. Here we will discuss the two most used diversity
metrics, α diversity (within one metagenome) and β (across
metagenomes).</p>
<ul>
<li>α Diversity: Can be represented only as richness (<em>, i.e.,</em>
the number of different species in an environment), or it can be
measured considering the abundance of the species in the environment as
well (<em>i.e.,</em> the number of individuals of each species inside
the environment). To measure α-diversity, we use indexes such as
Shannon’s, Simpson’s, Chao1, etc.</li>
</ul>
<p><a href="fig/03-08-01.png"></a></p>
<figure><img src="fig/03-08-01.png" alt="Alpha diversity diagram: In lake A, we have three fishes, each one of a different species. On lake B, we have two fish, each of a different species. Moreover, we have four fish in lake C, each of different species." class="figure mx-auto d-block"></figure><p></p>
<p><em> Figure 1. Alpha diversity is calculated according to fish
diversity in a pond. Here, alpha diversity is represented in its
simplest way: Richness. <em></em></em></p>
<ul>
<li>β diversity is the difference (measured as distance) between two or
more environments. It can be measured with metrics like Bray-Curtis
dissimilarity, Jaccard distance, or UniFrac distance, to name a few.
Each one of this measures are focused on a characteristic of the
community (<em>e.g.,</em> Unifrac distance measures the phylogenetic
relationship between the species of the community).</li>
</ul>
<p>In the next example, we will look at the α and the β components of
the diversity of a dataset of fishes in three lakes. The most simple way
to calculate the β-diversity is to calculate the distinct species
between two lakes (sites). Let us take as an example the diversity
between Lake A and Lake B. The number of species in Lake A is 3. To this
quantity, we will subtract the number of these species that are shared
with the Lake B: 2. So the number of unique species in Lake A compared
to Lake B is (3-2) = 1. To this number, we will sum the result of the
same operations but now take Lake B as our reference site. In the end,
the β diversity between Lake A and Lake B is (3-2) + (3-2) = 2. This
process can be repeated, taking each pair of lakes as the focused
sites.</p>
<p><a href="fig/03-08-02.png"></a></p>
<figure><img src="fig/03-08-02.png" alt=" Alpha and Beta diversity diagram: Each Lake has a different number of species, and each species has a different number of fish individuals. Both metrics are taken into account to measure alfa and beta diversity." class="figure mx-auto d-block"></figure><p></p>
<p><em> Figure 2. Alpha and beta diversity indexes of fishes in a
pond.<em></em></em></p>
<p>If you want to read more about diversity, we recommend to you this <a href="https://link.springer.com/article/10.1007/s00442-010-1812-0" class="external-link">paper</a>
on the concept of diversity.</p>
</section><section><h2 class="section-heading" id="α-diversity">α diversity<a class="anchor" aria-label="anchor" href="#%CE%B1-diversity"></a>
</h2>
<hr class="half-width">
<table class="table">
<colgroup>
<col width="59%">
<col width="40%">
</colgroup>
<thead><tr class="header">
<th>Diversity Indices</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Shannon (H)</td>
<td>Estimation of species richness and species evenness. More weight on
richness.</td>
</tr>
<tr class="even">
<td>Simpson’s (D)</td>
<td>Estimation of species richness and species evenness. More weigth on
evenness.</td>
</tr>
<tr class="odd">
<td>Chao1</td>
<td>Abundance based on species represented by a single individual
(singletons) and two individuals (doubletons).</td>
</tr>
</tbody>
</table>
<ul>
<li>Shannon (H):</li>
</ul>
<table class="table">
<colgroup>
<col width="60%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th align="center">Variable</th>
<th align="center">Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">$ H = - \sum_{i=1}^{S} p_{i} \ln{p_{i}} $</td>
<td align="center">Definition</td>
</tr>
<tr class="even">
<td align="center">$ S $</td>
<td align="center">Number of OTUs</td>
</tr>
<tr class="odd">
<td align="center">$ p_{i} $</td>
<td align="center">The proportion of the community represented by OTU
i</td>
</tr>
</tbody>
</table>
<!-- <img src="https://render.githubusercontent.com/render/math?math=H=-\sum_{i=1}^{S}p_i\:ln{p_i}"> | Definition
<img src="https://render.githubusercontent.com/render/math?math=S"> | Number of OTUs
<img src="https://render.githubusercontent.com/render/math?math=p_i">|  The proportion of the community represented by OTU i    --><ul>
<li>Simpson’s (D)</li>
</ul>
<table class="table">
<colgroup>
<col width="60%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th align="center">Variable</th>
<th align="center">Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">$ D = \frac{1}{\sum_{i=1}^{S} p_{i}^{2}} $</td>
<td align="center">Definition</td>
</tr>
<tr class="even">
<td align="center">$ S $</td>
<td align="center">Total number of the species in the community</td>
</tr>
<tr class="odd">
<td align="center">$ p_{i} $</td>
<td align="center">Proportion of community represented by OTU i</td>
</tr>
</tbody>
</table>
<!-- <img src="https://render.githubusercontent.com/render/math?math=D=\frac{1}{\sum_{i=1}^{S}p_i^2}">| Definition
<img src="https://render.githubusercontent.com/render/math?math=S"> | Total number of the species in the community
<img src="https://render.githubusercontent.com/render/math?math=p_i" align="middle"> | Proportion of community represented by OTU i     --><ul>
<li>Chao1</li>
</ul>
<table class="table">
<colgroup>
<col width="60%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th align="center">Variable</th>
<th align="center">Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">$ S_{chao1} = S_{Obs} + \frac{F_{1} \times (F_{1} -
1)}{2 \times (F_{2} + 1)} $</td>
<td align="center">Count of singletons and doubletons respectively</td>
</tr>
<tr class="even">
<td align="center">$ F_{1}, F_{2} $</td>
<td align="center">Count of singletons and doubletons respectively</td>
</tr>
<tr class="odd">
<td align="center">$ S_{chao1}=S_{Obs} $</td>
<td align="center">The number of observed species</td>
</tr>
</tbody>
</table>
<!-- <img src="fig/equation.svg">| Definition
<img src="https://render.githubusercontent.com/render/math?math=F_1,F_2">|Count of singletons and doubletons respectively
<img src="https://render.githubusercontent.com/render/math?math=S_{chao1}=S_{Obs}">| The number of observed species   --><p><!-- comment we use https://viereck.ch/latex-to-svg/ to convert from latex to SVG because Chao equation did not render correctly with GitHub math!--></p>
<div class="section level3">
<h3 id="β-diversity">β diversity<a class="anchor" aria-label="anchor" href="#%CE%B2-diversity"></a>
</h3>
<p>Diversity β measures how different two or more communities are,
either in their composition (richness) or in the abundance of the
organisms that compose it (abundance).</p>
<ul>
<li>Bray-Curtis dissimilarity: The difference in richness and abundance
across environments (samples). Weight on abundance. Measures the
differences from 0 (equal communities) to 1 (different communities)</li>
<li>Jaccard distance: Based on the presence/absence of species
(diversity). It goes from 0 (same species in the community) to 1 (no
species in common)</li>
<li>UniFrac: Measures the phylogenetic distance; how alike the trees in
each community are. There are two types, without weights (diversity) and
with weights (diversity and abundance)</li>
</ul>
<p>There are different ways to plot and show the results of such
analysis. Among others, PCA, PCoA, or NMDS analysis are widely used.</p>
<div id="exercise-1-simple-measure-of-alpha-and-beta-diversities." class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-simple-measure-of-alpha-and-beta-diversities." class="callout-inner">
<h3 class="callout-title">Exercise 1: Simple measure of alpha and beta diversities.</h3>
<div class="callout-content">
<p>In the next picture, there are two lakes with different fish species:
<a href="%7B%7B%20page.root%20%7D%7D/fig/03-08-03.png">
<img src="%7B%7B%20page.root%20%7D%7D/fig/03-08-03.png" alt="In lake A, we have four different species, two of these species have three specimens each one. This Lake also has two specimens of a third species and only one specimen of a fourth specie. We got nine fish in total. Lake B has only three different species, the most populated species is also present in lake A and has five specimens, and we have only one specimen of each of the other two species. We got seven species total in lake B " class="figure"></a> <em> Figure 3. <em></em></em></p>
<p>Which of the options below is true for the alpha diversity in lakes
A, B, and beta diversity between lakes A and B, respectively?</p>
<ol style="list-style-type: decimal">
<li>4, 3, 1</li>
<li>4, 3, 5</li>
<li>9, 7, 16</li>
</ol>
<p>Please, paste your result on the collaborative document provided by
instructors. <em>Hic Sunt Leones!</em> (<em>Here be Lions!</em>)</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Answer: 2. 4, 3, 5 <strong>Alpha diversity</strong> in this case, is
the sum of different species. Lake <strong>A</strong> has
<strong>4</strong> different species and lake <strong>B</strong> has
<strong>3</strong> different species. <strong>Beta diversity</strong>
refers to the difference between lake A and lake B. If we use the
formula in <em>Figure 2</em> we can see that to calculate beta
diversity, we have to detect the number of species and the number of
shared species in both lakes. There is only one shared species, so we
have to subtract the number of shared species to the total species and
sum the result. In this case, in lake A, we have 4 different species and
one shared species with lake B (4-1)=3, and in lake B we have three
species and one shared species with lake A (3-1)=2. If we add 3+2, the
result is 5.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="plot-alpha-diversity">Plot alpha diversity<a class="anchor" aria-label="anchor" href="#plot-alpha-diversity"></a>
</h2>
<hr class="half-width">
<p>We want to know the bacterial diversity, so we will prune all
non-bacterial organisms in our <code>merged_metagenomes</code> Phyloseq
object. To do this, we will make a subset of all bacterial groups and
save them.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(merged_metagenomes, Kingdom <span class="sc">==</span> <span class="st">"Bacteria"</span>)</span></code></pre>
</div>
<p>Now let us look at some statistics of our metagenomes. By the output
of the <code>sample_sums()</code> command, we can see how many reads
there are in the library. Library JC1A is the smallest with 18412 reads,
while library JP4D is the largest with 149590 reads.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 4024 taxa and 3 samples ]
tax_table()   Taxonomy Table:    [ 4024 taxa by 7 taxonomic ranks ]</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">sample_sums</span>(merged_metagenomes)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  JC1A   JP4D   JP41
 18412 149590  76589 </code></pre>
</div>
<p>Also, the Max, Min, and Mean output on <code>summary()</code> can
give us a sense of the evenness. For example, the OTU that occurs more
times in the sample JC1A occurs 399 times, and on average in sample
JP4D, an OTU occurs in 37.17 reads.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">summary</span>(merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      JC1A              JP4D              JP41
 Min.   :  0.000   Min.   :   0.00   Min.   :   0.00
 1st Qu.:  0.000   1st Qu.:   3.00   1st Qu.:   1.00
 Median :  0.000   Median :   7.00   Median :   5.00
 Mean   :  4.575   Mean   :  37.17   Mean   :  19.03
 3rd Qu.:  2.000   3rd Qu.:  21.00   3rd Qu.:  14.00
 Max.   :399.000   Max.   :6551.00   Max.   :1994.00  </code></pre>
</div>
<p>To have a more visual representation of the diversity inside the
samples (i.e., α diversity), we can now look at a graph created using
Phyloseq:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes, </span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>              <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>)) </span></code></pre>
</div>
<p><a href="fig/03-08-04.png"></a></p>
<figure><img src="%7B%7B%20page.root%20%7D%7D/fig/03-08-04.png" alt="A figure divided in three
panels. Each of these panels represents a different alpha diversity index.
Inside this section, each point represents the value assigned on this index to
the three different samples. The different indexes give
different values to the same sample." class="figure mx-auto d-block"></figure><p></p>
<p><em> Figure 4. Alpha diversity indexes for both samples. <em></em></em></p>
<p>Each of these metrics can give an insight into the distribution of
the OTUs inside our samples. For example, the Chao1 diversity index
gives more weight to singletons and doubletons observed in our samples,
while Shannon is an entropy index remarking the impossibility of taking
two reads out of the metagenome “bag” and that these two will belong to
the same OTU.</p>
<div id="exercise-2-exploring-function-flags." class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-exploring-function-flags." class="callout-inner">
<h3 class="callout-title">Exercise 2: Exploring function flags.</h3>
<div class="callout-content">
<p>While using the help provided, explore these options available for
the function in <code>plot_richness()</code>:</p>
<ol style="list-style-type: decimal">
<li><code>nrow</code></li>
<li><code>sortby</code></li>
<li><code>title</code></li>
</ol>
<p>Use these options to generate new figures that show you other ways to
present the data.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>The code and the plot using the three options will look as follows:
The “title” option adds a title to the figure.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes, </span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>             <span class="at">title =</span> <span class="st">"Alpha diversity indexes for three samples from Cuatro Cienegas"</span>,</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>             <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>))</span></code></pre>
</div>
<p><a href="fig/03-08-05.png"> <img src="fig/03-08-05.png" alt="" class="figure"></a> <em> Figure 5. Alpha diversity plot with the title. <em></em></em></p>
<p>The “nrow” option arranges the graphics horizontally.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes,</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>             <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>),</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>             <span class="at">nrow=</span><span class="dv">3</span>)</span></code></pre>
</div>
<p><a href="fig/03-08-06.png"> <img src="fig/03-08-06.png" alt="" class="figure"></a> <em> Figure 6. Alpha diversity plot with the three panels arranged
in rows. <em></em></em></p>
<p>The “sortby” option orders the samples from least to greatest
diversity depending on the parameter. In this case, it is ordered by
“Shannon” and tells us that the JP4D sample has the lowest diversity and
the JP41 sample the highest.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes, </span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>             <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>),</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>             <span class="at">sortby =</span> <span class="st">"Shannon"</span>) </span></code></pre>
</div>
<p><a href="fig/03-08-07.png">
<img src="fig/03-08-07.png" alt="The same panels as before, but now the samples are arranged horizontally according to the values in the Shannon index panel." class="figure"></a> <em> Figure 7. Samples sorted by Shannon in alpha diversity index
plots. <em></em></em></p>
<p>Considering those mentioned above, together with the three graphs, we
can say that JP41 and JP4D present a high diversity concerning the JC1A.
Moreover, the diversity of the sample JP41 is mainly given by singletons
or doubletons. Instead, the diversity of JP4D is given by species in
much greater abundance. Although the values of H (Shannon) above three
are considered to have a lot of diversity.</p>
</div>
</div>
</div>
</div>
<p>A caution when comparing samples is that differences in some alpha
indexes may be the consequence of the difference in the total number of
reads of the samples. A sample with more reads is more likely to have
more different OTUs, so some normalization is needed. Here we will work
with relative abundances, but other approaches could help reduce this
bias.</p>
</section><section><h2 class="section-heading" id="absolute-and-relative-abundances">Absolute and relative abundances<a class="anchor" aria-label="anchor" href="#absolute-and-relative-abundances"></a>
</h2>
<hr class="half-width">
<p>From the read counts that we just saw, it is evident that there is a
great difference in the number of total sequenced reads in each sample.
Before we further process our data, look if we have any non-identified
reads. Marked as blank (i.e.,““) on the different taxonomic levels:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">summary</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data<span class="sc">==</span> <span class="st">""</span>)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Kingdom          Phylum          Class           Order           Family          Genus          Species
 Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical
 FALSE:4024      FALSE:4024      FALSE:3886      FALSE:4015      FALSE:3967      FALSE:3866      FALSE:3540
                                 TRUE :138       TRUE :9         TRUE :57        TRUE :158       TRUE :484      </code></pre>
</div>
<p>With the command above, we can see blanks on different taxonomic
levels. For example, we have 158 blanks at the genus level. Although we
could expect to see some blanks at the species or even at the genus
level; we will get rid of the ones at the genus level to proceed with
the analysis:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(merged_metagenomes, Genus <span class="sc">!=</span> <span class="st">""</span>) <span class="co">#Only genus that are no blank</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">summary</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data<span class="sc">==</span> <span class="st">""</span>)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Kingdom          Phylum          Class           Order           Family          Genus          Species
 Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical
 FALSE:3866      FALSE:3866      FALSE:3739      FALSE:3860      FALSE:3858      FALSE:3866      FALSE:3527
                                 TRUE :127       TRUE :6         TRUE :8                         TRUE :339 </code></pre>
</div>
<p>Next, since our metagenomes have different sizes, it is imperative to
convert the number of assigned reads (i.e., absolute abundance) into
percentages (i.e., relative abundances) to compare them.</p>
<p>Right now, our OTU table looks like this:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        JC1A JP4D JP41
1060      32  420   84
1063     316 5733 1212
2033869  135 1232  146
1850250  114  846  538
1061      42 1004  355
265       42  975  205</code></pre>
</div>
<p>To make this transformation to percentages, we will take advantage of
a function of Phyloseq.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(merged_metagenomes, <span class="cf">function</span>(x) x<span class="sc">*</span><span class="dv">100</span> <span class="sc">/</span> <span class="fu">sum</span>(x) )</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(percentages<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             JC1A      JP4D      JP41
1060    0.1877383 0.3065134 0.1179709
1063    1.8539161 4.1839080 1.7021516
2033869 0.7920211 0.8991060 0.2050447
1850250 0.6688178 0.6174056 0.7555755
1061    0.2464066 0.7327130 0.4985675
265     0.2464066 0.7115490 0.2879052</code></pre>
</div>
<p>Now, we are ready to compare the abundaces given by percantages of
the samples with beta diversity indexes.</p>
</section><section><h2 class="section-heading" id="beta-diversity">Beta diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a>
</h2>
<hr class="half-width">
<p>As we mentioned before, the beta diversity is a measure of how alike
or different our samples are (overlap between discretely defined sets of
species or operational taxonomic units). To measure this, we need to
calculate an index that suits the objectives of our research. By the
next code, we can display all the possible distance metrics that
Phyloseq can use:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="sc">&gt;</span> distanceMethodList</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$UniFrac
[1] "unifrac"  "wunifrac"

$DPCoA
[1] "dpcoa"

$JSD
[1] "jsd"

$vegdist
 [1] "manhattan"  "euclidean"  "canberra"   "bray"       "kulczynski" "jaccard"    "gower"
 [8] "altGower"   "morisita"   "horn"       "mountford"  "raup"       "binomial"   "chao"
[15] "cao"

$betadiver
 [1] "w"   "-1"  "c"   "wb"  "r"   "I"   "e"   "t"   "me"  "j"   "sor" "m"   "-2"  "co"  "cc"  "g"
[17] "-3"  "l"   "19"  "hk"  "rlb" "sim" "gl"  "z"

$dist
[1] "maximum"   "binary"    "minkowski"

$designdist
[1] "ANY"</code></pre>
</div>
<p>Describing all these possible distance metrics is beyond the scope of
this lesson, but here we show which are the ones that need a
phylogenetic relationship between the species-OTUs present in our
samples:</p>
<ul>
<li>Unifrac</li>
<li>Weight-Unifrac</li>
<li>DPCoA</li>
</ul>
<p>We do not have a phylogenetic tree or phylogenetic relationships. So
we can not use any of those three. We will use <a href="https://www.pelagicos.net/MARS6300/readings/Bray_&amp;_Curtis_1957.pdf" class="external-link">Bray-curtis</a>
since it is one of the most robust and widely used distance metrics to
calculate beta diversity.</p>
<p><strong>Let’s keep this up!</strong> We already have all we need to
begin the beta diversity analysis. We will use the Phyloseq command
<code>ordinate</code> to generate a new object where the distances
between our samples will be allocated after calculating them. For this
command, we need to specify which method we will use to generate a
matrix. In this example, we will use Non-Metric Multidimensional Scaling
or <a href="https://academic.oup.com/bioinformatics/article/21/6/730/199398" class="external-link">NMDS</a>.
NMDS attempts to represent the pairwise dissimilarity between objects in
a low-dimensional space, in this case, a two-dimensional plot.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="sc">&gt;</span> meta_ord <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(<span class="at">physeq =</span> percentages, <span class="at">method =</span> <span class="st">"NMDS"</span>, </span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>                     <span class="at">distance =</span> <span class="st">"bray"</span>)</span></code></pre>
</div>
<p>If you get some warning messages after running this script, fear not.
It is because we only have three samples. Few samples make the algorithm
warn about the lack of difficulty in generating the distance matrix.</p>
<p>By now, we just need the command <code>plot_ordination()</code> to
see the results from our beta diversity analysis:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_ordination</span>(<span class="at">physeq =</span> percentages, <span class="at">ordination =</span> meta_ord)</span></code></pre>
</div>
<p><a href="fig/03-08-08.png">
<img src="fig/03-08-08.png" alt="Plot with NMDS1 as a label in x-axis that goes from -0.4 to 0.2 and NMDS2 in y-axis that goes from -0.2 to 0.1. The plot has three dots that are not clustered in any way." class="figure"></a> <em> Figure 8. Beta diversity with NMDS of our three samples.
<em></em></em></p>
<p>In this NMDS plot, each point represents the combined abundance of
all its OTUs. As depicted, each sample occupies space in the plot
without forming any clusters. This output is because each sample is
different enough to be considered its own point in the NMDS space.</p>
<div id="exercise-3-add-metadata-to-beta-diversity-visualization" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3-add-metadata-to-beta-diversity-visualization" class="callout-inner">
<h3 class="callout-title">Exercise 3: Add metadata to beta diversity visualization</h3>
<div class="callout-content">
<p>In the following figure, the beta diversity graph we produced earlier
has been enriched. Look at the code below and answer:</p>
<ol style="list-style-type: decimal">
<li>Which instruction colored the samples by their corresponding
treatment?</li>
<li>What did the instruction <code>geom_text</code> do?</li>
<li>What other difference do you notice with our previous graph?</li>
<li>Do you see some clustering of the samples according to their
treatment?</li>
</ol>
<p><a href="fig/03-08-09.png">
<img src="fig/03-08-09.png" alt="The distance between the three samples, JC1A, JP4D and JP41 is shown in a plane. Each sample has a legend and a color. The color is according to the metadata treatment. There are three possible treatments in the legend: Control mesocosm, Fertilized pond, and Unenriched pond" class="figure"></a></p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata_cuatroc</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>Sample<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"JC1A"</span>, <span class="st">"JP4D"</span>, <span class="st">"JP41"</span><span class="op">)</span>, Treatment<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"Control mesocosm"</span>, <span class="st">"Fertilized pond"</span>, <span class="st">"Unenriched pond"</span><span class="op">)</span><span class="op">)</span> <span class="co"># Making dataframe with metadata  </span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">metadata_cuatroc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">metadata_cuatroc</span><span class="op">$</span><span class="va">Sample</span> <span class="co"># Using sample names as row names  </span></span>
<span><span class="va">percentages</span><span class="op">@</span><span class="va">sam_data</span> <span class="op">&lt;-</span> <span class="fu">sample_data</span><span class="op">(</span><span class="va">metadata_cuatroc</span><span class="op">)</span> <span class="co"># Adding metadata to sam_data table of phyloseq object percentages  </span></span>
<span><span class="va">meta_ord</span> <span class="op">&lt;-</span> <span class="fu">ordinate</span><span class="op">(</span>physeq <span class="op">=</span> <span class="va">percentages</span>, method <span class="op">=</span> <span class="st">"NMDS"</span>, distance <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span> <span class="co"># Calculating beta diversity    </span></span>
<span><span class="fu">plot_ordination</span><span class="op">(</span>physeq <span class="op">=</span> <span class="va">percentages</span>, ordination <span class="op">=</span> <span class="va">meta_ord</span>, color <span class="op">=</span> <span class="st">"Treatment"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># Plotting beta diversity.  </span></span>
<span>    <span class="fu">geom_text</span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span>, vjust <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span>   </span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>The flag <code>color = "Treatment"</code> applied a color to each
sample according to its treatment, in the <code>plot_ordination</code>
of the object <code>percentages</code>.<br>
The <code>geom_text</code> instruction added the names of the sample to
the graph. This could have added any text, with the instruction
<code>label = Sample</code> we specified to add the names of the samples
as text. With <code>size</code> we adjusted the size of the text, and
with <code>vjust</code> we adjusted the position so the text would not
overlap with the dots.<br>
There are three possible treatments, Control mesocosm, Fertilized, and
Unfertilized pond.<br>
We do not observe any kind of clustering in these three samples. More
data would show if samples with similar treatments are clustered
together.</p>
</div>
</div>
</div>
</div>
<div id="discussion-indexes-of-diversity" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="discussion-indexes-of-diversity" class="callout-inner">
<h3 class="callout-title">Discussion: Indexes of diversity</h3>
<div class="callout-content">
<p>Why do you think we need different indexes to asses diversity? What
index will you use to assess the impact of rare, low-abundance taxa?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>It will be difficult (if not impossible) to take two communities and
observe the same distribution of all members. This outcome is because
there are a lot of <strong>factors</strong> affecting these lineages.
Some of the <strong>environmental factors</strong> are temperature, pH,
and nutrient concentration. Also, the interactions of these populations,
such as competence, inhibition of other populations, and growth speed,
are an important driver of variation (<strong>biotic factor</strong>). A
combination of the <strong>factors mentioned above</strong>, can
interact to maintain some populations with low abundance (<strong>rare
taxa</strong> In order to have ways to assess hypotheses regarding which
of these processes can be affecting the community, we use all these
different indexes. Some emphasize the number of species and other the
evenness of the OTUs. To assess the impact of low abundance lineages,
one alpha diversity index widely used is the Chao1 index.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Alpha diversity measures the intra-sample diversity.</li>
<li>Beta diversity measures the inter-sample diversity.</li>
<li>Phyloseq includes diversity analyses such as alpha and beta
diversity calculation.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-09-abundance-analyses"><p>Content from <a href="09-abundance-analyses.html">Taxonomic Analysis with R</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/09-abundance-analyses.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we know which taxa are in our samples?</li>
<li>How can we compare depth-contrasting samples?</li>
<li>How can we manipulate our data to deliver a message?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Manipulate data types inside your phyloseq object.</li>
<li>Extract specific information from taxonomic-assignation data.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="explore-our-samples-at-specific-taxonomic-levels">Explore our samples at specific taxonomic levels<a class="anchor" aria-label="anchor" href="#explore-our-samples-at-specific-taxonomic-levels"></a>
</h2>
<hr class="half-width">
<p>With the taxonomic assignment information that we obtained from
Kraken, we have measured diversity, and we have visualized the taxa
inside each sample with Krona and Pavian, but Phyloseq allows us to make
this visualization more flexible and personalized. So now, we will use
Phyloseq to make abundance plots of the taxa in our samples.</p>
<p>We will start our exploration at the Phylum level. In order to group
all the OTUs that have the same taxonomy at a certain taxonomic rank, we
will use the function <code>tax_glom()</code>.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_glom <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(percentages, <span class="at">taxrank =</span> <span class="st">'Phylum'</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(percentages_glom<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p><a href="fig/03-09-01.png">
<img src="fig/03-09-01.png" alt="Table containing the
  taxonomic information of each of the OTUs inside the three samples. Here,
  we can see only the Phylum column has information, leaving the other
  taxonomic levels blank." class="figure"></a> <em> Figure 1. Taxonomic-data table
after agglomeration at the phylum level. <em></em></em></p>
<p>Another phyloseq function is <code>psmelt()</code>, which melts
phyloseq objects into a <code>data.frame</code> to manipulate them with
packages like <code>ggplot2</code> and <code>vegan</code>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(percentages_glom)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">str</span>(percentages_df)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame': 99 obs. of  5 variables:
 $ OTU      : chr  "1063" "1063" "1063" "2350" ...
 $ Sample   : chr  "JP4D" "JC1A" "JP41" "JP41" ...
 $ Abundance: num  85 73.5 58.7 23.8 19.1 ...
 $ Kingdom  : chr  "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
 $ Phylum   : chr  "Proteobacteria" "Proteobacteria" "Proteobacteria" "Bacteroidetes" ...</code></pre>
</div>
<p>Now, let’s create another data frame with the original data. This
structure will help us to compare the absolute with the relative
abundance and have a complete picture of our samples.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> absolute_glom <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(<span class="at">physeq =</span> merged_metagenomes, <span class="at">taxrank =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="sc">&gt;</span> absolute_df <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(absolute_glom)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">str</span>(absolute_df)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame': 99 obs. of  5 variables:
 $ OTU      : chr  "1063" "1063" "2350" "1063" ...
 $ Sample   : chr  "JP4D" "JP41" "JP41" "JC1A" ...
 $ Abundance: num  116538 41798 16964 12524 9227 ...
 $ Kingdom  : chr  "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
 $ Phylum   : chr  "Proteobacteria" "Proteobacteria" "Bacteroidetes" "Proteobacteria" ...</code></pre>
</div>
<p>With these objects and what we have learned regarding R data
structures and <code>ggplot2</code>, we can compare them with a plot.
First, let’s take some steps that will allow us to personalize our plot,
making it accessible for color blindness. We will create a color
palette. With <code>colorRampPalette</code>, we will choose eight colors
from the Dark2 palette and make a “ramp” with it; that is, convert those
eight colors to the number of colors needed to have one for each phylum
in our data frame. We need to have our Phylum column in the factor
structure for this.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> absolute_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(absolute_df<span class="sc">$</span>Phylum)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="sc">&gt;</span> phylum_colors_abs<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(absolute_df<span class="sc">$</span>Phylum)))</span></code></pre>
</div>
<p>Now, let´s create the figure for the data with absolute abundances
(<em>, i.e.,</em> <code>absolute_plot</code> object)</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="sc">&gt;</span> absolute_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span> absolute_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Phylum))<span class="sc">+</span> </span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> phylum_colors_abs)</span></code></pre>
</div>
<p>With the <code>position="stack"</code> command, we are telling the
<code>ggplot</code> function that the values must stack each other for
each sample. In this way, we will have all of our different categories
(OTUs) stacked in one bar and not each in a separate one. For more info
<a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></p>
<p>Next, we will create the figure for the representation of the
relative abundance data and ask RStudio to show us both plots thanks to
the <code>|</code> function from the library <code>patchwork</code>:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(percentages_df<span class="sc">$</span>Phylum)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="sc">&gt;</span> phylum_colors_rel<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(percentages_df<span class="sc">$</span>Phylum)))</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="sc">&gt;</span> relative_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>percentages_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Phylum))<span class="sc">+</span> </span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> phylum_colors_rel)</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="sc">&gt;</span> absolute_plot <span class="sc">|</span> relative_plot</span></code></pre>
</div>
<p><a href="fig/03-09-02.png">
<img src="fig/03-09-02.png" alt="A two-part plot contrasting
  the absolute versus the relative abundance of the three samples. On the right
  side, we can see how each of the bars has its own height, making it difficult
  to compare the information between samples. The right side shows
  three bars with the same height after the abundance was transformed to
  percentage inside each sample." class="figure"></a> <em> Figure 2. Taxonomic
diversity of absolute and relative abundance. <em></em></em></p>
<p>At once, we can denote the difference between the two plots and how
processing the data can enhance the display of actual results. However,
it is noticeable that we have too many taxa to adequately distinguish
the color of each one, less of the ones that hold the most incredible
abundance. In order to change that, we will use the power of data frames
and R. We will change the identification of the OTUs whose relative
abundance is less than 0.2%:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.character</span>(percentages_df<span class="sc">$</span>Phylum) <span class="co"># Return the Phylum column to be of type character</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum[percentages_df<span class="sc">$</span>Abundance <span class="sc">&lt;</span> <span class="fl">0.5</span>] <span class="ot">&lt;-</span> <span class="st">"Phyla &lt; 0.5% abund."</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(percentages_df<span class="sc">$</span>Phylum)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Proteobacteria"    "Bacteroidetes"     "Actinobacteria"    "Firmicutes"        "Cyanobacteria"
[6] "Planctomycetes"    "Verrucomicrobia"   "Phyla &lt; 0.5 abund"</code></pre>
</div>
<p>Let’s ask R to display the figures again by re-running our code:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(percentages_df<span class="sc">$</span>Phylum)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="sc">&gt;</span> phylum_colors_rel<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(percentages_df<span class="sc">$</span>Phylum)))</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="sc">&gt;</span> relative_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>percentages_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Phylum))<span class="sc">+</span> </span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> phylum_colors_rel)</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="sc">&gt;</span> absolute_plot <span class="sc">|</span> relative_plot</span></code></pre>
</div>
<p><a href="fig/03-09-03.png">
<img src="fig/03-09-03.png" alt="A new two-part plot with
  a reassignment of the low-abundant taxa on the right side. Compared to the
  left legend, the one in the right has fewer groups because of the process of
  reassigning the taxa with an abundance lower than 0.5 % to just one
  group/color." class="figure"></a> <em> Figure 3. Taxonomic diversity of absolute
and relative abundance with corrections. <em></em></em></p>
</section><section><h2 class="section-heading" id="going-further-lets-take-an-exciting-lineage-and-explore-it-thoroughly">Going further, let’s take an exciting lineage and explore it
thoroughly<a class="anchor" aria-label="anchor" href="#going-further-lets-take-an-exciting-lineage-and-explore-it-thoroughly"></a>
</h2>
<hr class="half-width">
<p>As we have already reviewed, Phyloseq offers many tools to manage and
explore data. Let’s take a look at a function we already use but now
with guided exploration. The <code>subset_taxa</code> command is used to
extract specific lineages from a stated taxonomic level; we have used it
to get rid of the reads that do not belong to bacteria with
<code>merged_metagenomes &lt;- subset_taxa(merged_metagenomes, Kingdom == "Bacteria")</code>.</p>
<p>We will use it now to extract a specific phylum from our data and
explore it at a lower taxonomic level: Genus. We will take as an example
the phylum Cyanobacteria (indeed, this is a biased and arbitrary
decision, but who does not feel attracted to these incredible
microorganisms?):</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="sc">&gt;</span> cyanos <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(merged_metagenomes, Phylum <span class="sc">==</span> <span class="st">"Cyanobacteria"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(cyanos<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="dv">2</span>])</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Cyanobacteria"</code></pre>
</div>
<p>Let’s do a little review of all that we saw today:
<strong>Transformation of the data; Manipulation of the information; and
plotting</strong>:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_percentages <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(cyanos, <span class="cf">function</span>(x) x<span class="sc">*</span><span class="dv">100</span> <span class="sc">/</span> <span class="fu">sum</span>(x) )</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_glom <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(cyanos_percentages, <span class="at">taxrank =</span> <span class="st">"Genus"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_df <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(cyanos_glom)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_df<span class="sc">$</span>Genus[cyanos_df<span class="sc">$</span>Abundance <span class="sc">&lt;</span> <span class="dv">10</span>] <span class="ot">&lt;-</span> <span class="st">"Genera &lt; 10.0 abund"</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_df<span class="sc">$</span>Genus <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(cyanos_df<span class="sc">$</span>Genus)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="sc">&gt;</span> genus_colors_cyanos<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(cyanos_df<span class="sc">$</span>Genus)))  </span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="sc">&gt;</span> plot_cyanos <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>cyanos_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Genus))<span class="sc">+</span> </span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> genus_colors_cyanos)</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="sc">&gt;</span> plot_cyanos</span></code></pre>
</div>
<p><a href="fig/03-09-05.png">
<img src="fig/03-09-05.png" alt="A new plot with three bars
  representing the absolute abundance of Cyanobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height." class="figure"></a> <em> Figure 5.
Diversity of Cyanobacteria at genus level inside our samples.<em></em></em></p>
<div id="exercise-1-taxa-agglomeration" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-taxa-agglomeration" class="callout-inner">
<h3 class="callout-title">Exercise 1: Taxa agglomeration</h3>
<div class="callout-content">
<p>With the following code, in the dataset with absolute
abundances,<br>
group together the phyla with a small number of reads to have a better
visualization of the data. Remember to check the data classes inside
your data frame.<br>
According to the <a href="https://github.com/axismaps/colorbrewer/" class="external-link">ColorBrewer</a> package
it is recommended not to have more than nine different colors in a
plot.</p>
<p>What is the correct order to run the following chunks of code?
Compare your graphs with your partners’.</p>
<p>Hic Sunt Leones! (Here be Lions!):</p>
<ol style="list-style-type: upper-alpha">
<li><p><code>absolute_df$Phylum &lt;- as.factor(absolute_df$Phylum)</code></p></li>
<li><p><code>absolute_plot &lt;- ggplot(data= absolute_df, aes(x=Sample, y=Abundance, fill=Phylum))+</code><br><code>geom_bar(aes(), stat="identity", position="stack")+</code><br><code>scale_fill_manual(values = phylum_colors_abs)</code></p></li>
<li><p><code>absolute_$Phylum[absolute_$Abundance &lt; 300] &lt;- "Minoritary Phyla"</code></p></li>
<li><p><code>phylum_colors_abs&lt;- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(absolute_df$Phylum)))</code></p></li>
<li><p><code>absolute_df$Phylum &lt;- as.character(absolute_df$Phylum)</code></p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>By grouping the samples with less than 300 reads, we can get a more
decent plot. Certainly, this will be difficult since each sample has a
contrasting number of reads.</p>
<ol start="5" style="list-style-type: upper-alpha">
<li><p><code>absolute_df$Phylum &lt;- as.character(absolute_df$Phylum)</code></p></li>
<li><p><code>absolute_df$Phylum[absolute_$Abundance &lt; 300] &lt;- "Minoritary Phyla"</code></p></li>
<li><p><code>absolute_df$Phylum &lt;- as.factor(absolute_df$Phylum)</code></p></li>
<li><p><code>phylum_colors_abs&lt;- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(absolute_df$Phylum)))</code></p></li>
<li><p><code>absolute_plot &lt;- ggplot(data= absolute_df, aes(x=Sample, y=Abundance, fill=Phylum))+</code><br><code>geom_bar(aes(), stat="identity", position="stack")+</code><br><code>scale_fill_manual(values = phylum_colors_abs)</code></p></li>
</ol>
<p>Show your plots:</p>
<p><code>absolute_plot | relative_plot</code></p>
<p><a href="fig/03-09-04.png">
<img src="fig/03-09-04.png" alt="New reassignment to the low abundant taxa on the left part of the plot. A new class has been created that contains the taxa with less than 300 reads" class="figure"></a></p>
</div>
</div>
</div>
</div>
<div id="exercise-2-recap-of-abundance-plotting" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-recap-of-abundance-plotting" class="callout-inner">
<h3 class="callout-title">Exercise 2: Recap of abundance plotting</h3>
<div class="callout-content">
<p>Match the chunk of code with its description and put them in the
correct order to create a relative abundance plot at the genus level of
a particular phylum. がんばって! (ganbatte; <em>good luck</em>):</p>
<table class="table">
<colgroup>
<col width="91%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th>Description</th>
<th>Command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>plot the relative abundance at the genus levels.</td>
<td><code>plot_proteo</code></td>
</tr>
<tr class="even">
<td>Convert all the genera with less than 3% abundance into only one
label.</td>
<td><code>proteo_percentages &lt;- transform_sample_counts(proteo, function(x) &gt;x*100 / sum(x) )</code></td>
</tr>
<tr class="odd">
<td>Make just one row that groups all the observations of the same
genus.</td>
<td><code>proteo &lt;- subset_taxa(merged_metagenomes, Phylum == "Proteobacteria")</code></td>
</tr>
<tr class="even">
<td>Create a phyloseq object only with the reads assigned to a certain
phylum.</td>
<td><code>unique(proteo@tax_table@.Data[,2])</code></td>
</tr>
<tr class="odd">
<td>Show the plot.</td>
<td><code>proteo_glom &lt;- tax_glom(proteo_percentages, taxrank = "Genus")</code></td>
</tr>
<tr class="even">
<td>Transform the phyloseq object to a data frame.</td>
<td><code>plot_proteo &lt;- ggplot(data=proteo_df, aes(x=Sample, y=Abundance, fill=Genus))+</code></td>
</tr>
<tr class="odd">
<td></td>
<td><code>geom_bar(aes(), stat="identity", position="stack")+</code></td>
</tr>
<tr class="even">
<td></td>
<td><code>scale_fill_manual(values = genus_colors_proteo)</code></td>
</tr>
<tr class="odd">
<td>Convert the Genus column into the factor structure.</td>
<td><code>proteo_df$Genus[proteo_df$Abundance &lt; 3] &lt;- "Genera &lt; 3% abund"</code></td>
</tr>
<tr class="even">
<td>Look at the phyla present in your phyloseq object.</td>
<td><code>proteo_df &lt;- psmelt(proteo_glom)</code></td>
</tr>
<tr class="odd">
<td>Convert the abundance counts to relative abundance.</td>
<td><code>genus_colors_proteo&lt;- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(proteo_df$Genus)))</code></td>
</tr>
<tr class="even">
<td>Make a palette with the appropriate colors for the number of
genera.</td>
<td><code>proteo_df$Genus &lt;- as.factor(proteo_df$Genus)</code></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a phyloseq object only with the reads assigned to a certain phylum.</span></span>
<span><span class="va">proteo</span> <span class="op">&lt;-</span> <span class="fu">subset_taxa</span><span class="op">(</span><span class="va">merged_metagenomes</span>, <span class="va">Phylum</span> <span class="op">==</span> <span class="st">"Proteobacteria"</span><span class="op">)</span></span>
<span><span class="co"># Look at the phyla present in your phyloseq object</span></span>
<span><span class="fu">unique</span><span class="op">(</span><span class="va">proteo</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Convert the abundance counts to the relative abundance</span></span>
<span><span class="va">proteo_percentages</span> <span class="op">&lt;-</span> <span class="fu">transform_sample_counts</span><span class="op">(</span><span class="va">proteo</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">*</span><span class="fl">100</span> <span class="op">/</span> <span class="fu">sum</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="co"># Make just one row that groups all the observations of the same genus.</span></span>
<span><span class="va">proteo_glom</span> <span class="op">&lt;-</span> <span class="fu">tax_glom</span><span class="op">(</span><span class="va">proteo_percentages</span>, taxrank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span></span>
<span><span class="co"># Transform the phyloseq object to a data frame</span></span>
<span><span class="va">proteo_df</span> <span class="op">&lt;-</span> <span class="fu">psmelt</span><span class="op">(</span><span class="va">proteo_glom</span><span class="op">)</span></span>
<span><span class="co"># Convert all the genera that have less than 3% of abundance into only one label</span></span>
<span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span><span class="op">[</span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Abundance</span> <span class="op">&lt;</span> <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Genera &lt; 3% abund"</span></span>
<span><span class="co"># Convert the Genus column into the factor structure</span></span>
<span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span><span class="op">)</span></span>
<span><span class="co"># Make a palette with the appropriate colors for the number of genera</span></span>
<span><span class="va">genus_colors_proteo</span><span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>,<span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span> <span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="fu">levels</span><span class="op">(</span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Plot the relative abundance at the genus levels</span></span>
<span><span class="va">plot_proteo</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span>data<span class="op">=</span><span class="va">proteo_df</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">Sample</span>, y<span class="op">=</span><span class="va">Abundance</span>, fill<span class="op">=</span><span class="va">Genus</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="op">)</span>, stat<span class="op">=</span><span class="st">"identity"</span>, position<span class="op">=</span><span class="st">"stack"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">genus_colors_proteo</span><span class="op">)</span></span>
<span><span class="co"># Show the plot</span></span>
<span><span class="va">plot_proteo</span>  </span></code></pre>
</div>
<p><a href="%7B%7B%20page.root%20%7D%7D/fig/03-09-06.png">
<img src="%7B%7B%20page.root%20%7D%7D/fig/03-09-06.png" alt="A new plot with three bars
  representing the absolute abundance of Proteobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height." class="figure"></a></p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Depths and abundances can be visualized using phyloseq.</li>
<li>The library <code>phyloseq</code> lets you manipulate metagenomic
data in a taxonomic specific perspective.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-10-OtherResources"><p>Content from <a href="10-OtherResources.html">Other Resources</a></p>
<hr>
<p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/10-OtherResources.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Where are other metagenomic resources?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Know some places where to find more information.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="other-resources">Other resources<a class="anchor" aria-label="anchor" href="#other-resources"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="cuatro-ciénegas">Cuatro Ciénegas<a class="anchor" aria-label="anchor" href="#cuatro-ci%C3%A9negas"></a>
</h3>
<p>More information about 4 Ciénegas is available in these podcasts <a href="https://www.sciencemag.org/podcast/oasis-biodiversity-mexican-desert-and-making-sound-heat" class="external-link">Mexican
oasis in the desert</a> and <a href="https://youtu.be/xMMm_GKZsnU" class="external-link">Mexican Oasis</a>, in this <a href="https://www.youtube.com/embed/VzImXRI9wYE?autoplay=1&amp;rel=0" class="external-link">Cuatro
Ciénegas Video</a>, and finally, the manuscript <a href="https://www.sciencemag.org/news/2020/06/pools-mexican-desert-are-window-earth-s-early-life" class="external-link">Cuatro
Ciénegas</a>.</p>
</div>
<div class="section level3">
<h3 id="other-tutorials-and-books">Other tutorials and books<a class="anchor" aria-label="anchor" href="#other-tutorials-and-books"></a>
</h3>
<p>Now that you finished this metagenomic lesson, you are ready to
explore on your own the universe of available tutorials. <a href="https://joey711.github.io/phyloseq/" class="external-link">Phyloseq tutorial</a>
contains much more examples of metagenomic data manipulation.<br>
The <a href="https://genomics.sschmeier.com/" class="external-link">Computational Genomics
tutorial</a> by Schmeier explains each step of the process carefully.
There is a beautiful entry called [History of metagenomics] in the Meren
Lab blog (<a href="https://merenlab.org/2020/07/27/history-of-metagenomics/" class="external-link">http://merenlab.org/2020/07/27/history-of-metagenomics/</a>)
as well as several videos explaining <a href="https://youtu.be/C3fHlccFxJw" class="external-link">Metapangenomics: A nexus between
pangenomes and metagenomes</a>, <a href="https://youtu.be/MqD4aN1p1qA" class="external-link">The power of metagenomic read
recruitment</a>, <a href="https://youtu.be/RjNdHGK4ruo" class="external-link">Genome-resolved
metagenomics: key concepts in reconstructing genomes from
metagenomes</a>. Finally, for Spanish speakers, the ISME course is a
detailed tutorial for 16s metabarcoding <a href="https://www.castrolab.org/isme/biodiversity/biodiversity.html" class="external-link">ISME
Análisis de diversidad</a>.</p>
<p>The book <a href="https://link.springer.com/book/10.1007/978-3-030-65317-0" class="external-link">Microbiomes</a>
shares a contemporary concept of microbiomes. Books <a href="https://link.springer.com/book/10.1007/978-3-030-73351-3" class="external-link">Statistical
analysis of microbiome data</a> and <a href="https://link.springer.com/book/10.1007/978-981-13-1534-3" class="external-link">statistical
analysis of microbiome data with r</a> contain a state-of-the-art
compendium of the statistical and computational microbiome analysis
techniques beyond diversity analysis.</p>
</div>
<div class="section level3">
<h3 id="other-studies-and-databases">Other studies and databases<a class="anchor" aria-label="anchor" href="#other-studies-and-databases"></a>
</h3>
<p>Some databases are <a href="https://img.jgi.doe.gov/" class="external-link">jgi</a>, <a href="https://earthmicrobiome.org/" class="external-link">The Earth microbiome project</a>, <a href="https://metasub.org/" class="external-link">metaSUB</a>, <a href="https://www.science.org/doi/10.1126/science.aap9516" class="external-link">The atlas of
soil microbiome</a>, and <a href="https://hmpdacc.org/" class="external-link">The Human
microbiome project</a>.</p>
</div>
<div class="section level3">
<h3 id="mg-rast">MG-RAST<a class="anchor" aria-label="anchor" href="#mg-rast"></a>
</h3>
<p>A valuable and easy-to-use resource is <a href="https://www.mg-rast.org/" class="external-link">MG-RAST</a>, which is at the same time a
database and an online analysis tool. MG-RAST is an online metagenomic
platform where you can upload your raw data with its corresponding
metadata and get a complete taxonomic analysis of it. MG-RAST is a great
place to get started in this type of analysis, and it is also an
extensive repository of available data for future experiments. On the
downside, it is not possible to modify significantly the steps and
parameters in the MG-RAST workflow, so there is little room when it
comes to implementing our preferred analysis tools when using
MG-RAST.</p>
<p>The Cuatro Ciénegas data that we used in the workshop is already in
MG-RAST! You can check it out <a href="https://www.mg-rast.org/mgmain.html?mgpage=project&amp;project=mgp96823" class="external-link">here</a>.</p>
<p>We can check the taxonomical distribution of our sample at different
taxonomical levels.</p>
<p>The most abundant phylum is Proteobacteria.<br><a href="../fig/03-11-02.png">
<img src="../fig/03-11-02.png" alt="Pie chart showing the relative abundance at the phylum level, and the legend with the phylum names, read count, and percentages." class="figure"></a></p>
<p>Since we have a shotgun metagenome, we can also investigate the
metabolic functions present in our sample. MG-RAST can find genes and
annotate their function through the implementation of RAST or Rapid
Annotation using Subsystems Technology. By looking at the charts
generated by this analysis, we see that most of the genes are dedicated
to metabolism.</p>
<p><a href="fig/03-11-04.png">
<img src="fig/03-11-04.png" alt="Pie chart showing the relative abundance of general functional categories, and the legend with the category names, read count, and percentages." class="figure"></a></p>
<p><a href="fig/03-11-05.png">
<img src="fig/03-11-05.png" alt="Pie chart showing the relative abundance of specific functional categories, and the legend with the category names, read count, and percentages." class="figure"></a></p>
<p>MG-RAST has its own specific pipeline, so it is a handy tool to have
a quick look at your data and also to store and share it! However, it
does not keep you from making your own personalized analysis like we
just learned!</p>
<!-- [Evomics](http://evomics.org/learning/genomics/), [Data Carpentry in 16S Metagenomics](https://datacarpentry.org/blog/2017/11/16s-dc)  -->
<div id="discussion-exploring-more-resources" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion-exploring-more-resources" class="callout-inner">
<h3 class="callout-title">Discussion: Exploring more resources</h3>
<div class="callout-content">
<p>Explore one of the suggested resources and discuss your findings with
a neighbor.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="carpentries-philosophy">Carpentries Philosophy<a class="anchor" aria-label="anchor" href="#carpentries-philosophy"></a>
</h2>
<hr class="half-width">
<p>A good lesson should be as complete and transparent that becomes easy
to teach by any instructor. Carpentries lessons are developed for the
community, and now you are part of us. This lesson is being developed,
and we are sure that you can collaborate and help us improve it.</p>
<!--## How do our results compare with the original research-->
<!-- ## How can we improve the data analysis!--->
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Enjoy metagenomics.</li>
</ul>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/aio.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/aio.html",
  "dateCreated": "2020-08-24",
  "dateModified": "2025-05-20",
  "datePublished": "2025-05-20"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

