<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Processing and Visualization for Metagenomics: Binning, Refinamiento y Desreplicación</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/lab/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/lab/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/lab/favicon-16x16.png"><link rel="manifest" href="favicons/lab/site.webmanifest"><link rel="mask-icon" href="favicons/lab/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav lab"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="lab-logo" alt="Lesson Description" src="assets/images/lab-logo.svg"><abbr class="badge peer-reviewed" title="This lesson has passed peer review.">
          <a href="https://doi.org/10.21105/jose.00209" class="external-link alert-link">
            <i aria-hidden="true" class="icon" data-feather="user-check" style="border-radius: 5px"></i>
            DOI: 10.21105/jose.00209
          </a>
        </abbr>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/BinningRefinmentDereplication.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav lab" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/lab-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: NA%" class="percentage">
    NA%
  </div>
  <div class="progress lab">
    <div class="progress-bar lab" role="progressbar" style="width: NA%" aria-valuenow="NA" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/BinningRefinmentDereplication.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-assessing-read-quality.html">2. Assessing Read Quality</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-taxonomic.html">6. Taxonomic Assignment</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-phyloseq.html">7. Exploring Taxonomy with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-Diversity-tackled-with-R.html">8. Diversity Tackled With R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-abundance-analyses.html">9. Taxonomic Analysis with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Home
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Binning, Refinamiento y Desreplicación</h1>
        <p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/instructors/BinningRefinmentDereplication.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<section><h2 class="section-heading" id="mapeo">Mapeo<a class="anchor" aria-label="anchor" href="#mapeo"></a></h2>
<hr class="half-width"><p>En los <a href="https://carpentries-lab.github.io/metagenomics-analysis/" class="external-link">días
anteriores</a> aprendieron a evaluar la calidad de las lecturas,
filtrarlas y ensamblarlas, por lo que este apartado comenzará con un
ensamble ya generado.</p>
<p>De acuerdo con el flujo de análisis (Figura 1), debemos partir de un
ensamble, mapear las lecturas y obtener un archivo de profundidad de
cada contig en el ensamble. <br></p>
<div id="archivos-necesarios-para-hacer-binning" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="archivos-necesarios-para-hacer-binning" class="callout-inner">
<h3 class="callout-title">Archivos necesarios para hacer
<em>binning</em>
</h3>
<div class="callout-content">
<p>El proceso de ensamble y mapeo son demandantes en tiempo de ejecución
y recursos. Así que nos dimos a la tarea de generar el ensamble y el
archivo de profundidad para comenzar directamente con el
<em>binning</em>. El ensamble lo corrimos con Megahit, te dejamos la
línea que usamos:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Megahit</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">megahit</span> <span class="at">-1</span> data/full/48hrs_sm_R1.fastq <span class="at">-2</span> data/full/48hrs_sm_R2.fastq <span class="at">--min-contig-len</span> 500 <span class="at">-t</span> 40 <span class="at">--presets</span> meta-sensitive <span class="at">--out-prefix</span> 48hrs <span class="at">-o</span> results/02.assembliesfull/megahit/48hrs</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># Con Metaspades podría ser asi:</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">spades.py</span> <span class="at">--meta</span>  <span class="at">-1</span> data/48hrs_sm_R1_5M.fastq <span class="at">-2</span> data/48hrs_sm_R2_5M.fastq <span class="at">-o</span> results/02.assemblies/metaspades/48hrs </span></code></pre>
</div>
<p>El mapeo lo corrimos con <a href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#introduction" class="external-link">bowtie2</a>
que es una herramienta confiable y muy utilizada para alinear lecturas
cortas a una referencia, en nuestro caso, la referencia es el ensamble
metagenómico de la muestra de 48hrs. Bowtie2 genera un archivo de mapeo
(SAM) que debe convertirse a un formato binario (BAM), para esta
conversión usamos <a href="https://github.com/samtools/" class="external-link">samtools</a>
que contiene multiples subherramientas para trabajar con archivos de
mapeos. Para generar este archivo se utilizaron las siguientes lineas de
código:</p>
<div id="callout1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Formatear el ensamble</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">bowtie2-build</span> results/02.ensambles/megahit/48hrs/48hrs.fasta results/03.profundidad/48hrs <span class="at">--threads</span> 40</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co"># Mapear las lecturas contra el ensamble</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">bowtie2</span> <span class="at">-x</span> results/03.profundidad/48hrs <span class="at">-1</span> data/48hrs_sm_R1.fastq <span class="at">-2</span> data/48hrs_sm_R2.fastq <span class="at">-p</span> 40 <span class="at">-S</span> results/03.profundidad/48hrs.sam</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co"># Convertir de SAM a BAM y ordenar</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-Sb</span> <span class="at">-O</span> BAM <span class="at">-@</span> 40 results/03.profundidad/48hrs.sam <span class="kw">|</span> <span class="ex">samtools</span> sort <span class="at">-@</span> 40  <span class="at">-o</span> results/03.profundidad/48hrs_sorted.bam</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co"># Obtener el índice</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="ex">samtools</span> index results/03.profundidad/48hrs_sorted.bam</span></code></pre>
</div>
<p>Ya que generamos el archivo bam ordenado y el índice, obtuvimos un
archivo con la información de cobertura de cada contig dentro del
ensamble, este <strong>archivo de profundidad</strong> se generó con
<code>jgi_summarize_bam_contig_depths</code> que es una herramienta
desarrollada por el JGI.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Obtener el archivo de profundidad de cada contig</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">jgi_summarize_bam_contig_depths</span>  <span class="at">--outputDepth</span> results/03.profundidad/48hrs.mgh_depth.txt results/03.profundidad/48hrs_sorted.bam</span></code></pre>
</div>
</div>
</div>
</div>
<p>No las ejecutes, sólo son un ejemplo para que las puedas usar con tus
propios datos en el futuro.</p>
</div>
</div>
</div>
<p><br></p>
<div id="ejercicio-1.-qué-información-requieren-los-programas-de-binning" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="ejercicio-1.-qué-información-requieren-los-programas-de-binning" class="callout-inner">
<h3 class="callout-title">Ejercicio 1. ¿Qué información requieren los programas de binning?</h3>
<div class="callout-content">
<p>Antes de comenzar, reúnete con tu equipo y juntos:</p>
<ul><li>Revisen nuevamente el contenido de los directorios
<code>02.ensambles</code> y <code>03.profundidad.txt</code>
</li>
<li>En una diapositiva expliquen el <code>flujo teórico</code> que se
siguió para obtener los archivos que están en esos directorios.</li>
</ul><p>Usa <a href="https://docs.google.com/presentation/d/1IRqCKRA-5kS2NxkNQdTumLO-ZrD8KiTHXF92QJ4Bb_s/edit?usp=sharing" class="external-link">esta</a>
liga de drive para ir trabajando durante el taller.</p>
<p>Sólo un miembro de cada equipo escriba en la presentación</p>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="binning">Binning<a class="anchor" aria-label="anchor" href="#binning"></a></h2>
<hr class="half-width"><p>Ahora si, vamos a agrupar los contigs del metaensamble en
<em>bins</em> …</p>
<div class="section level3">
<h3 id="metabat2">Metabat2<a class="anchor" aria-label="anchor" href="#metabat2"></a></h3>
<p><a href="https://bitbucket.org/berkeleylab/metabat/src/master/" class="external-link">Metabat2</a>
es una herramienta que agrupa los contigs tomando la cobertura de cada
contig y calcula su composición nucleotídica. <br></p>
<p style="text-align: center;">
<a href="https://doi.org/10.7717/peerj.1165" target="_blank" class="external-link">
<img src="fig/extrasMAGs/04.Metabat.png" alt="Metabat2. Kang et al., 2015. DOI:10.7717/peerj.1165" width="573" class="figure"></a> <br><em>Metabat2. Kang et al., 2015. <a href="DOI:10.7717/peerj.1165" class="uri">DOI:10.7717/peerj.1165</a></em>
</p>
<p><br> Para correr metabat necesitamos activar el <a href="https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#activating-an-environment" class="external-link">ambiente
conda</a> donde se aloja. ~~~ conda activate binning ~~~ {: .bash} <br>
Ahora que ya tenemos el ambiente activado ejecutemos metabat:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">metabat2</span> <span class="at">-i</span> results/02.ensambles/48hrs.fasta <span class="at">-a</span> results/03.profundidad/48hrs.mgh_depth.txt <span class="at">-o</span> results/04.metabat/metabat <span class="at">-t</span> 4 <span class="at">-m</span> 1500</span></code></pre>
</div>
<p><br> Se sabe que el valor mínimo de contig para reducir errores es
2000, lo puedes ver en la <a href="https://static-content.springer.com/esm/art%3A10.1038%2Fnmeth.3103/MediaObjects/41592_2014_BFnmeth3103_MOESM187_ESM.pdf" class="external-link">figura
6 de este artículo</a>. <br> &gt; ## Responde &gt; ¿Cuántos bins se
formaron? &gt; ¿Qué parámetros cambiarías o agregarías? &gt;&gt; ##
Solución &gt;&gt; <code>ls results/04.metabat/</code> &gt;&gt; &gt;&gt;
<code>metabat2 –-help</code> &gt; {: .solution} {: .challenge}</p>
<p>Ya que corrimos Metabat2 vamos a ejecutar MaxBin2, pero primero
necesitamos desactivar el ambiente:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="maxbin2">MaxBin2<a class="anchor" aria-label="anchor" href="#maxbin2"></a></h3>
<p><a href="https://sourceforge.net/projects/maxbin/files/" class="external-link">MaxBin2</a>
agrupa los contigs de acuerdo a la información de cobertura, composición
nucleotídica y <code>marcadores de copia única</code>. <br></p>
<p style="text-align: center;">
<a href="https://doi.org/10.1186/2049-2618-2-26" target="_blank" class="external-link">
<img src="fig/extrasMAGs/05.Maxbin.png" alt="MaxBin2. Wu et al., 2014. https://doi.org/10.1186/2049-2618-2-26" width="573" class="figure"></a> <br><em>MaxBin2. Wu et al., 2014. <a href="https://doi.org/10.1186/2049-2618-2-26" class="external-link uri">https://doi.org/10.1186/2049-2618-2-26</a></em>
</p>
<p><br></p>
<p>Vamos a ejecutarlo, activemos el ambiente.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">conda</span> activate metagenomics</span></code></pre>
</div>
<p>Crea el directorio para los resultados de MaxBin2</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/05.maxbin</span></code></pre>
</div>
<p>Ahora si, vamos a ejecutarlo.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">run_MaxBin.pl</span> <span class="at">-thread</span> 4 <span class="at">-min_contig_length</span> 1500 <span class="at">-contig</span> results/02.ensambles/48hrs.fasta <span class="at">-out</span> results/05.maxbin/48hrs_maxbin <span class="at">-abund</span> results/03.profundidad/48hrs.mgh_depth.txt</span></code></pre>
</div>
<div id="responde" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="responde" class="callout-inner">
<h3 class="callout-title">Responde</h3>
<div class="callout-content">
<p>¿Cuántos bins se formaron? ¿Qué porcentaje de completitud tienen?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Solución
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p><code>ls results/05.maxbin/*.fasta | wc -l</code></p>
<p><code>cat results/05.maxbin/48hrs_maxbin.summary | column -t</code></p>
</div>
</div>
</div>
</div>
<p><br> Desactiva el ambiente</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre>
</div>
<p><br> ### Vamb</p>
<p><a href="https://vamb.readthedocs.io/en/latest/" class="external-link">VAMB</a> utiliza una
combinación de enfoques de aprendizaje profundo y técnicas de
agrupamiento basándose en sus patrones de composición de nucleótidos y
en la co-ocurrencia de sus frecuencias de cobertura.</p>
<p><br> Activa el ambiente binning</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">conda</span> activate binning</span></code></pre>
</div>
<p>Vamos a correr vamb, pero primero crea el directorio de
resultados</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/06.vamb</span></code></pre>
</div>
<p>Ejecutemos vamb:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">nohup</span> vamb <span class="at">--fasta</span> results/02.ensambles/48hrs.fasta <span class="at">--jgi</span> results/03.profundidad/48hrs.mgh_depth.txt <span class="at">--minfasta</span> 500000 <span class="at">--outdir</span> results/06.vamb/48hrs <span class="op">&gt;</span> outs/06.vamb.nohup <span class="kw">&amp;</span></span></code></pre>
</div>
<div id="responde-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="responde-1" class="callout-inner">
<h3 class="callout-title">Responde</h3>
<div class="callout-content">
<p>Si quisieras recuperar los genomas de virus ¿Qué parámetro
cambiarías?</p>
</div>
</div>
</div>
<div id="otros-programas-para-binning" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="otros-programas-para-binning" class="callout-inner">
<h3 class="callout-title">Otros programas para binning</h3>
<div class="callout-content">
<p>Recientemente se publicó COMEBin, que utiliza un enfoque distinto a
lo que hemos usado en este tutorial. En el siguiente <a href="https://github.com/ziyewang/COMEBin" class="external-link">link</a> encontrarás el
manual y una explicación general sobre su funcionamiento.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="refinamiento">Refinamiento<a class="anchor" aria-label="anchor" href="#refinamiento"></a></h2>
<hr class="half-width"><p>Ya corrimos tres programas de <em>binning</em>, pero, recordemos que
los agrupamientos pueden tener errores:</p>
<br><p style="text-align: center;">
<a href="fig/extrasMAGs/06.Refinamiento.png">
<img src="fig/extrasMAGs/06.Refinamiento.png" alt="Refinamiento" width="373" class="figure"></a>
</p>
<p><br></p>
<p>Para disminuir la contaminación e incrementar la completitud hay
algunos programas que pueden ayudarnos. Entre ellos están
Binning_refiner y DASTool.</p>
<div class="section level3">
<h3 id="checkm">CheckM<a class="anchor" aria-label="anchor" href="#checkm"></a></h3>
<p>Antes de proceder al refinamiento es necesario tener claro cómo se
evalúa la completitud y contaminación de los <em>bins.</em> Para esta
evaluación se usa <a href="https://github.com/Ecogenomics/CheckM/wiki/Workflows#lineage-specific-workflow" class="external-link">CheckM</a>
que se ha convertido en una herramienta estándar para la evaluación de
la calidad de genomas y MAGs, y es usada por la mayoría de programas de
refinamiento. <br> Para hacer esta evaluación, CheckM utiliza una serie
de herramientas: <code>tree</code> organiza los genomas en un árbol de
referencia. <code>tree_qa</code> evalúa la cantidad de genes marcadores
filogenéticos y su ubicación en el árbol. El comando
<code>lineage_set</code> crea un archivo de marcadores específicos de
linaje, que se usa en el comando <code>analyze</code> para evaluar la
integridad y contaminación de los genomas. Finalmente, el comando
<code>qa</code> genera tablas que resumen la calidad de los genomas.</p>
<br><p style="text-align: center;">
<a href="https://genome.cshlp.org/content/25/7/1043/T3.expansion.html" target="_blank" class="external-link">
<img src="fig/extrasMAGs/07.CheckM_workflow.png" alt="CheckM. Parks et al., 2015. https://genome.cshlp.org/content/25/7/1043.full" class="figure"></a> <br><em>CheckM. Parks et al., 2015. <a href="https://genome.cshlp.org/content/25/7/1043.full" class="external-link uri">https://genome.cshlp.org/content/25/7/1043.full</a></em>
</p>
<p><br></p>
<p>En este taller no vamos a correr CheckM porque los programas de
refinamiento que usaremos ya lo corren de forma interna, sin embargo, es
útil correrlo para tener una idea clara sobre la calidad de los bins que
obtengamos.</p>
<p>Te dejamos la siguiente línea para que la uses en tus proyectos.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="co">#ejemplo de como correrlo con los bins de vamb</span></span>
<span><span class="co">#se debe activar el ambiente metagenomics</span></span>
<span><span class="co">#checkm lineage_wf results/06.vamb/48hrs/bins results/checkm/ -x fna -t 4  -f results/checkm/checkm_vamb_bins.txt</span></span></code></pre>
</div>
<p>Y una captura de ejemplo de como se ve la salida de CheckM:</p>
<br><p style="text-align: center;">
<a href="fig/extrasMAGs/08.checkm_example.png">
<img src="fig/extrasMAGs/08.checkm_example.png" alt="CheckM ejemplo" width="828" class="figure"></a>
</p>
<p><br></p>
<p>Y ahora si, a refinar los <em>bins</em> …</p>
</div>
<div class="section level3">
<h3 id="binning_refiner">Binning_refiner<a class="anchor" aria-label="anchor" href="#binning_refiner"></a></h3>
<p><a href="https://doi.org/10.1093/bioinformatics/btx086" class="external-link">Binning_refiner</a>
se enfoca en refinar y fusionar los bins para mejorar la integridad y
reducir la contaminación. Identifica <em>bins</em> que pueden
representar el mismo genoma y los fusiona. Después elimina posibles
contaminaciones, durante el proceso, Binning_refiner evalúa la calidad
de los bins.</p>
<br><p style="text-align: center;">
<a href="https://doi.org/10.1093/bioinformatics/btx086" target="_blank" class="external-link">
<img src="fig/extrasMAGs/09.Binning_refiner.png" alt="Binning_refiner. Wei-Zhi &amp; Torsten, 2017." width="578" class="figure"></a> <br><em>Binning_refiner. Wei-Zhi &amp; Torsten, 2017. <a href="https://doi.org/10.1093/bioinformatics/btx086" class="external-link uri">https://doi.org/10.1093/bioinformatics/btx086</a></em>
</p>
<p><br></p>
<p>Necesitamos crear el directorio de resultados para binning_refiner y
un directorio con los bins generados por cada programa</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/07.binning_refiner/48hrsbins/<span class="dt">{metabat</span><span class="op">,</span><span class="dt">maxbin</span><span class="op">,</span><span class="dt">vamb}</span></span></code></pre>
</div>
<p><br> Ahora vamos a crear ligas simbólicas de los <em>bins</em>
generados por cada herramienta.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co">#metabat</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="bu">cd</span> results/07.binning_refiner/48hrsbins/metabat/</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../../04.metabat/<span class="pp">*</span>.fa .</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#maxbin</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="bu">cd</span> ../maxbin/</span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../../05.maxbin/<span class="pp">*</span>.fasta .</span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co"># vamb</span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="bu">cd</span> ../vamb/</span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../../06.vamb/48hrs/bins/<span class="pp">*</span>.fna .</span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="co">#regresar</span></span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="bu">cd</span> ../../</span></code></pre>
</div>
<p><br> Ahora si, corramos Binning_refiner</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">Binning_refiner</span> <span class="at">-i</span> 48hrsbins/ <span class="at">-p</span> 48hrs</span></code></pre>
</div>
<p><br> Y regresemos a nuestro directorio principal</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="bu">cd</span> <span class="kw">&amp;&amp;</span> <span class="bu">cd</span> taller_metagenomica_pozol/</span></code></pre>
</div>
<p><br></p>
<p>Exploremos los resultados!</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">cat</span> results/07.binning_refiner/48hrs_Binning_refiner_outputs/48hrs_sources_and_length.txt</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">MARKDOWN<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode markdown" tabindex="0"><code class="sourceCode markdown"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>Refined_bin     Size(Kbp)       Source</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>48hrs_1 1535.49 48hrs_maxbin.004.fasta,metabat.5.fa,676.fna</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>48hrs_2 1506.01 48hrs_maxbin.002.fasta,metabat.3.fa,6952.fna</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>48hrs_3 1319.12 48hrs_maxbin.008.fasta,metabat.2.fa,28067.fna</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>48hrs_4 1263.79 48hrs_maxbin.005.fasta,metabat.9.fa,3736.fna</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>48hrs_5 1185.25 48hrs_maxbin.001.fasta,metabat.11.fa,15732.fna</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>48hrs_6 1052.67 48hrs_maxbin.003.fasta,metabat.4.fa,15732.fna</span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>48hrs_7 557.49  48hrs_maxbin.006.fasta,metabat.1.fa,28990.fna</span></code></pre>
</div>
<p><br></p>
<div id="responde-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="responde-2" class="callout-inner">
<h3 class="callout-title">Responde</h3>
<div class="callout-content">
<p>¿Cuántos bins analizó Binning_refiner y cuántos quedaron después del
refinamiento? ¿Hubo separación de algunos bins para formar otros?</p>
</div>
</div>
</div>
<div id="sankey-plot" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="sankey-plot" class="callout-inner">
<h3 class="callout-title">Sankey plot</h3>
<div class="callout-content">
<p>Puedes generar tu propio sankey plot para visualizar los resultados
de Binning_refiner. Te dejamos el código por si te es útil.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  código
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cargar las librerias</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">networkD3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># revisa tu ubicación</span></span>
<span><span class="fu">getwd</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># OJO</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="st">"/home/ELALUMNOQUEERES/taller_metagenomica_pozol"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cargar los datos</span></span>
<span><span class="va">sankey_data</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"results/07.binning_refiner/48hrs_Binning_refiner_outputs/48hrs_sankey.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Crear una lista de nodos únicos</span></span>
<span><span class="va">nodes</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>name <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">sankey_data</span><span class="op">$</span><span class="va">C1</span>, <span class="va">sankey_data</span><span class="op">$</span><span class="va">C2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Crear el dataframe de enlaces</span></span>
<span><span class="va">links</span> <span class="op">&lt;-</span> <span class="va">sankey_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>source <span class="op">=</span> <span class="fu">match</span><span class="op">(</span><span class="va">C1</span>, <span class="va">nodes</span><span class="op">$</span><span class="va">name</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>         target <span class="op">=</span> <span class="fu">match</span><span class="op">(</span><span class="va">C2</span>, <span class="va">nodes</span><span class="op">$</span><span class="va">name</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>         value <span class="op">=</span> <span class="va">Length_Kbp</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">source</span>, <span class="va">target</span>, <span class="va">value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Crear el gráfico Sankey</span></span>
<span><span class="va">sankey_plot</span> <span class="op">&lt;-</span> <span class="fu">sankeyNetwork</span><span class="op">(</span>Links <span class="op">=</span> <span class="va">links</span>, Nodes <span class="op">=</span> <span class="va">nodes</span>,</span>
<span>                             Source <span class="op">=</span> <span class="st">"source"</span>, Target <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>                             Value <span class="op">=</span> <span class="st">"value"</span>, NodeID <span class="op">=</span> <span class="st">"name"</span>,</span>
<span>                             fontSize <span class="op">=</span> <span class="fl">12</span>, nodeWidth <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Mostrar el gráfico</span></span>
<span><span class="va">sankey_plot</span></span>
<span><span class="co"># Guardar</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">htmlwidgets</span><span class="op">)</span></span>
<span><span class="fu">saveWidget</span><span class="op">(</span><span class="va">sankey_plot</span>, file <span class="op">=</span> <span class="st">"48hrs_sankey_plot.html"</span><span class="op">)</span></span></code></pre>
</div>
<p style="text-align: center;">
<a href="fig/extrasMAGs/10.BRsankeyEx.png">
<img src="fig/extrasMAGs/10.BRsankeyEx.png" alt="Binning Refiner output" width="528" class="figure"></a>
</p>
</div>
</div>
</div>
</div>
<p><br><br></p>
</div>
<div class="section level3">
<h3 id="dastool">DASTool<a class="anchor" aria-label="anchor" href="#dastool"></a></h3>
<p><a href="https://github.com/cmks/DAS_Tool" class="external-link">DASTool</a> es una
herramienta utilizada para mejorar la calidad de los <em>bins</em>.
Evalúa la integridad, combina los resultados de diferentes
<em>bineadores</em> y por consenso selecciona los mejores <em>bins</em>
de cada herramienta. Una vez que DASTool ha seleccionado los mejores
bins, realiza un proceso de refinamiento para optimizar los
resultados.</p>
<br><p style="text-align: center;">
<a href="https://doi.org/10.1038/s41564-018-0171-1." target="_blank" class="external-link">
<img src="fig/extrasMAGs/11.DASTool.png" alt="DASTool. Sieber et al., 2018." width="678" class="figure"></a> <br><em>DASTool. Sieber et al., 2018. <a href="https://doi.org/10.1038/s41564-018-0171-1" class="external-link uri">https://doi.org/10.1038/s41564-018-0171-1</a>.</em>
</p>
<p><br></p>
<p>Vamos a correr DASTool … <br> Primero crea el directorio para los
resultados</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/08.dastool</span></code></pre>
</div>
<p>DASTool necesita como entrada un archivo tabular con información de
los resultados de cada programa de <em>binning.</em></p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">Fasta_to_Contig2Bin.sh</span> <span class="at">-i</span> results/04.metabat/ <span class="at">-e</span> fa <span class="op">&gt;</span> results/08.dastool/48hrs_metabat.dastool.tsv</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="ex">Fasta_to_Contig2Bin.sh</span> <span class="at">-i</span> results/05.maxbin/ <span class="at">-e</span> fasta <span class="op">&gt;</span> results/08.dastool/48hrs_maxbin.dastool.tsv</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="ex">Fasta_to_Contig2Bin.sh</span> <span class="at">-i</span> results/06.vamb/48hrs/bins/ <span class="at">-e</span> fna <span class="op">&gt;</span> results/08.dastool/48hrs_vamb.dastool.tsv</span></code></pre>
</div>
<p>Ya que tenemos los archivos tsv podemos empezar con el
refinamiento!!</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">DAS_Tool</span> <span class="at">-i</span> results/08.dastool/48hrs_metabat.dastool.tsv,results/08.dastool/48hrs_maxbin.dastool.tsv,results/08.dastool/48hrs_vamb.dastool.tsv <span class="at">-l</span> metabat,maxbin,vamb <span class="at">-c</span> results/02.ensambles/48hrs.fasta <span class="at">-o</span> results/08.dastool/48hrs <span class="at">-t</span> 4 <span class="at">--write_bins</span></span></code></pre>
</div>
<p><br></p>
<div id="responde-3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="responde-3" class="callout-inner">
<h3 class="callout-title">Responde</h3>
<div class="callout-content">
<p>¿Cuántos bins quedaron después del refinamiento con DASTool? ¿Más o
menos que con Binning_refiner? ¿Qué calidad tienen?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="desreplicación">Desreplicación<a class="anchor" aria-label="anchor" href="#desreplicaci%C3%B3n"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="drep">dRep<a class="anchor" aria-label="anchor" href="#drep"></a></h3>
<p>La desreplicación es el proceso de identificar conjuntos de genomas
que son “iguales” en una lista de genomas y eliminar todos los genomas
excepto el “mejor” de cada conjunto redundante. <a href="https://drep.readthedocs.io/en/latest/overview.html" class="external-link">dRep</a> es
una herramienta útil para esto. Utiliza distancias Mash y ANIm para
discriminar entre MAGs iguales.</p>
<br><p style="text-align: center;">
<a href="https://academic.oup.com/ismej/article/11/12/2864/7537826" target="_blank" class="external-link">
<img src="fig/extrasMAGs/12.dRep.png" alt="DASTool. Olm et al., 2017." width="678" class="figure"></a> <br><em>DASTool. Olm et al., 2017. <a href="https://academic.oup.com/ismej/article/11/12/2864/7537826" class="external-link uri">https://academic.oup.com/ismej/article/11/12/2864/7537826</a>.</em>
</p>
<p><br></p>
<p>Ya que tenemos los resultados de los dos refinadores ejecutaremos
dRep para desreplicar y seleccionar el mejor representante de cada
<em>bin</em>.</p>
<p><br> Primero vamos a crear el directorio de resultados para dRep.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/09.drep/bins</span></code></pre>
</div>
<p>Y entraremos al directorio bins dentro del directorio de resultados
para colocar los bins que queremos comparar. En este caso los generados
por ambos refinadores (pero podrían ser los bins refinados de cada punto
de muestreo).</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="bu">cd</span> results/09.drep/bins/</span></code></pre>
</div>
<p>Con las siguientes lineas podemos copiar los bins en este directorio:
<br> Primero los de DASTool</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="va">$(</span><span class="fu">ls</span> ../../08.dastool/48hrs_DASTool_bins/<span class="pp">*</span>.fa<span class="va">)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>    <span class="va">name</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$i</span> .fa<span class="va">)</span><span class="kw">;</span> <span class="fu">cp</span> <span class="va">$i</span> <span class="va">$name</span><span class="st">".fasta"</span></span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p><br> Y ahora los de Binning_refiner</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">cp</span> ../../07.binning_refiner/48hrs_Binning_refiner_outputs/48hrs_refined_bins/<span class="pp">*</span>.fasta .</span></code></pre>
</div>
<p><br> Ya que los copiamos, regresemos al directorio principal.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="bu">cd</span> <span class="kw">&amp;&amp;</span> <span class="bu">cd</span> taller_metagenomica_pozol/</span></code></pre>
</div>
<p><br> Y ahora si, vamos a correr dRep …</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="bu">export</span> <span class="va">PATH</span><span class="op">=</span>/miniconda3/envs/metagenomics/bin:<span class="va">$PATH</span></span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a><span class="fu">nohup</span> dRep dereplicate results/09.drep/ <span class="at">-d</span> <span class="at">-comp</span> 50 <span class="at">-con</span> 10 <span class="at">--SkipSecondary</span> <span class="at">-g</span> results/09.drep/bins/<span class="pp">*</span>.fasta <span class="op">&gt;</span> outs/09.drep.nohup <span class="kw">&amp;</span></span></code></pre>
</div>
<p><br> &gt; ### Nota &gt; &gt; El argumento
<code>--SkipSecondary</code> no se aconseja poner, en la vida real
queremos que se hagan todas las agrupaciones &gt; para discriminar
genomas. En este ejemplo fue necesario ponerlo porque no logramos llamar
a ANIm {: .callout} <br></p>
<p>Este es uno de los plots generados por dRep, que representa los
mejores bins desreplicados.</p>
<br><p style="text-align: center;">
<a href="fig/extrasMAGs/13.dRepWinningGenomes.png">
<img src="fig/extrasMAGs/13.dRepWinningGenomes.png" alt="dRep output" class="figure"></a>
</p>
<p><br></p>
<p>Vamos a desactivar el ambiente de dRep</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre>
</div>
<div id="responde-4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="responde-4" class="callout-inner">
<h3 class="callout-title">Responde</h3>
<div class="callout-content">
<p>¿Cuántos clústers se formaron? ¿Cuántos MAGs son únicos?</p>
</div>
</div>
</div>
<hr><div id="discusión" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="discusión" class="callout-inner">
<h3 class="callout-title">Discusión</h3>
<div class="callout-content">
<p>Para tomar en cuenta</p>
<ul><li>En la vida real, si el proyecto de metagenómica que estás
desarrollando tiene librerías de diferentes muestras usarías dRep entre
todos los conjuntos de <em>bins</em> ya refinados para no tener
redundancia de genomas.</li>
<li>Qué harías si antes de desreplicar tienes un bin que tiene 98 % de
completitud y 11 % de contaminación?. dRep en automático lo
descartaría.</li>
</ul><p>Propondrías alguna manera para quedarte con este bin y curarlo para
reducir su contaminación?</p>
<p>Por suerte hay más programas que pueden ayudarnos a curar nuestros
bins manualmente, una herramienta útil para esto es <a href="https://kasperskytte.github.io/mmgenome2/articles/mmgenome2.html" class="external-link"><strong>mmgenome2</strong></a></p>
<p>Ya que tenemos los <em>bins</em> refinados y desreplicados
opcionalmente podrías reensamblarlos. La manera sería mapear las
lecturas de toda la muestra a los <em>bins</em> finales y con las
lecturas mapeadas y el <em>bin,</em> generar un ensamble genómico para
cada uno. Con esta aproximación se genera un MAG más pulido y la
contaminación se reduce.</p>
<p>Aunque en muchos reportes verás que los autores reensamblan sus MAGs,
en otros no lo hacen y no hacerlo no está mal, pero hacerlo mejora la
calidad.</p>
</div>
</div>
</div>
<hr><p><br></p>
<p>Ahora te toca a tí</p>
<div id="ejercicio-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="ejercicio-2" class="callout-inner">
<h3 class="callout-title">Ejercicio 2</h3>
<div class="callout-content">
<p>Ahora te toca a tí.</p>
<ul><li>Reúnanse en equipos y repliquen todo el flujo hasta este punto con
la muestra que les toca.</li>
<li>Discutan cada resultado obtenido.</li>
<li>En la <a href="https://docs.google.com/presentation/d/1IRqCKRA-5kS2NxkNQdTumLO-ZrD8KiTHXF92QJ4Bb_s/edit?usp=sharing" class="external-link">carpeta
compartida de Drive</a> busquen la diapositiva para el Ejercicio 2.</li>
<li>En la diapositiva correspondiente resuman sus resultados
obtenidos.</li>
</ul><p>Tiempo de actividad (2 hr)</p>
<p>Tiempo de presentación de resultados (5 min por equipo)</p>
</div>
</div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Home
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/instructors/BinningRefinmentDereplication.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/BinningRefinmentDereplication.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Binning, Refinamiento y Desreplicación",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/BinningRefinmentDereplication.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/BinningRefinmentDereplication.html",
  "dateCreated": "2020-08-24",
  "dateModified": "2025-04-06",
  "datePublished": "2025-06-10"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

