<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Data Processing and Visualization for Metagenomics: All Images</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/lab/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/lab/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/lab/favicon-16x16.png">
<link rel="manifest" href="favicons/lab/site.webmanifest">
<link rel="mask-icon" href="favicons/lab/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav lab"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="lab-logo" alt="Lesson Description" src="assets/images/lab-logo.svg"><abbr class="badge peer-reviewed" title="This lesson has passed peer review.">
          <a href="https://doi.org/10.21105/jose.00209" class="external-link alert-link">
            <i aria-hidden="true" class="icon" data-feather="user-check" style="border-radius: 5px"></i>
            DOI: 10.21105/jose.00209
          </a>
        </abbr>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/images.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav lab" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/lab-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress lab">
    <div class="progress-bar lab" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/images.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-assessing-read-quality.html">2. Assessing Read Quality</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-taxonomic.html">6. Taxonomic Assignment</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-phyloseq.html">7. Exploring Taxonomy with R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-Diversity-tackled-with-R.html">8. Diversity Tackled With R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-abundance-analyses.html">9. Taxonomic Analysis with R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="01-background-metadata"><h2 class="section-heading"><a href="01-background-metadata.html">Starting a Metagenomics Project</a></h2>
<hr class="half-width">
<h3 id="01-background-metadata-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Flow chart that shows the steps of a metagenomics project: Experimental design, Sampling, DNA extraction, Sequencing, Read quality, Assembly, Binning, Bin quality and Data analysis ’</p>
<a href="fig/03-01-01.png">
<img src="fig/03-01-01.png" alt="Flow chart that shows the steps of a metagenomics project: Experimental design, Sampling, DNA extraction, Sequencing, Read quality, Assembly, Binning, Bin quality and Data analysis " class="figure"></a><hr>
<h3 id="01-background-metadata-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Photography of a pond in Cuatro Ciénegas’</p>
<a href="fig/03-01-02.jpeg">
<img src="fig/03-01-02.jpeg" alt="Photography of a pond in Cuatro Ciénegas" class="figure"></a><hr></section><section id="02-assessing-read-quality"><h2 class="section-heading"><a href="02-assessing-read-quality.html">Assessing Read Quality</a></h2>
<hr class="half-width">
<h3 id="02-assessing-read-quality-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Flow diagram that shows the steps: Sequence reads, Quality control, Assembly, Binning and Taxonomy’</p>
<a href="fig/03-02-01.png">
<img src="fig/03-02-01.png" alt="Flow diagram that shows the steps: Sequence reads, Quality control, Assembly, Binning and Taxonomy" class="figure"></a><hr>
<h3 id="02-assessing-read-quality-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Flow diagram that shows the steps: Sequence reads and Quality control.’</p>
<a href="fig/03-02-02.png">
<img src="fig/03-02-02.png" alt="Flow diagram that shows the steps: Sequence reads and Quality control." class="figure"></a><hr>
<h3 id="02-assessing-read-quality-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Quality graph showing a boxplot for each sequence position in the x-axis, and the Phred score in the y-axis. The background is colored red for the Phred scores 0 to 20, yellow for the scores 20 to 28, and green for the scores 28 to 38. All of the boxes for each position are in the green area.’</p>
<a href="fig/03-02-03.png">
<img src="fig/03-02-03.png" alt="Quality graph showing a boxplot for each sequence position in the x-axis, and the Phred score in the y-axis. The background is colored red for the Phred scores 0 to 20, yellow for the scores 20 to 28, and green for the scores 28 to 38. All of the boxes for each position are in the green area." class="figure"></a><hr>
<h3 id="02-assessing-read-quality-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘Graphic of boxplots, where the first ones are in the good range of scores of the y-axis and extend to the acceptable and bad ranges of scores toward the right of the x-axis’</p>
<a href="fig/03-02-04.png">
<img src="fig/03-02-04.png" alt="Graphic of boxplots, where the first ones are in the good range of scores of the y-axis and extend to the acceptable and bad ranges of scores toward the right of the x-axis" class="figure"></a><hr></section><section id="03-trimming-filtering"><h2 class="section-heading"><a href="03-trimming-filtering.html">Trimming and Filtering</a></h2>
<hr class="half-width">
<h3 id="03-trimming-filtering-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Diagram showing the parts of the sequence that are reviewed by each parameter and the parts that are maintained or discarded at the end of the process. The Illuminaclip parameter removes the adapters, and the SlidingWindow scans the read by sections and removes a part of the read below the quality threshold. We remain with a trimmed read that has a valid quality.’</p>
<a href="fig/03-03-01.png">
<img src="fig/03-03-01.png" alt="Diagram showing the parts of the sequence that are reviewed by each parameter and the parts that are maintained or discarded at the end of the process. The Illuminaclip parameter removes the adapters, and the SlidingWindow scans the read by sections and removes a part of the read below the quality threshold. We remain with a trimmed read that has a valid quality." class="figure"></a><hr></section><section id="04-assembly"><h2 class="section-heading"><a href="04-assembly.html">Metagenome Assembly</a></h2>
<hr class="half-width">
<h3 id="04-assembly-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Three diagrams depicting the three assembly algorithms: The Greedy extension starts with any read, extends it whit the reads that make a match to make a contig, it continues with a different read when the previous contig can not be extended anymore. The Overlap Layout consensus finds every pairwise overlap, makes a layout graph with all the overlaps and chooses consensus sequences to make the contigs. The De Bruijn Graphs divides the reads in k-mers, makes a k-mer graph that shows all the overlapping k-mers, and chooses paths from the graph to make the contigs. ’</p>
<figure><img src="fig/03-04-01.png" width="868" height="777" alt="Three diagrams depicting the three assembly algorithms: The Greedy extension starts with any read, extends it whit the reads that make a match to make a contig, it continues with a different read when the previous contig can not be extended anymore. The Overlap Layout consensus finds every pairwise overlap, makes a layout graph with all the overlaps and chooses consensus sequences to make the contigs. The De Bruijn Graphs divides the reads in k-mers, makes a k-mer graph that shows all the overlapping k-mers, and chooses paths from the graph to make the contigs. " class="figure mx-auto d-block"></figure><hr></section><section id="05-binning"><h2 class="section-heading"><a href="05-binning.html">Metagenome Binning</a></h2>
<hr class="half-width">
<h3 id="05-binning-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Diagram depicts the DNA sequences in the original sample as circular chromosomes of three different taxa. After sequencing, the DNA sequences of the three different taxa are mixed as small linear reads; after the assembly, we have contigs, each corresponding to a single taxon, except for the ones with a bad assembly that has sequences of different taxa in the same contig, after the binning taxa separate the contigs.’</p>
<a href="fig/03-05-01.png">
<img src="fig/03-05-01.png" width="435" height="631" alt="Diagram depicts the DNA sequences in the original sample as circular chromosomes of three different taxa. After sequencing, the DNA sequences of the three different taxa are mixed as small linear reads; after the assembly, we have contigs, each corresponding to a single taxon, except for the ones with a bad assembly that has sequences of different taxa in the same contig, after the binning taxa separate the contigs." class="figure"></a><hr></section><section id="06-taxonomic"><h2 class="section-heading"><a href="06-taxonomic.html">Taxonomic Assignment</a></h2>
<hr class="half-width">
<h3 id="06-taxonomic-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Diagram of a taxonomic tree with four levels of nodes, some nodes have a number from 1 to 3, and some do not. From the most recent nodes, one has a three, and its parent nodes do not have numbers. This node with a three is selected.’</p>
<a href="fig/03-06-01.png">
<img src="fig/03-06-01.png" alt="Diagram of a taxonomic tree with four levels of nodes, some nodes have a number from 1 to 3, and some do not. From the most recent nodes, one has a three, and its parent nodes do not have numbers. This node with a three is selected." class="figure"></a><hr>
<h3 id="06-taxonomic-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Flow diagram that shows how the initial composition of 33% for each of the three taxa in the sample ends up being 4%, 72%, and 24% after the biases imposed by the extraction, PCR, sequencing and bioinformatics steps.’</p>
<a href="fig/03-06-02.png">
<img src="fig/03-06-02.png" alt="Flow diagram that shows how the initial composition of 33% for each of the three taxa in the sample ends up being 4%, 72%, and 24% after the biases imposed by the extraction, PCR, sequencing and bioinformatics steps." class="figure"></a><hr>
<h3 id="06-taxonomic-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Krona displays a circled-shape bacterial taxonomy plot with abundance percentages of each taxon’</p>
<a href="fig/03-06-03.png">
<img src="fig/03-06-03.png" alt="Krona displays a circled-shape bacterial taxonomy plot with abundance percentages of each taxon" class="figure"></a><hr>
<h3 id="06-taxonomic-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘Pavian website showing the upload of two reports’</p>
<a href="fig/03-06-04.png">
<img src="fig/03-06-04.png" alt="Pavian website showing the upload of two reports" class="figure"></a><hr>
<h3 id="06-taxonomic-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘Results Overview tab of the Pavian website where it shows the number of reads classified to several categories for the two samples’</p>
<a href="fig/03-06-05.png">
<img src="fig/03-06-05.png" alt="Results Overview tab of the Pavian website where it shows the number of reads classified to several categories for the two samples" class="figure"></a><hr>
<h3 id="06-taxonomic-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘Sankey type visualization that shows the abundance of each taxonomic label in a tree-like manner’</p>
<a href="fig/03-06-06.png">
<img src="fig/03-06-06.png" alt="Sankey type visualization that shows the abundance of each taxonomic label in a tree-like manner" class="figure"></a><hr>
<h3 id="06-taxonomic-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘A bar chart of the abundance of reads of the two samples, showing a segment for the read identified at the specific taxon and another segment for the number of reads identifies at children of the specified taxon’</p>
<a href="fig/03-06-07.png">
<img src="fig/03-06-07.png" alt="A bar chart of the abundance of reads of the two samples, showing a segment for the read identified at the specific taxon and another segment for the number of reads identifies at children of the specified taxon" class="figure"></a><hr>
<h3 id="06-taxonomic-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table of the same format as the Kraken report but for both samples at once.’</p>
<a href="fig/03-06-08.png">
<img src="fig/03-06-08.png" alt="A table of the same format as the Kraken report but for both samples at once." class="figure"></a><hr></section><section id="07-phyloseq"><h2 class="section-heading"><a href="07-phyloseq.html">Exploring Taxonomy with R</a></h2>
<hr class="half-width">
<h3 id="07-phyloseq-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table where the taxonomic
  identification information of all OTUs is displayed. Each row represents one
  OTU and the columns represent its identification at different levels in the taxonomic classification ranks, begging with Kingdom until we reach Species
  in the seventh column.’</p>
<a href="fig/03-07-01.png">
<img src="fig/03-07-01.png" alt="A table where the taxonomic
  identification information of all OTUs is displayed. Each row represents one
  OTU and the columns represent its identification at different levels in the taxonomic classification ranks, begging with Kingdom until we reach Species
  in the seventh column." class="figure"></a><hr>
<h3 id="07-phyloseq-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘The same table we saw in Figure 3 but with informative names in each of the columns. Now, we can see which of
  the columns are associated with which taxonomic classification rank’</p>
<a href="fig/03-07-02.png">
<img src="fig/03-07-02.png" alt="The same table we saw in Figure 3 but with informative names in each of the columns. Now, we can see which of
  the columns are associated with which taxonomic classification rank" class="figure"></a><hr>
<h3 id="07-phyloseq-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table where the abundance of each OTU in each sample is shown. Each row represents one
  OTU and the columns represent the samples. In the intersection, a number indicates how many sequenced reads of that OTU are present in that sample.’</p>
<a href="fig/03-07-03.png">
<img src="fig/03-07-03.png" alt="A table where the abundance of each OTU in each sample is shown. Each row represents one
  OTU and the columns represent the samples. In the intersection, a number indicates how many sequenced reads of that OTU are present in that sample." class="figure"></a><hr>
<h3 id="07-phyloseq-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘The OTU number is in the leftmost space of the table as a row name for the searched species.’</p>
<a href="%7B%7B%20page.root%20%7D%7D/fig/03-07-04.png">
<img src="%7B%7B%20page.root%20%7D%7D/fig/03-07-04.png" alt="The OTU number is in the leftmost space of the table as a row name for the searched species." class="figure"></a><hr></section><section id="08-Diversity-tackled-with-R"><h2 class="section-heading"><a href="08-Diversity-tackled-with-R.html">Diversity Tackled With R</a></h2>
<hr class="half-width">
<h3 id="08-Diversity-tackled-with-R-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Alpha diversity diagram: In lake A, we have three fishes, each one of a different species. On lake B, we have two fish, each of a different species. Moreover, we have four fish in lake C, each of different species.’</p>
<figure><img src="fig/03-08-01.png" alt="Alpha diversity diagram: In lake A, we have three fishes, each one of a different species. On lake B, we have two fish, each of a different species. Moreover, we have four fish in lake C, each of different species." class="figure mx-auto d-block"></figure><hr>
<h3 id="08-Diversity-tackled-with-R-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘ Alpha and Beta diversity diagram: Each Lake has a different number of species, and each species has a different number of fish individuals. Both metrics are taken into account to measure alfa and beta diversity.’</p>
<figure><img src="fig/03-08-02.png" alt=" Alpha and Beta diversity diagram: Each Lake has a different number of species, and each species has a different number of fish individuals. Both metrics are taken into account to measure alfa and beta diversity." class="figure mx-auto d-block"></figure><hr>
<h3 id="08-Diversity-tackled-with-R-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘In lake A, we have four different species, two of these species have three specimens each one. This Lake also has two specimens of a third species and only one specimen of a fourth specie. We got nine fish in total. Lake B has only three different species, the most populated species is also present in lake A and has five specimens, and we have only one specimen of each of the other two species. We got seven species total in lake B ’</p>
<a href="%7B%7B%20page.root%20%7D%7D/fig/03-08-03.png">
<img src="%7B%7B%20page.root%20%7D%7D/fig/03-08-03.png" alt="In lake A, we have four different species, two of these species have three specimens each one. This Lake also has two specimens of a third species and only one specimen of a fourth specie. We got nine fish in total. Lake B has only three different species, the most populated species is also present in lake A and has five specimens, and we have only one specimen of each of the other two species. We got seven species total in lake B " class="figure"></a><hr>
<h3 id="08-Diversity-tackled-with-R-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘A figure divided in three
panels. Each of these panels represents a different alpha diversity index.
Inside this section, each point represents the value assigned on this index to
the three different samples. The different indexes give
different values to the same sample.’</p>
<figure><img src="%7B%7B%20page.root%20%7D%7D/fig/03-08-04.png" alt="A figure divided in three
panels. Each of these panels represents a different alpha diversity index.
Inside this section, each point represents the value assigned on this index to
the three different samples. The different indexes give
different values to the same sample." class="figure mx-auto d-block"></figure><hr>
<h3 id="08-Diversity-tackled-with-R-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘[decorative]’</p>
<a href="fig/03-08-05.png"> <img src="fig/03-08-05.png" alt="" class="figure"></a><hr>
<h3 id="08-Diversity-tackled-with-R-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘[decorative]’</p>
<a href="fig/03-08-06.png"> <img src="fig/03-08-06.png" alt="" class="figure"></a><hr>
<h3 id="08-Diversity-tackled-with-R-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘The same panels as before, but now the samples are arranged horizontally according to the values in the Shannon index panel.’</p>
<a href="fig/03-08-07.png">
<img src="fig/03-08-07.png" alt="The same panels as before, but now the samples are arranged horizontally according to the values in the Shannon index panel." class="figure"></a><hr>
<h3 id="08-Diversity-tackled-with-R-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plot with NMDS1 as a label in x-axis that goes from -0.4 to 0.2 and NMDS2 in y-axis that goes from -0.2 to 0.1. The plot has three dots that are not clustered in any way.’</p>
<a href="fig/03-08-08.png">
<img src="fig/03-08-08.png" alt="Plot with NMDS1 as a label in x-axis that goes from -0.4 to 0.2 and NMDS2 in y-axis that goes from -0.2 to 0.1. The plot has three dots that are not clustered in any way." class="figure"></a><hr>
<h3 id="08-Diversity-tackled-with-R-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘The distance between the three samples, JC1A, JP4D and JP41 is shown in a plane. Each sample has a legend and a color. The color is according to the metadata treatment. There are three possible treatments in the legend: Control mesocosm, Fertilized pond, and Unenriched pond’</p>
<a href="fig/03-08-09.png">
<img src="fig/03-08-09.png" alt="The distance between the three samples, JC1A, JP4D and JP41 is shown in a plane. Each sample has a legend and a color. The color is according to the metadata treatment. There are three possible treatments in the legend: Control mesocosm, Fertilized pond, and Unenriched pond" class="figure"></a><hr></section><section id="09-abundance-analyses"><h2 class="section-heading"><a href="09-abundance-analyses.html">Taxonomic Analysis with R</a></h2>
<hr class="half-width">
<h3 id="09-abundance-analyses-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Table containing the
  taxonomic information of each of the OTUs inside the three samples. Here,
  we can see only the Phylum column has information, leaving the other
  taxonomic levels blank.’</p>
<a href="fig/03-09-01.png">
<img src="fig/03-09-01.png" alt="Table containing the
  taxonomic information of each of the OTUs inside the three samples. Here,
  we can see only the Phylum column has information, leaving the other
  taxonomic levels blank." class="figure"></a><hr>
<h3 id="09-abundance-analyses-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘A two-part plot contrasting
  the absolute versus the relative abundance of the three samples. On the right
  side, we can see how each of the bars has its own height, making it difficult
  to compare the information between samples. The right side shows
  three bars with the same height after the abundance was transformed to
  percentage inside each sample.’</p>
<a href="fig/03-09-02.png">
<img src="fig/03-09-02.png" alt="A two-part plot contrasting
  the absolute versus the relative abundance of the three samples. On the right
  side, we can see how each of the bars has its own height, making it difficult
  to compare the information between samples. The right side shows
  three bars with the same height after the abundance was transformed to
  percentage inside each sample." class="figure"></a><hr>
<h3 id="09-abundance-analyses-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘A new two-part plot with
  a reassignment of the low-abundant taxa on the right side. Compared to the
  left legend, the one in the right has fewer groups because of the process of
  reassigning the taxa with an abundance lower than 0.5 % to just one
  group/color.’</p>
<a href="fig/03-09-03.png">
<img src="fig/03-09-03.png" alt="A new two-part plot with
  a reassignment of the low-abundant taxa on the right side. Compared to the
  left legend, the one in the right has fewer groups because of the process of
  reassigning the taxa with an abundance lower than 0.5 % to just one
  group/color." class="figure"></a><hr>
<h3 id="09-abundance-analyses-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘A new plot with three bars
  representing the absolute abundance of Cyanobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height.’</p>
<a href="fig/03-09-05.png">
<img src="fig/03-09-05.png" alt="A new plot with three bars
  representing the absolute abundance of Cyanobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height." class="figure"></a><hr>
<h3 id="09-abundance-analyses-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘New reassignment to the low abundant taxa on the left part of the plot. A new class has been created that contains the taxa with less than 300 reads’</p>
<a href="fig/03-09-04.png">
<img src="fig/03-09-04.png" alt="New reassignment to the low abundant taxa on the left part of the plot. A new class has been created that contains the taxa with less than 300 reads" class="figure"></a><hr>
<h3 id="09-abundance-analyses-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘A new plot with three bars
  representing the absolute abundance of Proteobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height.’</p>
<a href="%7B%7B%20page.root%20%7D%7D/fig/03-09-06.png">
<img src="%7B%7B%20page.root%20%7D%7D/fig/03-09-06.png" alt="A new plot with three bars
  representing the absolute abundance of Proteobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height." class="figure"></a><hr></section><section id="10-OtherResources"><h2 class="section-heading"><a href="10-OtherResources.html">Other Resources</a></h2>
<hr class="half-width">
<h3 id="10-OtherResources-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Pie chart showing the relative abundance at the phylum level, and the legend with the phylum names, read count, and percentages.’</p>
<a href="../fig/03-11-02.png">
<img src="../fig/03-11-02.png" alt="Pie chart showing the relative abundance at the phylum level, and the legend with the phylum names, read count, and percentages." class="figure"></a><hr>
<h3 id="10-OtherResources-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Pie chart showing the relative abundance of general functional categories, and the legend with the category names, read count, and percentages.’</p>
<a href="fig/03-11-04.png">
<img src="fig/03-11-04.png" alt="Pie chart showing the relative abundance of general functional categories, and the legend with the category names, read count, and percentages." class="figure"></a><hr>
<h3 id="10-OtherResources-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Pie chart showing the relative abundance of specific functional categories, and the legend with the category names, read count, and percentages.’</p>
<a href="fig/03-11-05.png">
<img src="fig/03-11-05.png" alt="Pie chart showing the relative abundance of specific functional categories, and the legend with the category names, read count, and percentages." class="figure"></a><hr></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/images.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All Images",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/images.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/images.html",
  "dateCreated": "2020-08-24",
  "dateModified": "2025-04-29",
  "datePublished": "2025-04-29"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

