<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Processing and Visualization for Metagenomics: Assessing Read Quality</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/lab/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/lab/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/lab/favicon-16x16.png"><link rel="manifest" href="../favicons/lab/site.webmanifest"><link rel="mask-icon" href="../favicons/lab/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav lab"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="lab-logo" alt="Lesson Description" src="../assets/images/lab-logo.svg"><abbr class="badge peer-reviewed" title="This lesson has passed peer review.">
          <a href="https://doi.org/10.21105/jose.00209" class="external-link alert-link">
            <i aria-hidden="true" class="icon" data-feather="user-check" style="border-radius: 5px"></i>
            DOI: 10.21105/jose.00209
          </a>
        </abbr>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../02-assessing-read-quality.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav lab" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/lab-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="BinningRefinmentDereplication.html">Binning, Refinamiento y Desreplicación</a></li><li><a class="dropdown-item" href="ImprovingMAG.html">Reconstrucción de Genomas a partir de Metagenomas (MAGs)</a></li><li><a class="dropdown-item" href="MAGsTaxonomyMetabolism.html">Clasificación taxonómica y Metabolismo</a></li><li><a class="dropdown-item" href="amplicon.html">Amplicon metagenomics</a></li><li><a class="dropdown-item" href="data.html">Data</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 7%" class="percentage">
    7%
  </div>
  <div class="progress lab">
    <div class="progress-bar lab" role="progressbar" style="width: 7%" aria-valuenow="7" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../02-assessing-read-quality.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Assessing Read Quality
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#bioinformatic-workflows">Bioinformatic workflows</a></li>
<li><a href="#quality-control">Quality control</a></li>
<li><a href="#activating-an-environment">Activating an environment</a></li>
<li><a href="#assessing-quality-using-fastqc">Assessing quality using FastQC</a></li>
<li><a href="#running-fastqc">Running FastQC</a></li>
<li><a href="#viewing-the-fastqc-results">Viewing the FastQC results</a></li>
<li><a href="#decoding-the-other-fastqc-outputs">Decoding the other FastQC outputs</a></li>
<li><a href="#working-with-the-fastqc-text-output">Working with the FastQC text output</a></li>
<li><a href="#documenting-our-work">Documenting our work</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-taxonomic.html">6. Taxonomic Assignment</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-phyloseq.html">7. Exploring Taxonomy with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-Diversity-tackled-with-R.html">8. Diversity Tackled With R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-abundance-analyses.html">9. Taxonomic Analysis with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="BinningRefinmentDereplication.html">Binning, Refinamiento y Desreplicación</a></li><li><a href="ImprovingMAG.html">Reconstrucción de Genomas a partir de Metagenomas (MAGs)</a></li><li><a href="MAGsTaxonomyMetabolism.html">Clasificación taxonómica y Metabolismo</a></li><li><a href="amplicon.html">Amplicon metagenomics</a></li><li><a href="data.html">Data</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-background-metadata.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-trimming-filtering.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-background-metadata.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Starting a
        </a>
        <a class="chapter-link float-end" href="../instructor/03-trimming-filtering.html" rel="next">
          Next: Trimming and...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Assessing Read Quality</h1>
        <p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/02-assessing-read-quality.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I describe the quality of my data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain how a FASTQ file encodes per-base quality scores.</li>
<li>Interpret a FastQC plot summarizing per-base quality across all
reads.</li>
<li>Use <code>for</code> loops to automate operations on multiple
files.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="bioinformatic-workflows">Bioinformatic workflows<a class="anchor" aria-label="anchor" href="#bioinformatic-workflows"></a></h2>
<hr class="half-width"><p>When working with high-throughput sequencing data, the raw reads you
get off the sequencer must pass through several different tools to
generate your final desired output. The execution of this set of tools
in a specified order is commonly referred to as a <em>workflow</em> or a
<em>pipeline</em>.</p>
<p>An example of the workflow we will be using for our analysis is
provided below, with a brief description of each step.</p>
<p><a href="../fig/03-02-01.png">
<img src="../fig/03-02-01.png" alt="Flow diagram that shows the steps: Sequence reads, Quality control, Assembly, Binning and Taxonomy" class="figure"></a></p>
<ol style="list-style-type: decimal"><li>Quality control - Assessing quality using FastQC and Trimming and/or
filtering reads (if necessary)</li>
<li>Assembly of metagenome</li>
<li>Binning</li>
<li>Taxonomic assignation</li>
</ol><p>These workflows in bioinformatics adopt a plug-and-play approach in
that the output of one tool can be easily used as input to another tool
without any extensive configuration. Having standards for data formats
is what makes this feasible. Standards ensure that data is stored in a
way that is generally accepted and agreed upon within the community.
Therefore, the tools used to analyze data at different workflow stages
are<br>
built, assuming that the data will be provided in a specific format.</p>
</section><section><h2 class="section-heading" id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a></h2>
<hr class="half-width"><p>We will now assess the quality of the sequence reads contained in our
FASTQ files.</p>
<p><a href="../fig/03-02-02.png">
<img src="../fig/03-02-02.png" alt="Flow diagram that shows the steps: Sequence reads and Quality control." class="figure"></a></p>
<div class="section level3">
<h3 id="details-on-the-fastq-format">Details on the FASTQ format<a class="anchor" aria-label="anchor" href="#details-on-the-fastq-format"></a></h3>
<p>Although it looks complicated (and it is), we can understand the <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">FASTQ</a> format with
a little decoding. Some rules about the format include the
following:</p>
<table class="table"><colgroup><col width="3%"><col width="96%"></colgroup><thead><tr class="header"><th>Line</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>1</td>
<td>Always begins with ‘@’ followed by the information about the
read</td>
</tr><tr class="even"><td>2</td>
<td>The actual DNA sequence</td>
</tr><tr class="odd"><td>3</td>
<td>Always begins with a ‘+’ and sometimes contains the same info as in
line 1</td>
</tr><tr class="even"><td>4</td>
<td>Has a string of characters which represent the quality scores; must
have same number of characters as line 2</td>
</tr></tbody></table><p>We can view the first complete read in one of the files from our
dataset using <code>head</code> to look at the first four lines. But we
have to decompress one of the files first.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq/</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">$</span> gunzip JP4D_R1.fastq.gz</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n</span> 4 JP4D_R1.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:12622:2006 1:N:0:CTCAGA
CCCGTTCCTCGGGCGTGCAGTCGGGCTTGCGGTCTGCCATGTCGTGTTCGGCGTCGGTGGTGCCGATCAGGGTGAAATCCGTCTCGTAGGGGATCGCGAAGATGATCCGCCCGTCCGTGCCCTGAAAGAAATAGCACTTGTCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCTCAGAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAAGCAAACCTCTCACTCCCTCTACTCTACTCCCTT
+
A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### </code></pre>
</div>
<p>Line 4 shows the quality of each nucleotide in the read. Quality is
interpreted as the probability of an incorrect base call (e.g., 1 in 10)
or, equivalently, the base call accuracy (e.g., 90%). Each nucleotide’s
numerical score’s value is converted into a character code where every
single character represents a quality score for an individual
nucleotide. This conversion allows the alignment of each individual
nucleotide with its quality score. For example, in the line above, the
quality score line is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### </code></pre>
</div>
<p>The numerical value assigned to each character depends on the
sequencing platform that generated the reads. The sequencing machine
used to generate our data uses the standard Sanger quality PHRED score
encoding, using Illumina version 1.8 onwards. Each character is assigned
a quality score between 0 and 41, as shown in the chart below.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Quality encoding: !"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJ
                   |         |         |         |         |
Quality score:    01........11........21........31........41                                </code></pre>
</div>
<p>Each quality score represents the probability that the corresponding
nucleotide call is incorrect. These probability values are the results
of the base calling algorithm and depend on how much signal was captured
for the base incorporation. This quality score is logarithmically based,
so a quality score of 10 reflects a base call accuracy of 90%, but a
quality score of 20 reflects a base call accuracy of 99%. In this <a href="https://drive5.com/usearch/manual/quality_score.html" class="external-link">link</a> you
can find more information about quality scores.</p>
<p>Looking back at our read:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:12622:2006 1:N:0:CTCAGA
CCCGTTCCTCGGGCGTGCAGTCGGGCTTGCGGTCTGCCATGTCGTGTTCGGCGTCGGTGGTGCCGATCAGGGTGAAATCCGTCTCGTAGGGGATCGCGAAGATGATCCGCCCGTCCGTGCCCTGAAAGAAATAGCACTTGTCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCTCAGAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAAGCAAACCTCTCACTCCCTCTACTCTACTCCCTT
+
A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### </code></pre>
</div>
<p>We can now see that there is a range of quality scores but that the
end of the sequence is very poor (<code>#</code> = a quality score of
2).</p>
<div id="exercise-1-looking-at-specific-reads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-looking-at-specific-reads" class="callout-inner">
<h3 class="callout-title">Exercise 1: Looking at specific reads</h3>
<div class="callout-content">
<p>In the terminal, how would you show the ID and quality of the last
read <code>JP4D_R1.fastq</code>?<br>
a) <code>tail JP4D_R1.fastq</code><br>
b) <code>head -n 4 JP4D_R1.fastq</code><br>
c) <code>more JP4D_R1.fastq</code><br>
d) <code>tail -n4 JP4D_R1.fastq</code><br>
e) <code>tail -n4 JP4D_R1.fastq | head -n2</code></p>
<p>Do you trust the sequence in this read?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>  </span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">a</span><span class="er">)</span> <span class="ex">It</span> shows the ID and quality of the last read but also unnecessary lines from previous reads.  </span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="ex">b</span><span class="er">)</span> <span class="ex">No.</span> It shows the first read<span class="st">'s info.  </span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="st">c) It shows the text of the entire file.  </span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="st">d) This option is the best answer as it only shows the last read'</span>s information.  </span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="ex">e</span><span class="er">)</span> <span class="ex">It</span> does show the ID of the last read but not the quality.  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@MISEQ-LAB244-W7:156:000000000-A80CV:1:2114:17866:28868 1:N:0:CTCAGA

CCCGTTCTCCACCTCGGCGCGCGCCAGCTGCGGCTCGTCCTTCCACAGGAACTTCCACGTCGCCGTCAGCCGCGACACGTTCTCCCCCCTCGCATGCTCGTCCTGTCTCTCGTGCTTGGCCGACGCCTGCGCCTCGCACTGCGCCCGCTCGGTGTCGTTCATGTTGATCTTCACCGTGGCGTGCATGAAGCGGTTCCCGGCCTCGTCGCCACCCACGCCATCCGCGTCGGCCAGCCACTCTCACTGCTCGC

+

AA11AC1&gt;3@DC1F1111000A0/A///BB#############################################################################################################################################################################################################################          </code></pre>
</div>
<p>This read has more consistent quality at its first than at the end
but still has a range of quality scores, most of them are low. We will
look at variations in position-based quality in just a moment.</p>
</div>
</div>
</div>
</div>
<p>In real life, you won’t be assessing the quality of your reads by
visually inspecting your FASTQ files. Instead, you’ll use a software
program to assess read quality and filter out poor reads. We’ll first
use a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
to visualize the quality of our reads. Later in our workflow, we’ll use
another program to filter out poor-quality reads.</p>
<p>First, let’s make available our metagenomics software:</p>
</div>
</section><section><h2 class="section-heading" id="activating-an-environment">Activating an environment<a class="anchor" aria-label="anchor" href="#activating-an-environment"></a></h2>
<hr class="half-width"><p>Environments are part of a bioinformatic tendency to do reproducible
research; they are a way to share and maintain our programs in their
needed versions used for a pipeline with our colleagues and our future
self. FastQC has not been activated in the (base) environment, but this
AWS instance came with an environment called metagenomics. We need to
activate it in order to start using FastQC.</p>
<p>We will use <a href="https://docs.conda.io/en/latest/" class="external-link">Conda</a> as
our environment manager. Conda is an open-source package and environment
management system that runs on Windows, macOS and Linux. Conda
environments are activated with the <code>conda activate</code>
direction:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> conda activate metagenomics  </span></code></pre>
</div>
<p>After the environment has been activated, a label is shown before the
<code>$</code> sign.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>(metagenomics) $</code></pre>
</div>
<p>Now, if we call FastQC, a long help page will be displayed on our
screen.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="at">-h</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

        fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam]
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of many different modules, each one of
    which will help to identify a different potential type of problem in your
    data.
.
.
.</code></pre>
</div>
<p>If FastQC is not installed, then you would expect to see an error
like</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>The program 'fastqc' is currently not installed. You can install it by typing:
sudo apt-get install fastqc</code></pre>
</div>
<p>If this happens, check with your instructor before trying to install
it.</p>
</section><section><h2 class="section-heading" id="assessing-quality-using-fastqc">Assessing quality using FastQC<a class="anchor" aria-label="anchor" href="#assessing-quality-using-fastqc"></a></h2>
<hr class="half-width"><p>FastQC has several features that can give you a quick impression of
any problems your data may have, so you can consider these issues before
moving forward with your analyses. Rather than looking at quality scores
for each read, FastQC looks at quality collectively across all reads
within a sample. The image below shows one FastQC-generated plot that
indicates a very high-quality sample:</p>
<p><a href="../fig/03-02-03.png">
<img src="../fig/03-02-03.png" alt="Quality graph showing a boxplot for each sequence position in the x-axis, and the Phred score in the y-axis. The background is colored red for the Phred scores 0 to 20, yellow for the scores 20 to 28, and green for the scores 28 to 38. All of the boxes for each position are in the green area." class="figure"></a></p>
<p>The x-axis displays the base position in the read, and the y-axis
shows quality scores. In this example, the sample contains reads that
are 40 bp long. This length is much shorter than the reads we are
working on within our workflow. For each position, there is a
box-and-whisker plot showing the distribution of quality scores for all
reads at that position. The horizontal red line indicates the median
quality score, and the yellow box shows the 1st to 3rd quartile range.
This range means that 50% of reads have a quality score that falls
within the range of the yellow box at that position. The whiskers show
the whole range covering the lowest (0th quartile) to highest (4th
quartile) values.</p>
<p>The quality values for each position in this sample do not drop much
lower than 32, which is a high-quality score. The plot background is
also color-coded to identify good (green), acceptable (yellow) and bad
(red) quality scores.</p>
<p>Now let’s look at a quality plot on the other end of the
spectrum.</p>
<p><a href="../fig/03-02-04.png">
<img src="../fig/03-02-04.png" alt="Graphic of boxplots, where the first ones are in the good range of scores of the y-axis and extend to the acceptable and bad ranges of scores toward the right of the x-axis" class="figure"></a></p>
<p>The FastQC tool produces several other diagnostic plots to assess
sample quality and the one plotted above. Here, we see positions within
the read in which the boxes span a much more comprehensive range. Also,
quality scores drop pretty low into the “bad” range, particularly on the
tail end of the reads.</p>
</section><section><h2 class="section-heading" id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a></h2>
<hr class="half-width"><p>We will now assess the quality of the reads that we downloaded.
First, make sure you’re still in the <code>untrimmed_fastq</code>
directory.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq/ </span></code></pre>
</div>
<div id="exercise-2-looking-at-metadata-about-the-untrimmed-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-looking-at-metadata-about-the-untrimmed-files" class="callout-inner">
<h3 class="callout-title">Exercise 2: Looking at metadata about the untrimmed-files</h3>
<div class="callout-content">
<p>To know which files have more data, you need to see metadata about
untrimmed files. In files, metadata includes owners of the file, state
of the write, read, and execute permissions, size, and modification
date. Using the <code>ls</code> command, how would you get the size of
the files in the <code>untrimmed_fastq\</code> directory?<br>
(Hint: Look at the options for the <code>ls</code> command to see how to
show file sizes.)<br>
a) <code>ls -a</code><br>
b) <code>ls -S</code><br>
c) <code>ls -l</code><br>
d) <code>ls -lh</code><br>
e) <code>ls -ahlS</code></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<ol style="list-style-type: lower-alpha"><li>No. The flag <code>-a</code> shows all the contents, including
hidden files and directories, but not the sizes.<br></li>
<li>No. The flag <code>-S</code> shows the content Sorted by size,
starting with the most extensive file, but not the sizes.<br></li>
<li>Yes. The flag <code>-l</code> shows the contents with metadata,
including file size. Other metadata are permissions, owners, and
modification dates.<br></li>
<li>Yes. The flag <code>-lh</code> shows the content with metadata in a
human-readable manner.<br></li>
<li>Yes. The combination of all the flags shows all the contents with
metadata, including hidden files, sorted by size.</li>
</ol><div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-ahls</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-rw-r--r-- 1 dcuser dcuser  24M Nov 26 21:34 JC1A_R1.fastq.gz
-rw-r--r-- 1 dcuser dcuser  24M Nov 26 21:34 JC1A_R2.fastq.gz
-rw-r--r-- 1 dcuser dcuser 616M Nov 26 21:34 JP4D_R1.fastq
-rw-r--r-- 1 dcuser dcuser 203M Nov 26 21:35 JP4D_R2.fastq.gz   </code></pre>
</div>
<p>Four FASTQ files oscillate between 24M (24MB) to 616M. The largest
file is JP4D_R1.fastq with 616M.</p>
</div>
</div>
</div>
</div>
<p>FastQC can accept multiple file names as input, and on both zipped
and unzipped files, so we can use the <code>\*.fastq*</code>wildcard to
run FastQC on all FASTQ files in this directory.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="pp">*</span>.fastq<span class="pp">*</span> </span></code></pre>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Started analysis of JC1A_R1.fastq.gz
Approx 5% complete for JC1A_R1.fastq.gz
Approx 10% complete for JC1A_R1.fastq.gz
Approx 15% complete for JC1A_R1.fastq.gz
Approx 20% complete for JC1A_R1.fastq.gz
Approx 25% complete for JC1A_R1.fastq.gz
Approx 30% complete for JC1A_R1.fastq.gz
Approx 35% complete for JC1A_R1.fastq.gz  </code></pre>
</div>
<p>It should take around five minutes for FastQC to run on all four of
our FASTQ files. When the analysis completes, your prompt will return.
So your screen will look something like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Approx 80% complete for JP4D_R2.fastq.gz
Approx 85% complete for JP4D_R2.fastq.gz
Approx 90% complete for JP4D_R2.fastq.gz
Approx 95% complete for JP4D_R2.fastq.gz
Analysis complete for JP4D_R2.fastq.gz
$</code></pre>
</div>
<p>The FastQC program has created several new files within our
<code>data/untrimmed_fastq/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">$</span> ls </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc.html  JC1A_R2_fastqc.html  JP4D_R1.fastq        JP4D_R2_fastqc.html  TruSeq3-PE.fa
JC1A_R1_fastqc.zip   JC1A_R2_fastqc.zip   JP4D_R1_fastqc.html  JP4D_R2_fastqc.zip
JC1A_R1.fastq.gz     JC1A_R2.fastq.gz     JP4D_R1_fastqc.zip   JP4D_R2.fastq.gz    </code></pre>
</div>
<p>For each input FASTQ file, FastQC has created a <code>.zip</code>
file and a <code>.html</code> file. The <code>.zip</code> file extension
indicates that this is a compressed set of multiple output files. We’ll
be working with these output files soon. The <code>.html</code> file is
a stable webpage displaying the summary report for each of our
samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into a new directory within our
<code>results/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/dc_workshop/results/fastqc_untrimmed_reads </span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.zip ~/dc_workshop/results/fastqc_untrimmed_reads/ </span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.html ~/dc_workshop/results/fastqc_untrimmed_reads/ </span></code></pre>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/results/fastqc_untrimmed_reads/ </span></code></pre>
</div>
</section><section><h2 class="section-heading" id="viewing-the-fastqc-results">Viewing the FastQC results<a class="anchor" aria-label="anchor" href="#viewing-the-fastqc-results"></a></h2>
<hr class="half-width"><p>If we were working on our local computers, we’d be able to look at
each of these HTML files by opening them in a web browser. However,
these files are currently sitting on our remote AWS instance, where our
local computer can’t see them. If we are working with the terminal
provided by r-studio we can either: select the html files and with the
secondary click chose the option open in a browser or export the files
to our local computer as we learned in the Introduction to the Command
Line lesson.</p>
<div id="exercise-3-downloading-files-by-scp-optional" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3-downloading-files-by-scp-optional" class="callout-inner">
<h3 class="callout-title">Exercise 3: Downloading files by scp (optional)</h3>
<div class="callout-content">
<p>Suppose you only have access to a terminal and there is not a web
browser available at the AWS remote machine. The following are the steps
needed to download your files to your computer. Observe the third step
and answer false/true for each question about the <code>scp</code>
command.</p>
<ol style="list-style-type: decimal"><li>Open a new terminal on your local computer.</li>
<li>Make a new local directory on our computer to store the HTML files
<code>$ mkdir -p ~/Desktop/fastqc_html</code>
</li>
<li>Transfer our HTML files to our local computer using
<code>scp</code>.
<code>$ scp dcuser@ec2-34-238-162-94.compute-1.amazonaws.com:~/dc_workshop/results/fastqc_untrimmed_reads/*.html ~/Desktop/fastqc_html</code><br></li>
</ol><ol style="list-style-type: upper-alpha"><li>
<code>dcuser</code> is your local user<br></li>
<li>
<code>ec2-34-238-162-94.compute-1.amazonaws.com</code> is the
address of your remote machine<br></li>
<li>the current adress of the file goes after the second space in the
<code>scp</code> command.<br></li>
<li>
<code>~/dc_workshop/results/fastqc_untrimmed_reads/*.html</code> is
the path of the file you want to download<br></li>
<li>
<code>~/Desktop/fastqc_html</code> is a remote path.<br></li>
<li>
<code>:</code> Divides the host name of your local computer and the
path of the file.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha"><li>False. <code>dcuser</code> is your remote user.<br></li>
<li>True. <code>ec2-34-238-162-94.compute-1.amazonaws.com</code> is the
adress of your remote machine<br></li>
<li>False. The current adress of the file goes after the first space in
the <code>scp</code> command.<br></li>
<li>True.
<code>~/dc_workshop/results/fastqc_untrimmed_reads/*.html</code> is the
path of the file you want to download in the remote machine.</li>
<li>False. <code>~/Desktop/fastqc_html</code> is a local path where your
file will be downloaded.</li>
<li>False. <code>:</code> Divides the host name of a <em>remote</em>
computer and the path of the file on the remote computer.</li>
</ol><p>You should see a status output like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc.html     100%  253KB 320.0KB/s   00:00
JC1A_R2_fastqc.html     100%  262KB 390.1KB/s   00:00
JP4D_R1_fastqc.html     100%  237KB 360.8KB/s   00:00
JP4D_R2_fastqc.html     100%  244KB 385.2KB/s   00:00</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Now we can open the 4 HTML files.</p>
<p>Depending on your system, you should be able to select and open them
all at once via a right-click menu in your file browser.</p>
<div id="exercise-4-discuss-the-quality-of-sequencing-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4-discuss-the-quality-of-sequencing-files" class="callout-inner">
<h3 class="callout-title">Exercise 4: Discuss the quality of sequencing files</h3>
<div class="callout-content">
<p>Discuss your results with a neighbor. Which sample(s) looks the best
per base sequence quality? Which sample(s) look the worst?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>All of the reads contain usable data, but the quality decreases
toward the end of the reads. File JC1A_R2_fastqc shows the lowest
quality.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="decoding-the-other-fastqc-outputs">Decoding the other FastQC outputs<a class="anchor" aria-label="anchor" href="#decoding-the-other-fastqc-outputs"></a></h2>
<hr class="half-width"><p>We’ve now looked at quite a few “Per base sequence quality” FastQC
graphs, but there are nine other graphs that we haven’t talked about!
Below we have provided a brief overview of interpretations for each
plot. For more information, please see the FastQC documentation <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/" class="external-link">here</a></p>
<ul><li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html" class="external-link"><strong>Per
tile sequence quality</strong></a>: the machines that perform sequencing
are divided into tiles. This plot displays patterns in base quality
along these tiles. Consistently low scores are often found around the
edges, but hot spots could also occur in the middle if an air bubble was
introduced during the run.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/3%20Per%20Sequence%20Quality%20Scores.html" class="external-link"><strong>Per
sequence quality scores</strong></a>: a density plot of quality for all
reads at all positions. This plot shows what quality scores are most
common.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html" class="external-link"><strong>Per
base sequence content</strong></a>: plots the proportion of each base
position over all of the reads. Typically, we expect to see each base
roughly 25% of the time at each position, but this often fails at the
beginning or end of the read due to quality or adapter content.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/5%20Per%20Sequence%20GC%20Content.html" class="external-link"><strong>Per
sequence GC content</strong></a>: a density plot of average GC content
in each of the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/6%20Per%20Base%20N%20Content.html" class="external-link"><strong>Per
base N content</strong></a>: the percent of times that ‘N’ occurs at a
position in all reads. If there is an increase at a particular position,
this might indicate that something went wrong during sequencing.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html" class="external-link"><strong>Sequence
Length Distribution</strong></a>: the distribution of sequence lengths
of all reads in the file. If the data is raw, there is often a sharp
peak; however, if the reads have been trimmed, there may be a
distribution of shorter lengths.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/8%20Duplicate%20Sequences.html" class="external-link"><strong>Sequence
Duplication Levels</strong></a>: a distribution of duplicated sequences.
In sequencing, we expect most reads to only occur once. If some
sequences are occurring more than once, it might indicate enrichment
bias (e.g. from PCR). This might not be true if the samples are high
coverage (or RNA-seq or amplicon).</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/9%20Overrepresented%20Sequences.html" class="external-link"><strong>Overrepresented
sequences</strong></a>: a list of sequences that occur more frequently
than would be expected by chance.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/10%20Adapter%20Content.html" class="external-link"><strong>Adapter
Content</strong></a>: a graph indicating where adapter sequences occur
in the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/11%20Kmer%20Content.html" class="external-link"><strong>K-mer
Content</strong></a>: a graph showing any sequences which may show a
positional bias within the reads.</li>
</ul></section><section><h2 class="section-heading" id="working-with-the-fastqc-text-output">Working with the FastQC text output<a class="anchor" aria-label="anchor" href="#working-with-the-fastqc-text-output"></a></h2>
<hr class="half-width"><p>Now that we’ve looked at our HTML reports getting a feel for the
data, let’s look more closely at the other output files. Go back to the
tab in your terminal program that is connected to your AWS instance (the
tab label will start with <code>dcuser@ip</code>) and make sure you’re
in our results subdirectory.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/results/fastqc_untrimmed_reads/ </span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="ex">$</span> ls </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc.html           JP4D_R1_fastqc.html
JC1A_R1_fastqc.zip            JP4D_R1_fastqc.zip
JC1A_R2_fastqc.html           JP4D_R2_fastqc.html
JC1A_R2_fastqc.zip            JP4D_R2_fastqc.zip </code></pre>
</div>
<p>Our <code>.zip</code> files are compressed files. Each contains
multiple different types of output files for a single input FASTQ file.
To view the contents of a <code>.zip</code> file, we can use the program
<code>unzip</code> to decompress these files. Let’s try doing them all
at once using a wildcard.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">$</span> unzip <span class="pp">*</span>.zip </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Archive:  JC1A_R1_fastqc.zip
caution: filename not matched:  JC1A_R2_fastqc.zip
caution: filename not matched:  JP4D_R1_fastqc.zip
caution: filename not matched:  JP4D_R2_fastqc.zip  </code></pre>
</div>
<p>This decompression didn’t work. It identified the first file and got
a warning message for the other <code>.zip</code> files. This is because
<code>unzip</code> expects to get only one zip file as input. We could
go through and unzip each file one at a time, but this is very
time-consuming and error-prone. Someday you may have 500 files to
unzip!</p>
<p>A more efficient way is to use a <code>for</code> loop like we
learned in the Command Line lesson to iterate through all of our
<code>.zip</code> files. Let’s see what that looks like, and then we’ll
discuss what we’re doing with each line of our loop.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="ex">$</span> for filename in <span class="pp">*</span>.zip</span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="op">&gt;</span> unzip <span class="va">$filename</span></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>In this example, the input is the four filenames (one filename for
each of our <code>.zip</code> files). Each time the loop iterates, it
will assign a file name to the variable <code>filename</code> and run
the <code>unzip</code> command. The first time through the loop,
<code>$filename</code> is <code>JC1A_R1_fastqc.zip</code>. The
interpreter runs the command <code>unzip</code> on
<code>JC1A_R1_fastqc.zip</code>. For the second iteration,
<code>$filename</code> becomes <code>JC1A_R2_fastqc.zip</code>. This
time, the shell runs <code>unzip</code> on
<code>JC1A_R2_fastqc.zip</code>. It then repeats this process for the
other <code>.zip</code> files in our directory.</p>
<p>When we run the <code>for</code> loop, you will see an output that
starts like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">Archive</span><span class="op">:</span>  <span class="va">JC1A_R1_fastqc.zip</span></span>
<span><span class="va">creating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span></span>
<span><span class="va">creating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span></span>
<span><span class="va">creating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">fastqc_icon.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">warning.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">error.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">tick.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">summary.txt</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_quality.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_tile_quality.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_quality.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_sequence_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_gc_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_n_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">sequence_length_distribution.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">duplication_levels.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">adapter_content.png</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">fastqc_report.html</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">fastqc_data.txt</span></span>
<span><span class="va">inflating</span><span class="op">:</span> <span class="va">JC1A_R1_fastqc</span><span class="op">/</span><span class="va">fastqc.fo</span>  </span></code></pre>
</div>
<p>The <code>unzip</code> program is decompressing the <code>.zip</code>
files and creates a new directory (with subdirectories) for each of our
samples, to store all of the different output that is produced by
FastQC. There are a lot of files here. We’re going to focus on the
<code>summary.txt</code> file.</p>
<p>If you list the files in our directory, now you will see the
following:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">$</span> ls </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc                  JP4D_R1_fastqc
JC1A_R1_fastqc.html             JP4D_R1_fastqc.html
JC1A_R1_fastqc.zip              JP4D_R1_fastqc.zip
JC1A_R2_fastqc                  JP4D_R2_fastqc
JC1A_R2_fastqc.html             JP4D_R2_fastqc.html
JC1A_R2_fastqc.zip              JP4D_R2_fastqc.zip                                         </code></pre>
</div>
<p>The <code>.html</code> files and the uncompressed <code>.zip</code>
files are still present, but now we also have a new directory for each
sample. We can see that it’s a directory if we use the <code>-F</code>
flag for <code>ls</code>.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>JC1A_R1_fastqc/                  JP4D_R1_fastqc/
JC1A_R1_fastqc.html             JP4D_R1_fastqc.html
JC1A_R1_fastqc.zip              JP4D_R1_fastqc.zip
JC1A_R2_fastqc/                  JP4D_R2_fastqc/
JC1A_R2_fastqc.html             JP4D_R2_fastqc.html
JC1A_R2_fastqc.zip              JP4D_R2_fastqc.zip                                         </code></pre>
</div>
<p>Let’s see what files are present within one of these output
directories.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span> JC1A_R1_fastqc/ </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fastqc_data.txt  fastqc.fo  fastqc_report.html	Icons/	Images/  summary.txt</code></pre>
</div>
<p>Use <code>less</code> to preview the <code>summary.txt</code> file
for this sample.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="ex">$</span> less JC1A_R1_fastqc/summary.txt </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>PASS    Basic Statistics        JC1A_R1.fastq.gz
FAIL    Per base sequence quality       JC1A_R1.fastq.gz
PASS    Per tile sequence quality       JC1A_R1.fastq.gz
PASS    Per sequence quality scores     JC1A_R1.fastq.gz
WARN    Per base sequence content       JC1A_R1.fastq.gz
FAIL    Per sequence GC content JC1A_R1.fastq.gz
PASS    Per base N content      JC1A_R1.fastq.gz
PASS    Sequence Length Distribution    JC1A_R1.fastq.gz
FAIL    Sequence Duplication Levels     JC1A_R1.fastq.gz
PASS    Overrepresented sequences       JC1A_R1.fastq.gz
FAIL    Adapter Content JC1A_R1.fastq.gz  </code></pre>
</div>
<p>The summary file gives us a list of tests that FastQC ran and tells
us whether this sample passed, failed, or is borderline
(<code>WARN</code>). Remember, to quit from <code>less</code>, you must
type <code>q</code>.</p>
</section><section><h2 class="section-heading" id="documenting-our-work">Documenting our work<a class="anchor" aria-label="anchor" href="#documenting-our-work"></a></h2>
<hr class="half-width"><p>We can make a record of the results we obtained for all our samples
by concatenating all of our <code>summary.txt</code> files into a single
file using the <code>cat</code> command. We’ll call this
<code>fastqc_summaries.txt</code> and store it to
<code>~/dc_workshop/docs</code>.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/dc_workshop/docs</span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a><span class="ex">$</span> cat <span class="pp">*</span>/summary.txt <span class="op">&gt;</span> ~/dc_workshop/docs/fastqc_summaries.txt</span></code></pre>
</div>
<div id="exercise-4-quality-tests" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4-quality-tests" class="callout-inner">
<h3 class="callout-title">Exercise 4: Quality tests</h3>
<div class="callout-content">
<p>Which samples failed at least one of FastQC’s quality tests? What
test(s) did those samples failed</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p>We can get the list of all failed tests using <code>grep</code>.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/docs</span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a><span class="ex">$</span> grep FAIL fastqc_summaries.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>FAIL    Per base sequence quality       JC1A_R1.fastq.gz
FAIL    Per sequence GC content JC1A_R1.fastq.gz
FAIL    Sequence Duplication Levels     JC1A_R1.fastq.gz
FAIL    Adapter Content JC1A_R1.fastq.gz
FAIL    Per base sequence quality       JC1A_R2.fastq.gz
FAIL    Per sequence GC content JC1A_R2.fastq.gz
FAIL    Sequence Duplication Levels     JC1A_R2.fastq.gz
FAIL    Adapter Content JC1A_R2.fastq.gz
FAIL    Per base sequence content       JP4D_R1.fastq
FAIL    Adapter Content JP4D_R1.fastq
FAIL    Per base sequence quality       JP4D_R2.fastq.gz
FAIL    Per base sequence content       JP4D_R2.fastq.gz
FAIL    Adapter Content JP4D_R2.fastq.gz</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="quality-of-large-datasets" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="quality-of-large-datasets" class="callout-inner">
<h3 class="callout-title">Quality of large datasets</h3>
<div class="callout-content">
<p>Explore <a href="https://multiqc.info/" class="external-link">MultiQC</a> if you want a
tool that can show the quality of many samples at once.</p>
</div>
</div>
</div>
<div id="quality-encodings-vary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="quality-encodings-vary" class="callout-inner">
<h3 class="callout-title">Quality Encodings Vary</h3>
<div class="callout-content">
<p>Although we’ve used a particular quality encoding system to
demonstrate the interpretation of read quality, different sequencing
machines use different encoding systems. This means that depending on
which sequencer you use to generate your data, a <code>#</code> may not
indicate a poor quality base call.</p>
<p>This mainly relates to older Solexa/Illumina data. However, it’s
essential that you know which sequencing platform was used to generate
your data to tell your quality control program which encoding to use. If
you choose the wrong encoding, you run the risk of throwing away good
reads or (even worse) not throwing away bad reads!</p>
</div>
</div>
</div>
<div id="bonus-exercise-automating-a-quality-control-workflow" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="bonus-exercise-automating-a-quality-control-workflow" class="callout-inner">
<h3 class="callout-title">Bonus Exercise: Automating a quality control workflow</h3>
<div class="callout-content">
<p>If you lose your FastQC analysis results. How would you do it again
but faster than the first time? As we have seen in a previous lesson,
making scripts for repetitive tasks is a very efficient practice during
bioinformatic pipelines.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<p>Make a new script with nano</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="fu">nano</span> quality_control.sh</span></code></pre>
</div>
<p>Paste inside the commands that we used along with <code>echo</code>
commands that shows you how the script is running.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="bu">set</span> <span class="at">-e</span> <span class="co"># This will ensure that our script will exit if an error occurs</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a><span class="bu">cd</span> ~/dc_workshop/data/untrimmed_fastq/</span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Running FastQC ..."</span></span>
<span id="cb42-5"><a href="#cb42-5" tabindex="-1"></a><span class="ex">fastqc</span> <span class="pp">*</span>.fastq<span class="pp">*</span></span>
<span id="cb42-6"><a href="#cb42-6" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/dc_workshop/results/fastqc_untrimmed_reads</span>
<span id="cb42-8"><a href="#cb42-8" tabindex="-1"></a></span>
<span id="cb42-9"><a href="#cb42-9" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Saving FastQC results..."</span></span>
<span id="cb42-10"><a href="#cb42-10" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.zip ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb42-11"><a href="#cb42-11" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.html ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb42-12"><a href="#cb42-12" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" tabindex="-1"></a><span class="bu">cd</span> ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb42-14"><a href="#cb42-14" tabindex="-1"></a></span>
<span id="cb42-15"><a href="#cb42-15" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Unzipping..."</span></span>
<span id="cb42-16"><a href="#cb42-16" tabindex="-1"></a><span class="cf">for</span> filename <span class="kw">in</span> <span class="pp">*</span>.zip</span>
<span id="cb42-17"><a href="#cb42-17" tabindex="-1"></a>    <span class="cf">do</span></span>
<span id="cb42-18"><a href="#cb42-18" tabindex="-1"></a>    <span class="fu">unzip</span> <span class="va">$filename</span></span>
<span id="cb42-19"><a href="#cb42-19" tabindex="-1"></a>    <span class="cf">done</span></span>
<span id="cb42-20"><a href="#cb42-20" tabindex="-1"></a></span>
<span id="cb42-21"><a href="#cb42-21" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Saving summary..."</span></span>
<span id="cb42-22"><a href="#cb42-22" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/dc_workshop/docs</span>
<span id="cb42-23"><a href="#cb42-23" tabindex="-1"></a><span class="fu">cat</span> <span class="pp">*</span>/summary.txt <span class="op">&gt;</span> ~/dc_workshop/docs/fastqc_summaries.txt</span></code></pre>
</div>
<p>If we were to run this script, it would ask us for confirmation to
redo several steps because we already did all of them. If you want to,
you can run it to check that it works, but it is not necessary if you
already completed every step of the previous episode.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>It is important to know the quality of our data to make decisions in
the subsequent steps.</li>
<li>FastQC is a program that allows us to know the quality of FASTQ
files.</li>
<li>
<code>for</code> loops let you perform the same operations on
multiple files with a single command.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-background-metadata.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-trimming-filtering.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-background-metadata.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Starting a
        </a>
        <a class="chapter-link float-end" href="../instructor/03-trimming-filtering.html" rel="next">
          Next: Trimming and...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/02-assessing-read-quality.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/02-assessing-read-quality.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Assessing Read Quality",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/02-assessing-read-quality.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/02-assessing-read-quality.html",
  "dateCreated": "2020-08-24",
  "dateModified": "2025-04-06",
  "datePublished": "2025-04-22"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

