<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Processing and Visualization for Metagenomics: Diversity Tackled With R</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/cp/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/cp/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/cp/favicon-16x16.png"><link rel="manifest" href="../favicons/cp/site.webmanifest"><link rel="mask-icon" href="../favicons/cp/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav carpentries"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="cp-logo" alt="Lesson Description" src="../assets/images/carpentries-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../08-Diversity-tackled-with-R.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav carpentries" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/carpentries-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="BinningRefinmentDereplication.html">Binning, Refinamiento y Desreplicación</a></li><li><a class="dropdown-item" href="ImprovingMAG.html">Reconstrucción de Genomas a partir de Metagenomas (MAGs)</a></li><li><a class="dropdown-item" href="MAGsTaxonomyMetabolism.html">Clasificación taxonómica y Metabolismo</a></li><li><a class="dropdown-item" href="amplicon.html">Amplicon metagenomics</a></li><li><a class="dropdown-item" href="data.html">Data</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 71%" class="percentage">
    71%
  </div>
  <div class="progress carpentries">
    <div class="progress-bar carpentries" role="progressbar" style="width: 71%" aria-valuenow="71" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../08-Diversity-tackled-with-R.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-assessing-read-quality.html">2. Assessing Read Quality</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-taxonomic.html">6. Taxonomic Assignment</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-phyloseq.html">7. Exploring Taxonomy with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        8. Diversity Tackled With R
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#first-plunge-into-diversity">First plunge into diversity</a></li>
<li><a href="#%CE%B1-diversity">α diversity</a></li>
<li><a href="#plot-alpha-diversity">Plot alpha diversity</a></li>
<li><a href="#absolute-and-relative-abundances">Absolute and relative abundances</a></li>
<li><a href="#beta-diversity">Beta diversity</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-abundance-analyses.html">9. Taxonomic Analysis with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="BinningRefinmentDereplication.html">Binning, Refinamiento y Desreplicación</a></li><li><a href="ImprovingMAG.html">Reconstrucción de Genomas a partir de Metagenomas (MAGs)</a></li><li><a href="MAGsTaxonomyMetabolism.html">Clasificación taxonómica y Metabolismo</a></li><li><a href="amplicon.html">Amplicon metagenomics</a></li><li><a href="data.html">Data</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/07-phyloseq.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/09-abundance-analyses.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/07-phyloseq.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Exploring Taxonomy
        </a>
        <a class="chapter-link float-end" href="../instructor/09-abundance-analyses.html" rel="next">
          Next: Taxonomic Analysis...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Diversity Tackled With R</h1>
        <p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/08-Diversity-tackled-with-R.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can we measure diversity?</li>
<li>How can I use R to analyze diversity?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Plot alpha and beta diversity.</li>
</ul></div>
</div>
</div>
</div>
</div>
<p><em>Look at your fingers; controlled by the mind can do great things.
However, imagine if each one has a little brain of its own, with
different ideas, desires, and fears ¡How wonderful things will be made
out of an artist with such hands!</em> -Ode to multidisciplinarity</p>
<section><h2 class="section-heading" id="first-plunge-into-diversity">First plunge into diversity<a class="anchor" aria-label="anchor" href="#first-plunge-into-diversity"></a></h2>
<hr class="half-width"><p>Species diversity, in its simplest definition, is the number of
species in a particular area and their relative abundance (evenness).
Once we know the taxonomic composition of our metagenomes, we can do
diversity analyses. Here we will discuss the two most used diversity
metrics, α diversity (within one metagenome) and β (across
metagenomes).</p>
<ul><li>α Diversity: Can be represented only as richness (<em>, i.e.,</em>
the number of different species in an environment), or it can be
measured considering the abundance of the species in the environment as
well (<em>i.e.,</em> the number of individuals of each species inside
the environment). To measure α-diversity, we use indexes such as
Shannon’s, Simpson’s, Chao1, etc.</li>
</ul><p><a href="../fig/03-08-01.png"></a></p>
<figure><img src="../fig/03-08-01.png" alt="Alpha diversity diagram: In lake A, we have three fishes, each one of a different species. On lake B, we have two fish, each of a different species. Moreover, we have four fish in lake C, each of different species." class="figure mx-auto d-block"></figure><p></p>
<p><em> Figure 1. Alpha diversity is calculated according to fish
diversity in a pond. Here, alpha diversity is represented in its
simplest way: Richness. <em></em></em></p>
<ul><li>β diversity is the difference (measured as distance) between two or
more environments. It can be measured with metrics like Bray-Curtis
dissimilarity, Jaccard distance, or UniFrac distance, to name a few.
Each one of this measures are focused on a characteristic of the
community (<em>e.g.,</em> Unifrac distance measures the phylogenetic
relationship between the species of the community).</li>
</ul><p>In the next example, we will look at the α and the β components of
the diversity of a dataset of fishes in three lakes. The most simple way
to calculate the β-diversity is to calculate the distinct species
between two lakes (sites). Let us take as an example the diversity
between Lake A and Lake B. The number of species in Lake A is 3. To this
quantity, we will subtract the number of these species that are shared
with the Lake B: 2. So the number of unique species in Lake A compared
to Lake B is (3-2) = 1. To this number, we will sum the result of the
same operations but now take Lake B as our reference site. In the end,
the β diversity between Lake A and Lake B is (3-2) + (3-2) = 2. This
process can be repeated, taking each pair of lakes as the focused
sites.</p>
<p><a href="../fig/03-08-02.png"></a></p>
<figure><img src="../fig/03-08-02.png" alt=" Alpha and Beta diversity diagram: Each Lake has a different number of species, and each species has a different number of fish individuals. Both metrics are taken into account to measure alfa and beta diversity." class="figure mx-auto d-block"></figure><p></p>
<p><em> Figure 2. Alpha and beta diversity indexes of fishes in a
pond.<em></em></em></p>
<p>If you want to read more about diversity, we recommend to you this <a href="https://link.springer.com/article/10.1007/s00442-010-1812-0" class="external-link">paper</a>
on the concept of diversity.</p>
</section><section><h2 class="section-heading" id="α-diversity">α diversity<a class="anchor" aria-label="anchor" href="#%CE%B1-diversity"></a></h2>
<hr class="half-width"><p>|——————-+—————————————————————————————————————–|<br>
| Diversity Indices | Description |<br>
|——————-+—————————————————————————————————————–|<br>
| Shannon (H) | Estimation of species richness and species evenness.
More weight on richness. |<br>
|——————-+—————————————————————————————————————–|<br>
| Simpson’s (D) |Estimation of species richness and species evenness.
More weigth on evenness. |<br>
|——————-+—————————————————————————————————————–|<br>
| Chao1 | Abundance based on species represented by a single individual
(singletons) and two individuals (doubletons). |<br>
|——————-+—————————————————————————————————————–|</p>
<ul><li>Shannon (H):</li>
</ul><table class="table"><colgroup><col width="60%"><col width="39%"></colgroup><thead><tr class="header"><th align="center">Variable</th>
<th align="center">Definition</th>
</tr></thead><tbody><tr class="odd"><td align="center">$ H = - \sum_{i=1}^{S} p_{i} \ln{p_{i}} $</td>
<td align="center">Definition</td>
</tr><tr class="even"><td align="center">$ S $</td>
<td align="center">Number of OTUs</td>
</tr><tr class="odd"><td align="center">$ p_{i} $</td>
<td align="center">The proportion of the community represented by OTU
i</td>
</tr></tbody></table><!-- <img src="https://render.githubusercontent.com/render/math?math=H=-\sum_{i=1}^{S}p_i\:ln{p_i}"> | Definition
<img src="https://render.githubusercontent.com/render/math?math=S"> | Number of OTUs
<img src="https://render.githubusercontent.com/render/math?math=p_i">|  The proportion of the community represented by OTU i    --><ul><li>Simpson’s (D)</li>
</ul><table class="table"><colgroup><col width="60%"><col width="39%"></colgroup><thead><tr class="header"><th align="center">Variable</th>
<th align="center">Definition</th>
</tr></thead><tbody><tr class="odd"><td align="center">$ D = \frac{1}{\sum_{i=1}^{S} p_{i}^{2}} $</td>
<td align="center">Definition</td>
</tr><tr class="even"><td align="center">$ S $</td>
<td align="center">Total number of the species in the community</td>
</tr><tr class="odd"><td align="center">$ p_{i} $</td>
<td align="center">Proportion of community represented by OTU i</td>
</tr></tbody></table><!-- <img src="https://render.githubusercontent.com/render/math?math=D=\frac{1}{\sum_{i=1}^{S}p_i^2}">| Definition
<img src="https://render.githubusercontent.com/render/math?math=S"> | Total number of the species in the community
<img src="https://render.githubusercontent.com/render/math?math=p_i" align="middle"> | Proportion of community represented by OTU i     --><ul><li>Chao1</li>
</ul><table class="table"><colgroup><col width="60%"><col width="39%"></colgroup><thead><tr class="header"><th align="center">Variable</th>
<th align="center">Definition</th>
</tr></thead><tbody><tr class="odd"><td align="center">$ S_{chao1} = S_{Obs} + \frac{F_{1} \times (F_{1} -
1)}{2 \times (F_{2} + 1)} $</td>
<td align="center">Count of singletons and doubletons respectively</td>
</tr><tr class="even"><td align="center">$ F_{1}, F_{2} $</td>
<td align="center">Count of singletons and doubletons respectively</td>
</tr><tr class="odd"><td align="center">$ S_{chao1}=S_{Obs} $</td>
<td align="center">The number of observed species</td>
</tr></tbody></table><!-- <img src="fig/equation.svg">| Definition
<img src="https://render.githubusercontent.com/render/math?math=F_1,F_2">|Count of singletons and doubletons respectively
<img src="https://render.githubusercontent.com/render/math?math=S_{chao1}=S_{Obs}">| The number of observed species   --><p><!-- comment we use https://viereck.ch/latex-to-svg/ to convert from latex to SVG because Chao equation did not render correctly with GitHub math!--></p>
<div class="section level3">
<h3 id="β-diversity">β diversity<a class="anchor" aria-label="anchor" href="#%CE%B2-diversity"></a></h3>
<p>Diversity β measures how different two or more communities are,
either in their composition (richness) or in the abundance of the
organisms that compose it (abundance).</p>
<ul><li>Bray-Curtis dissimilarity: The difference in richness and abundance
across environments (samples). Weight on abundance. Measures the
differences from 0 (equal communities) to 1 (different communities)</li>
<li>Jaccard distance: Based on the presence/absence of species
(diversity). It goes from 0 (same species in the community) to 1 (no
species in common)</li>
<li>UniFrac: Measures the phylogenetic distance; how alike the trees in
each community are. There are two types, without weights (diversity) and
with weights (diversity and abundance)</li>
</ul><p>There are different ways to plot and show the results of such
analysis. Among others, PCA, PCoA, or NMDS analysis are widely used.</p>
<div id="exercise-1-simple-measure-of-alpha-and-beta-diversities." class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-simple-measure-of-alpha-and-beta-diversities." class="callout-inner">
<h3 class="callout-title">Exercise 1: Simple measure of alpha and beta diversities.</h3>
<div class="callout-content">
<p>In the next picture, there are two lakes with different fish species:
<a href="../%7B%7B%20page.root%20%7D%7D/fig/03-08-03.png">
<img src="../%7B%7B%20page.root%20%7D%7D/fig/03-08-03.png" alt="In lake A, we have four different species, two of these species have three specimens each one. This Lake also has two specimens of a third species and only one specimen of a fourth specie. We got nine fish in total. Lake B has only three different species, the most populated species is also present in lake A and has five specimens, and we have only one specimen of each of the other two species. We got seven species total in lake B " class="figure"></a> <em> Figure 3. <em></em></em></p>
<p>Which of the options below is true for the alpha diversity in lakes
A, B, and beta diversity between lakes A and B, respectively?</p>
<ol style="list-style-type: decimal"><li>4, 3, 1</li>
<li>4, 3, 5</li>
<li>9, 7, 16</li>
</ol><p>Please, paste your result on the collaborative document provided by
instructors. <em>Hic Sunt Leones!</em> (<em>Here be Lions!</em>)</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Answer: 2. 4, 3, 5 <strong>Alpha diversity</strong> in this case, is
the sum of different species. Lake <strong>A</strong> has
<strong>4</strong> different species and lake <strong>B</strong> has
<strong>3</strong> different species. <strong>Beta diversity</strong>
refers to the difference between lake A and lake B. If we use the
formula in <em>Figure 2</em> we can see that to calculate beta
diversity, we have to detect the number of species and the number of
shared species in both lakes. There is only one shared species, so we
have to subtract the number of shared species to the total species and
sum the result. In this case, in lake A, we have 4 different species and
one shared species with lake B (4-1)=3, and in lake B we have three
species and one shared species with lake A (3-1)=2. If we add 3+2, the
result is 5.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="plot-alpha-diversity">Plot alpha diversity<a class="anchor" aria-label="anchor" href="#plot-alpha-diversity"></a></h2>
<hr class="half-width"><p>We want to know the bacterial diversity, so we will prune all
non-bacterial organisms in our <code>merged_metagenomes</code> Phyloseq
object. To do this, we will make a subset of all bacterial groups and
save them.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(merged_metagenomes, Kingdom <span class="sc">==</span> <span class="st">"Bacteria"</span>)</span></code></pre>
</div>
<p>Now let us look at some statistics of our metagenomes. By the output
of the <code>sample_sums()</code> command, we can see how many reads
there are in the library. Library JC1A is the smallest with 18412 reads,
while library JP4D is the largest with 149590 reads.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 4024 taxa and 3 samples ]
tax_table()   Taxonomy Table:    [ 4024 taxa by 7 taxonomic ranks ]</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">sample_sums</span>(merged_metagenomes)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  JC1A   JP4D   JP41
 18412 149590  76589 </code></pre>
</div>
<p>Also, the Max, Min, and Mean output on <code>summary()</code> can
give us a sense of the evenness. For example, the OTU that occurs more
times in the sample JC1A occurs 399 times, and on average in sample
JP4D, an OTU occurs in 37.17 reads.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">summary</span>(merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      JC1A              JP4D              JP41
 Min.   :  0.000   Min.   :   0.00   Min.   :   0.00
 1st Qu.:  0.000   1st Qu.:   3.00   1st Qu.:   1.00
 Median :  0.000   Median :   7.00   Median :   5.00
 Mean   :  4.575   Mean   :  37.17   Mean   :  19.03
 3rd Qu.:  2.000   3rd Qu.:  21.00   3rd Qu.:  14.00
 Max.   :399.000   Max.   :6551.00   Max.   :1994.00  </code></pre>
</div>
<p>To have a more visual representation of the diversity inside the
samples (i.e., α diversity), we can now look at a graph created using
Phyloseq:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes, </span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>              <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>)) </span></code></pre>
</div>
<p><a href="../fig/03-08-04.png"></a></p>
<figure><img src="../%7B%7B%20page.root%20%7D%7D/fig/03-08-04.png" alt="A figure divided in three
panels. Each of these panels represents a different alpha diversity index.
Inside this section, each point represents the value assigned on this index to
the three different samples. The different indexes give
different values to the same sample." class="figure mx-auto d-block"></figure><p></p>
<p><em> Figure 4. Alpha diversity indexes for both samples. <em></em></em></p>
<p>Each of these metrics can give an insight into the distribution of
the OTUs inside our samples. For example, the Chao1 diversity index
gives more weight to singletons and doubletons observed in our samples,
while Shannon is an entropy index remarking the impossibility of taking
two reads out of the metagenome “bag” and that these two will belong to
the same OTU.</p>
<div id="exercise-2-exploring-function-flags." class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-exploring-function-flags." class="callout-inner">
<h3 class="callout-title">Exercise 2: Exploring function flags.</h3>
<div class="callout-content">
<p>While using the help provided, explore these options available for
the function in <code>plot_richness()</code>:</p>
<ol style="list-style-type: decimal"><li><code>nrow</code></li>
<li><code>sortby</code></li>
<li><code>title</code></li>
</ol><p>Use these options to generate new figures that show you other ways to
present the data.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The code and the plot using the three options will look as follows:
The “title” option adds a title to the figure.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes, </span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>             <span class="at">title =</span> <span class="st">"Alpha diversity indexes for three samples from Cuatro Cienegas"</span>,</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>             <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>))</span></code></pre>
</div>
<p><a href="../fig/03-08-05.png"> <img src="../fig/03-08-05.png" alt="" class="figure"></a> <em> Figure 5. Alpha diversity plot with the title. <em></em></em></p>
<p>The “nrow” option arranges the graphics horizontally.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes,</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>             <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>),</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>             <span class="at">nrow=</span><span class="dv">3</span>)</span></code></pre>
</div>
<p><a href="../fig/03-08-06.png"> <img src="../fig/03-08-06.png" alt="" class="figure"></a> <em> Figure 6. Alpha diversity plot with the three panels arranged
in rows. <em></em></em></p>
<p>The “sortby” option orders the samples from least to greatest
diversity depending on the parameter. In this case, it is ordered by
“Shannon” and tells us that the JP4D sample has the lowest diversity and
the JP41 sample the highest.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_richness</span>(<span class="at">physeq =</span> merged_metagenomes, </span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>             <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>,<span class="st">"Chao1"</span>,<span class="st">"Shannon"</span>),</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>             <span class="at">sortby =</span> <span class="st">"Shannon"</span>) </span></code></pre>
</div>
<p><a href="../fig/03-08-07.png">
<img src="../fig/03-08-07.png" alt="The same panels as before, but now the samples are arranged horizontally according to the values in the Shannon index panel." class="figure"></a> <em> Figure 7. Samples sorted by Shannon in alpha diversity index
plots. <em></em></em></p>
<p>Considering those mentioned above, together with the three graphs, we
can say that JP41 and JP4D present a high diversity concerning the JC1A.
Moreover, the diversity of the sample JP41 is mainly given by singletons
or doubletons. Instead, the diversity of JP4D is given by species in
much greater abundance. Although the values of H (Shannon) above three
are considered to have a lot of diversity.</p>
</div>
</div>
</div>
</div>
<p>A caution when comparing samples is that differences in some alpha
indexes may be the consequence of the difference in the total number of
reads of the samples. A sample with more reads is more likely to have
more different OTUs, so some normalization is needed. Here we will work
with relative abundances, but other approaches could help reduce this
bias.</p>
</section><section><h2 class="section-heading" id="absolute-and-relative-abundances">Absolute and relative abundances<a class="anchor" aria-label="anchor" href="#absolute-and-relative-abundances"></a></h2>
<hr class="half-width"><p>From the read counts that we just saw, it is evident that there is a
great difference in the number of total sequenced reads in each sample.
Before we further process our data, look if we have any non-identified
reads. Marked as blank (i.e.,““) on the different taxonomic levels:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">summary</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data<span class="sc">==</span> <span class="st">""</span>)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Kingdom          Phylum          Class           Order           Family          Genus          Species
 Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical
 FALSE:4024      FALSE:4024      FALSE:3886      FALSE:4015      FALSE:3967      FALSE:3866      FALSE:3540
                                 TRUE :138       TRUE :9         TRUE :57        TRUE :158       TRUE :484      </code></pre>
</div>
<p>With the command above, we can see blanks on different taxonomic
levels. For example, we have 158 blanks at the genus level. Although we
could expect to see some blanks at the species or even at the genus
level; we will get rid of the ones at the genus level to proceed with
the analysis:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="sc">&gt;</span> merged_metagenomes <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(merged_metagenomes, Genus <span class="sc">!=</span> <span class="st">""</span>) <span class="co">#Only genus that are no blank</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">summary</span>(merged_metagenomes<span class="sc">@</span>tax_table<span class="sc">@</span>.Data<span class="sc">==</span> <span class="st">""</span>)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Kingdom          Phylum          Class           Order           Family          Genus          Species
 Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical   Mode :logical
 FALSE:3866      FALSE:3866      FALSE:3739      FALSE:3860      FALSE:3858      FALSE:3866      FALSE:3527
                                 TRUE :127       TRUE :6         TRUE :8                         TRUE :339 </code></pre>
</div>
<p>Next, since our metagenomes have different sizes, it is imperative to
convert the number of assigned reads (i.e., absolute abundance) into
percentages (i.e., relative abundances) to compare them.</p>
<p>Right now, our OTU table looks like this:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(merged_metagenomes<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        JC1A JP4D JP41
1060      32  420   84
1063     316 5733 1212
2033869  135 1232  146
1850250  114  846  538
1061      42 1004  355
265       42  975  205</code></pre>
</div>
<p>To make this transformation to percentages, we will take advantage of
a function of Phyloseq.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(merged_metagenomes, <span class="cf">function</span>(x) x<span class="sc">*</span><span class="dv">100</span> <span class="sc">/</span> <span class="fu">sum</span>(x) )</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(percentages<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             JC1A      JP4D      JP41
1060    0.1877383 0.3065134 0.1179709
1063    1.8539161 4.1839080 1.7021516
2033869 0.7920211 0.8991060 0.2050447
1850250 0.6688178 0.6174056 0.7555755
1061    0.2464066 0.7327130 0.4985675
265     0.2464066 0.7115490 0.2879052</code></pre>
</div>
<p>Now, we are ready to compare the abundaces given by percantages of
the samples with beta diversity indexes.</p>
</section><section><h2 class="section-heading" id="beta-diversity">Beta diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a></h2>
<hr class="half-width"><p>As we mentioned before, the beta diversity is a measure of how alike
or different our samples are (overlap between discretely defined sets of
species or operational taxonomic units). To measure this, we need to
calculate an index that suits the objectives of our research. By the
next code, we can display all the possible distance metrics that
Phyloseq can use:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="sc">&gt;</span> distanceMethodList</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$UniFrac
[1] "unifrac"  "wunifrac"

$DPCoA
[1] "dpcoa"

$JSD
[1] "jsd"

$vegdist
 [1] "manhattan"  "euclidean"  "canberra"   "bray"       "kulczynski" "jaccard"    "gower"
 [8] "altGower"   "morisita"   "horn"       "mountford"  "raup"       "binomial"   "chao"
[15] "cao"

$betadiver
 [1] "w"   "-1"  "c"   "wb"  "r"   "I"   "e"   "t"   "me"  "j"   "sor" "m"   "-2"  "co"  "cc"  "g"
[17] "-3"  "l"   "19"  "hk"  "rlb" "sim" "gl"  "z"

$dist
[1] "maximum"   "binary"    "minkowski"

$designdist
[1] "ANY"</code></pre>
</div>
<p>Describing all these possible distance metrics is beyond the scope of
this lesson, but here we show which are the ones that need a
phylogenetic relationship between the species-OTUs present in our
samples:</p>
<ul><li>Unifrac</li>
<li>Weight-Unifrac</li>
<li>DPCoA</li>
</ul><p>We do not have a phylogenetic tree or phylogenetic relationships. So
we can not use any of those three. We will use <a href="https://www.pelagicos.net/MARS6300/readings/Bray_&amp;_Curtis_1957.pdf" class="external-link">Bray-curtis</a>
since it is one of the most robust and widely used distance metrics to
calculate beta diversity.</p>
<p><strong>Let’s keep this up!</strong> We already have all we need to
begin the beta diversity analysis. We will use the Phyloseq command
<code>ordinate</code> to generate a new object where the distances
between our samples will be allocated after calculating them. For this
command, we need to specify which method we will use to generate a
matrix. In this example, we will use Non-Metric Multidimensional Scaling
or <a href="https://academic.oup.com/bioinformatics/article/21/6/730/199398" class="external-link">NMDS</a>.
NMDS attempts to represent the pairwise dissimilarity between objects in
a low-dimensional space, in this case, a two-dimensional plot.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="sc">&gt;</span> meta_ord <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(<span class="at">physeq =</span> percentages, <span class="at">method =</span> <span class="st">"NMDS"</span>, </span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>                     <span class="at">distance =</span> <span class="st">"bray"</span>)</span></code></pre>
</div>
<p>If you get some warning messages after running this script, fear not.
It is because we only have three samples. Few samples make the algorithm
warn about the lack of difficulty in generating the distance matrix.</p>
<p>By now, we just need the command <code>plot_ordination()</code> to
see the results from our beta diversity analysis:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_ordination</span>(<span class="at">physeq =</span> percentages, <span class="at">ordination =</span> meta_ord)</span></code></pre>
</div>
<p><a href="../fig/03-08-08.png">
<img src="../fig/03-08-08.png" alt="Plot with NMDS1 as a label in x-axis that goes from -0.4 to 0.2 and NMDS2 in y-axis that goes from -0.2 to 0.1. The plot has three dots that are not clustered in any way." class="figure"></a> <em> Figure 8. Beta diversity with NMDS of our three samples.
<em></em></em></p>
<p>In this NMDS plot, each point represents the combined abundance of
all its OTUs. As depicted, each sample occupies space in the plot
without forming any clusters. This output is because each sample is
different enough to be considered its own point in the NMDS space.</p>
<div id="exercise-3-add-metadata-to-beta-diversity-visualization" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-3-add-metadata-to-beta-diversity-visualization" class="callout-inner">
<h3 class="callout-title">Exercise 3: Add metadata to beta diversity visualization</h3>
<div class="callout-content">
<p>In the following figure, the beta diversity graph we produced earlier
has been enriched. Look at the code below and answer:</p>
<ol style="list-style-type: decimal"><li>Which instruction colored the samples by their corresponding
treatment?</li>
<li>What did the instruction <code>geom_text</code> do?</li>
<li>What other difference do you notice with our previous graph?</li>
<li>Do you see some clustering of the samples according to their
treatment?</li>
</ol><p><a href="../fig/03-08-09.png">
<img src="../fig/03-08-09.png" alt="The distance between the three samples, JC1A, JP4D and JP41 is shown in a plane. Each sample has a legend and a color. The color is according to the metadata treatment. There are three possible treatments in the legend: Control mesocosm, Fertilized pond, and Unenriched pond" class="figure">
~~~ metadata_cuatroc &lt;- data.frame(Sample=c(“JC1A”, “JP4D”, “JP41”),
Treatment=c(“Control mesocosm”, “Fertilized pond”, “Unenriched pond”)) #
Making dataframe with metadata<br>
rownames(metadata_cuatroc) &lt;- metadata_cuatroc$Sample # Using sample
names as row names<br></a><a href="mailto:percentages@sam" class="email">percentages@sam</a>_data
&lt;- sample_data(metadata_cuatroc) # Adding metadata to sam_data table
of phyloseq object percentages<br>
meta_ord &lt;- ordinate(physeq = percentages, method = “NMDS”, distance
= “bray”) # Calculating beta diversity<br>
plot_ordination(physeq = percentages, ordination = meta_ord, color =
“Treatment”) + # Plotting beta diversity.<br>
geom_text(mapping = aes(label = Sample), size = 3, vjust = 1.5)<br>
~~~ {: .language-r} &gt; ## Solution &gt; The flag
<code>color = "Treatment"</code> applied a color to each sample
according to its treatment, in the <code>plot_ordination</code> of the
object <code>percentages</code>.<br>
&gt; The <code>geom_text</code> instruction added the names of the
sample to the graph. This could have added any text, with the
instruction <code>label = Sample</code> we specified to add the names of
the samples as text. With <code>size</code> we adjusted the size of the
text, and with <code>vjust</code> we adjusted the position &gt;so the
text would not overlap with the dots.<br>
&gt; There are three possible treatments, Control mesocosm, Fertilized,
and Unfertilized pond.<br>
&gt; We do not observe any kind of clustering in these three samples.
More data would show if samples with similar treatments are clustered
together.<br>
{: .solution}</p>
</div>
</div>
</div>
<div id="discussion-indexes-of-diversity" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion-indexes-of-diversity" class="callout-inner">
<h3 class="callout-title">Discussion: Indexes of diversity</h3>
<div class="callout-content">
<p>Why do you think we need different indexes to asses diversity? What
index will you use to assess the impact of rare, low-abundance taxa?</p>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>It will be difficult (if not impossible) to take two communities and
observe the same distribution of all members. This outcome is because
there are a lot of <strong>factors</strong> affecting these lineages.
Some of the <strong>environmental factors</strong> are temperature, pH,
and nutrient concentration. Also, the interactions of these populations,
such as competence, inhibition of other populations, and growth speed,
are an important driver of variation (<strong>biotic factor</strong>). A
combination of the <strong>factors mentioned above</strong>, can
interact to maintain some populations with low abundance (<strong>rare
taxa</strong> In order to have ways to assess hypotheses regarding which
of these processes can be affecting the community, we use all these
different indexes. Some emphasize the number of species and other the
evenness of the OTUs. To assess the impact of low abundance lineages,
one alpha diversity index widely used is the Chao1 index.</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Alpha diversity measures the intra-sample diversity.</li>
<li>Beta diversity measures the inter-sample diversity.</li>
<li>Phyloseq includes diversity analyses such as alpha and beta
diversity calculation.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/07-phyloseq.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/09-abundance-analyses.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/07-phyloseq.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Exploring Taxonomy
        </a>
        <a class="chapter-link float-end" href="../instructor/09-abundance-analyses.html" rel="next">
          Next: Taxonomic Analysis...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/08-Diversity-tackled-with-R.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/08-Diversity-tackled-with-R.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Diversity Tackled With R",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/08-Diversity-tackled-with-R.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/08-Diversity-tackled-with-R.html",
  "dateCreated": "2025-04-06",
  "dateModified": "2025-04-06",
  "datePublished": "2025-04-06"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

