<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Processing and Visualization for Metagenomics: Taxonomic Analysis with R</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/lab/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/lab/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/lab/favicon-16x16.png"><link rel="manifest" href="../favicons/lab/site.webmanifest"><link rel="mask-icon" href="../favicons/lab/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav lab"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="lab-logo" alt="Lesson Description" src="../assets/images/lab-logo.svg"><abbr class="badge peer-reviewed" title="This lesson has passed peer review.">
          <a href="https://doi.org/10.21105/jose.00209" class="external-link alert-link">
            <i aria-hidden="true" class="icon" data-feather="user-check" style="border-radius: 5px"></i>
            DOI: 10.21105/jose.00209
          </a>
        </abbr>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../09-abundance-analyses.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav lab" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/lab-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Processing and Visualization for Metagenomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Processing and Visualization for Metagenomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="BinningRefinmentDereplication.html">Binning, Refinamiento y Desreplicación</a></li><li><a class="dropdown-item" href="ImprovingMAG.html">Reconstrucción de Genomas a partir de Metagenomas (MAGs)</a></li><li><a class="dropdown-item" href="MAGsTaxonomyMetabolism.html">Clasificación taxonómica y Metabolismo</a></li><li><a class="dropdown-item" href="amplicon.html">Amplicon metagenomics</a></li><li><a class="dropdown-item" href="data.html">Data</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Processing and Visualization for Metagenomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 83%" class="percentage">
    83%
  </div>
  <div class="progress lab">
    <div class="progress-bar lab" role="progressbar" style="width: 83%" aria-valuenow="83" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../09-abundance-analyses.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background-metadata.html">1. Starting a Metagenomics Project</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-assessing-read-quality.html">2. Assessing Read Quality</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming-filtering.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-assembly.html">4. Metagenome Assembly</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-binning.html">5. Metagenome Binning</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-taxonomic.html">6. Taxonomic Assignment</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-phyloseq.html">7. Exploring Taxonomy with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-Diversity-tackled-with-R.html">8. Diversity Tackled With R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        9. Taxonomic Analysis with R
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#explore-our-samples-at-specific-taxonomic-levels">Explore our samples at specific taxonomic levels</a></li>
<li><a href="#going-further-lets-take-an-exciting-lineage-and-explore-it-thoroughly">Going further, let’s take an exciting lineage and explore it
thoroughly</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-OtherResources.html">10. Other Resources</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="BinningRefinmentDereplication.html">Binning, Refinamiento y Desreplicación</a></li><li><a href="ImprovingMAG.html">Reconstrucción de Genomas a partir de Metagenomas (MAGs)</a></li><li><a href="MAGsTaxonomyMetabolism.html">Clasificación taxonómica y Metabolismo</a></li><li><a href="amplicon.html">Amplicon metagenomics</a></li><li><a href="data.html">Data</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary - Data processing and visualization for metagenomics</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/08-Diversity-tackled-with-R.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/10-OtherResources.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/08-Diversity-tackled-with-R.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Diversity Tackled
        </a>
        <a class="chapter-link float-end" href="../instructor/10-OtherResources.html" rel="next">
          Next: Other Resources...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Taxonomic Analysis with R</h1>
        <p>Last updated on 2025-04-06 |

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/09-abundance-analyses.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can we know which taxa are in our samples?</li>
<li>How can we compare depth-contrasting samples?</li>
<li>How can we manipulate our data to deliver a message?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Manipulate data types inside your phyloseq object.</li>
<li>Extract specific information from taxonomic-assignation data.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="explore-our-samples-at-specific-taxonomic-levels">Explore our samples at specific taxonomic levels<a class="anchor" aria-label="anchor" href="#explore-our-samples-at-specific-taxonomic-levels"></a></h2>
<hr class="half-width"><p>With the taxonomic assignment information that we obtained from
Kraken, we have measured diversity, and we have visualized the taxa
inside each sample with Krona and Pavian, but Phyloseq allows us to make
this visualization more flexible and personalized. So now, we will use
Phyloseq to make abundance plots of the taxa in our samples.</p>
<p>We will start our exploration at the Phylum level. In order to group
all the OTUs that have the same taxonomy at a certain taxonomic rank, we
will use the function <code>tax_glom()</code>.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_glom <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(percentages, <span class="at">taxrank =</span> <span class="st">'Phylum'</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">View</span>(percentages_glom<span class="sc">@</span>tax_table<span class="sc">@</span>.Data)</span></code></pre>
</div>
<p><a href="../fig/03-09-01.png">
<img src="../fig/03-09-01.png" alt="Table containing the
  taxonomic information of each of the OTUs inside the three samples. Here,
  we can see only the Phylum column has information, leaving the other
  taxonomic levels blank." class="figure"></a> <em> Figure 1. Taxonomic-data table
after agglomeration at the phylum level. <em></em></em></p>
<p>Another phyloseq function is <code>psmelt()</code>, which melts
phyloseq objects into a <code>data.frame</code> to manipulate them with
packages like <code>ggplot2</code> and <code>vegan</code>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(percentages_glom)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">str</span>(percentages_df)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame': 99 obs. of  5 variables:
 $ OTU      : chr  "1063" "1063" "1063" "2350" ...
 $ Sample   : chr  "JP4D" "JC1A" "JP41" "JP41" ...
 $ Abundance: num  85 73.5 58.7 23.8 19.1 ...
 $ Kingdom  : chr  "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
 $ Phylum   : chr  "Proteobacteria" "Proteobacteria" "Proteobacteria" "Bacteroidetes" ...</code></pre>
</div>
<p>Now, let’s create another data frame with the original data. This
structure will help us to compare the absolute with the relative
abundance and have a complete picture of our samples.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> absolute_glom <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(<span class="at">physeq =</span> merged_metagenomes, <span class="at">taxrank =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="sc">&gt;</span> absolute_df <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(absolute_glom)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">str</span>(absolute_df)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame': 99 obs. of  5 variables:
 $ OTU      : chr  "1063" "1063" "2350" "1063" ...
 $ Sample   : chr  "JP4D" "JP41" "JP41" "JC1A" ...
 $ Abundance: num  116538 41798 16964 12524 9227 ...
 $ Kingdom  : chr  "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
 $ Phylum   : chr  "Proteobacteria" "Proteobacteria" "Bacteroidetes" "Proteobacteria" ...</code></pre>
</div>
<p>With these objects and what we have learned regarding R data
structures and <code>ggplot2</code>, we can compare them with a plot.
First, let’s take some steps that will allow us to personalize our plot,
making it accessible for color blindness. We will create a color
palette. With <code>colorRampPalette</code>, we will choose eight colors
from the Dark2 palette and make a “ramp” with it; that is, convert those
eight colors to the number of colors needed to have one for each phylum
in our data frame. We need to have our Phylum column in the factor
structure for this.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> absolute_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(absolute_df<span class="sc">$</span>Phylum)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="sc">&gt;</span> phylum_colors_abs<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(absolute_df<span class="sc">$</span>Phylum)))</span></code></pre>
</div>
<p>Now, let´s create the figure for the data with absolute abundances
(<em>, i.e.,</em> <code>absolute_plot</code> object)</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="sc">&gt;</span> absolute_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span> absolute_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Phylum))<span class="sc">+</span> </span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> phylum_colors_abs)</span></code></pre>
</div>
<p>With the <code>position="stack"</code> command, we are telling the
<code>ggplot</code> function that the values must stack each other for
each sample. In this way, we will have all of our different categories
(OTUs) stacked in one bar and not each in a separate one. For more info
<a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></p>
<p>Next, we will create the figure for the representation of the
relative abundance data and ask RStudio to show us both plots thanks to
the <code>|</code> function from the library <code>patchwork</code>:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(percentages_df<span class="sc">$</span>Phylum)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="sc">&gt;</span> phylum_colors_rel<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(percentages_df<span class="sc">$</span>Phylum)))</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="sc">&gt;</span> relative_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>percentages_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Phylum))<span class="sc">+</span> </span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> phylum_colors_rel)</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="sc">&gt;</span> absolute_plot <span class="sc">|</span> relative_plot</span></code></pre>
</div>
<p><a href="../fig/03-09-02.png">
<img src="../fig/03-09-02.png" alt="A two-part plot contrasting
  the absolute versus the relative abundance of the three samples. On the right
  side, we can see how each of the bars has its own height, making it difficult
  to compare the information between samples. The right side shows
  three bars with the same height after the abundance was transformed to
  percentage inside each sample." class="figure"></a> <em> Figure 2. Taxonomic
diversity of absolute and relative abundance. <em></em></em></p>
<p>At once, we can denote the difference between the two plots and how
processing the data can enhance the display of actual results. However,
it is noticeable that we have too many taxa to adequately distinguish
the color of each one, less of the ones that hold the most incredible
abundance. In order to change that, we will use the power of data frames
and R. We will change the identification of the OTUs whose relative
abundance is less than 0.2%:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.character</span>(percentages_df<span class="sc">$</span>Phylum) <span class="co"># Return the Phylum column to be of type character</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum[percentages_df<span class="sc">$</span>Abundance <span class="sc">&lt;</span> <span class="fl">0.5</span>] <span class="ot">&lt;-</span> <span class="st">"Phyla &lt; 0.5% abund."</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(percentages_df<span class="sc">$</span>Phylum)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Proteobacteria"    "Bacteroidetes"     "Actinobacteria"    "Firmicutes"        "Cyanobacteria"
[6] "Planctomycetes"    "Verrucomicrobia"   "Phyla &lt; 0.5 abund"</code></pre>
</div>
<p>Let’s ask R to display the figures again by re-running our code:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span> percentages_df<span class="sc">$</span>Phylum <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(percentages_df<span class="sc">$</span>Phylum)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="sc">&gt;</span> phylum_colors_rel<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(percentages_df<span class="sc">$</span>Phylum)))</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="sc">&gt;</span> relative_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>percentages_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Phylum))<span class="sc">+</span> </span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> phylum_colors_rel)</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="sc">&gt;</span> absolute_plot <span class="sc">|</span> relative_plot</span></code></pre>
</div>
<p><a href="../fig/03-09-03.png">
<img src="../fig/03-09-03.png" alt="A new two-part plot with
  a reassignment of the low-abundant taxa on the right side. Compared to the
  left legend, the one in the right has fewer groups because of the process of
  reassigning the taxa with an abundance lower than 0.5 % to just one
  group/color." class="figure"></a> <em> Figure 3. Taxonomic diversity of absolute
and relative abundance with corrections. <em></em></em></p>
</section><section><h2 class="section-heading" id="going-further-lets-take-an-exciting-lineage-and-explore-it-thoroughly">Going further, let’s take an exciting lineage and explore it
thoroughly<a class="anchor" aria-label="anchor" href="#going-further-lets-take-an-exciting-lineage-and-explore-it-thoroughly"></a></h2>
<hr class="half-width"><p>As we have already reviewed, Phyloseq offers many tools to manage and
explore data. Let’s take a look at a function we already use but now
with guided exploration. The <code>subset_taxa</code> command is used to
extract specific lineages from a stated taxonomic level; we have used it
to get rid of the reads that do not belong to bacteria with
<code>merged_metagenomes &lt;- subset_taxa(merged_metagenomes, Kingdom == "Bacteria")</code>.</p>
<p>We will use it now to extract a specific phylum from our data and
explore it at a lower taxonomic level: Genus. We will take as an example
the phylum Cyanobacteria (indeed, this is a biased and arbitrary
decision, but who does not feel attracted to these incredible
microorganisms?):</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="sc">&gt;</span> cyanos <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(merged_metagenomes, Phylum <span class="sc">==</span> <span class="st">"Cyanobacteria"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">unique</span>(cyanos<span class="sc">@</span>tax_table<span class="sc">@</span>.Data[,<span class="dv">2</span>])</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Cyanobacteria"</code></pre>
</div>
<p>Let’s do a little review of all that we saw today:
<strong>Transformation of the data; Manipulation of the information; and
plotting</strong>:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_percentages <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(cyanos, <span class="cf">function</span>(x) x<span class="sc">*</span><span class="dv">100</span> <span class="sc">/</span> <span class="fu">sum</span>(x) )</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_glom <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(cyanos_percentages, <span class="at">taxrank =</span> <span class="st">"Genus"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_df <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(cyanos_glom)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_df<span class="sc">$</span>Genus[cyanos_df<span class="sc">$</span>Abundance <span class="sc">&lt;</span> <span class="dv">10</span>] <span class="ot">&lt;-</span> <span class="st">"Genera &lt; 10.0 abund"</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="sc">&gt;</span> cyanos_df<span class="sc">$</span>Genus <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(cyanos_df<span class="sc">$</span>Genus)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="sc">&gt;</span> genus_colors_cyanos<span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)) (<span class="fu">length</span>(<span class="fu">levels</span>(cyanos_df<span class="sc">$</span>Genus)))  </span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="sc">&gt;</span> plot_cyanos <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>cyanos_df, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Genus))<span class="sc">+</span> </span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> genus_colors_cyanos)</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="sc">&gt;</span> plot_cyanos</span></code></pre>
</div>
<p><a href="../fig/03-09-05.png">
<img src="../fig/03-09-05.png" alt="A new plot with three bars
  representing the absolute abundance of Cyanobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height." class="figure"></a> <em> Figure 5.
Diversity of Cyanobacteria at genus level inside our samples.<em></em></em></p>
<div id="exercise-1-taxa-agglomeration" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-taxa-agglomeration" class="callout-inner">
<h3 class="callout-title">Exercise 1: Taxa agglomeration</h3>
<div class="callout-content">
<p>With the following code, in the dataset with absolute
abundances,<br>
group together the phyla with a small number of reads to have a better
visualization of the data. Remember to check the data classes inside
your data frame.<br>
According to the <a href="https://github.com/axismaps/colorbrewer/" class="external-link">ColorBrewer</a> package
it is recommended not to have more than nine different colors in a
plot.</p>
<p>What is the correct order to run the following chunks of code?
Compare your graphs with your partners’.</p>
<p>Hic Sunt Leones! (Here be Lions!):</p>
<ol style="list-style-type: upper-alpha"><li><p><code>absolute_df$Phylum &lt;- as.factor(absolute_df$Phylum)</code></p></li>
<li><p><code>absolute_plot &lt;- ggplot(data= absolute_df, aes(x=Sample, y=Abundance, fill=Phylum))+</code><br><code>geom_bar(aes(), stat="identity", position="stack")+</code><br><code>scale_fill_manual(values = phylum_colors_abs)</code></p></li>
<li><p><code>absolute_$Phylum[absolute_$Abundance &lt; 300] &lt;- "Minoritary Phyla"</code></p></li>
<li><p><code>phylum_colors_abs&lt;- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(absolute_df$Phylum)))</code></p></li>
<li><p><code>absolute_df$Phylum &lt;- as.character(absolute_df$Phylum)</code></p></li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>By grouping the samples with less than 300 reads, we can get a more
decent plot. Certainly, this will be difficult since each sample has a
contrasting number of reads.</p>
<ol start="5" style="list-style-type: upper-alpha"><li><p><code>absolute_df$Phylum &lt;- as.character(absolute_df$Phylum)</code></p></li>
<li><p><code>absolute_df$Phylum[absolute_$Abundance &lt; 300] &lt;- "Minoritary Phyla"</code></p></li>
<li><p><code>absolute_df$Phylum &lt;- as.factor(absolute_df$Phylum)</code></p></li>
<li><p><code>phylum_colors_abs&lt;- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(absolute_df$Phylum)))</code></p></li>
<li><p><code>absolute_plot &lt;- ggplot(data= absolute_df, aes(x=Sample, y=Abundance, fill=Phylum))+</code><br><code>geom_bar(aes(), stat="identity", position="stack")+</code><br><code>scale_fill_manual(values = phylum_colors_abs)</code></p></li>
</ol><p>Show your plots:</p>
<p><code>absolute_plot | relative_plot</code></p>
<p><a href="../fig/03-09-04.png">
<img src="../fig/03-09-04.png" alt="New reassignment to the low abundant taxa on the left part of the plot. A new class has been created that contains the taxa with less than 300 reads" class="figure"></a></p>
</div>
</div>
</div>
</div>
<div id="exercise-2-recap-of-abundance-plotting" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-recap-of-abundance-plotting" class="callout-inner">
<h3 class="callout-title">Exercise 2: Recap of abundance plotting</h3>
<div class="callout-content">
<p>Match the chunk of code with its description and put them in the
correct order to create a relative abundance plot at the genus level of
a particular phylum. がんばって! (ganbatte; <em>good luck</em>):</p>
<table class="table"><colgroup><col width="91%"><col width="8%"></colgroup><thead><tr class="header"><th>Description</th>
<th>Command</th>
</tr></thead><tbody><tr class="odd"><td>plot the relative abundance at the genus levels.</td>
<td><code>plot_proteo</code></td>
</tr><tr class="even"><td>Convert all the genera with less than 3% abundance into only one
label.</td>
<td><code>proteo_percentages &lt;- transform_sample_counts(proteo, function(x) &gt;x*100 / sum(x) )</code></td>
</tr><tr class="odd"><td>Make just one row that groups all the observations of the same
genus.</td>
<td><code>proteo &lt;- subset_taxa(merged_metagenomes, Phylum == "Proteobacteria")</code></td>
</tr><tr class="even"><td>Create a phyloseq object only with the reads assigned to a certain
phylum.</td>
<td><code>unique(proteo@tax_table@.Data[,2])</code></td>
</tr><tr class="odd"><td>Show the plot.</td>
<td><code>proteo_glom &lt;- tax_glom(proteo_percentages, taxrank = "Genus")</code></td>
</tr><tr class="even"><td>Transform the phyloseq object to a data frame.</td>
<td><code>plot_proteo &lt;- ggplot(data=proteo_df, aes(x=Sample, y=Abundance, fill=Genus))+</code></td>
</tr><tr class="odd"><td></td>
<td><code>geom_bar(aes(), stat="identity", position="stack")+</code></td>
</tr><tr class="even"><td></td>
<td><code>scale_fill_manual(values = genus_colors_proteo)</code></td>
</tr><tr class="odd"><td>Convert the Genus column into the factor structure.</td>
<td><code>proteo_df$Genus[proteo_df$Abundance &lt; 3] &lt;- "Genera &lt; 3% abund"</code></td>
</tr><tr class="even"><td>Look at the phyla present in your phyloseq object.</td>
<td><code>proteo_df &lt;- psmelt(proteo_glom)</code></td>
</tr><tr class="odd"><td>Convert the abundance counts to relative abundance.</td>
<td><code>genus_colors_proteo&lt;- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(proteo_df$Genus)))</code></td>
</tr><tr class="even"><td>Make a palette with the appropriate colors for the number of
genera.</td>
<td><code>proteo_df$Genus &lt;- as.factor(proteo_df$Genus)</code></td>
</tr></tbody></table></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a phyloseq object only with the reads assigned to a certain phylum.</span></span>
<span><span class="va">proteo</span> <span class="op">&lt;-</span> <span class="fu">subset_taxa</span><span class="op">(</span><span class="va">merged_metagenomes</span>, <span class="va">Phylum</span> <span class="op">==</span> <span class="st">"Proteobacteria"</span><span class="op">)</span></span>
<span><span class="co"># Look at the phyla present in your phyloseq object</span></span>
<span><span class="fu">unique</span><span class="op">(</span><span class="va">proteo</span><span class="op">@</span><span class="va">tax_table</span><span class="op">@</span><span class="va">.Data</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Convert the abundance counts to the relative abundance</span></span>
<span><span class="va">proteo_percentages</span> <span class="op">&lt;-</span> <span class="fu">transform_sample_counts</span><span class="op">(</span><span class="va">proteo</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">*</span><span class="fl">100</span> <span class="op">/</span> <span class="fu">sum</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="co"># Make just one row that groups all the observations of the same genus.</span></span>
<span><span class="va">proteo_glom</span> <span class="op">&lt;-</span> <span class="fu">tax_glom</span><span class="op">(</span><span class="va">proteo_percentages</span>, taxrank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span></span>
<span><span class="co"># Transform the phyloseq object to a data frame</span></span>
<span><span class="va">proteo_df</span> <span class="op">&lt;-</span> <span class="fu">psmelt</span><span class="op">(</span><span class="va">proteo_glom</span><span class="op">)</span></span>
<span><span class="co"># Convert all the genera that have less than 3% of abundance into only one label</span></span>
<span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span><span class="op">[</span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Abundance</span> <span class="op">&lt;</span> <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Genera &lt; 3% abund"</span></span>
<span><span class="co"># Convert the Genus column into the factor structure</span></span>
<span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span><span class="op">)</span></span>
<span><span class="co"># Make a palette with the appropriate colors for the number of genera</span></span>
<span><span class="va">genus_colors_proteo</span><span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>,<span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span> <span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="fu">levels</span><span class="op">(</span><span class="va">proteo_df</span><span class="op">$</span><span class="va">Genus</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Plot the relative abundance at the genus levels</span></span>
<span><span class="va">plot_proteo</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span>data<span class="op">=</span><span class="va">proteo_df</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">Sample</span>, y<span class="op">=</span><span class="va">Abundance</span>, fill<span class="op">=</span><span class="va">Genus</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="op">)</span>, stat<span class="op">=</span><span class="st">"identity"</span>, position<span class="op">=</span><span class="st">"stack"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">genus_colors_proteo</span><span class="op">)</span></span>
<span><span class="co"># Show the plot</span></span>
<span><span class="va">plot_proteo</span>  </span></code></pre>
</div>
<p><a href="../%7B%7B%20page.root%20%7D%7D/fig/03-09-06.png">
<img src="../%7B%7B%20page.root%20%7D%7D/fig/03-09-06.png" alt="A new plot with three bars
  representing the absolute abundance of Proteobacteria in each of the samples.
  Each of the colors represents a Genus. Because we see relative
  abundances, all the bars have the same height." class="figure"></a></p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Depths and abundances can be visualized using phyloseq.</li>
<li>The library <code>phyloseq</code> lets you manipulate metagenomic
data in a taxonomic specific perspective.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/08-Diversity-tackled-with-R.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/10-OtherResources.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/08-Diversity-tackled-with-R.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Diversity Tackled
        </a>
        <a class="chapter-link float-end" href="../instructor/10-OtherResources.html" rel="next">
          Next: Other Resources...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-lab/metagenomics-analysis/edit/main/episodes/09-abundance-analyses.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-lab/metagenomics-analysis/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-lab/metagenomics-analysis/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:nselem@matmor.unam.mx">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/09-abundance-analyses.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Taxonomic Analysis with R",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/09-abundance-analyses.html",
  "identifier": "https://carpentries-lab.github.io/metagenomics-analysis/instructor/09-abundance-analyses.html",
  "dateCreated": "2020-08-24",
  "dateModified": "2025-04-06",
  "datePublished": "2025-05-13"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

