---
title: "Other Resources"
teaching: 5
exercises: 0
questions:
- "Where are other metagenomic resources?"
- "How can lessons be previewed?"
objectives:
- "Know some places where to find more information."
keypoints:
- "Enjoy metagenomics."
---

## Other resources 
Now that we finished togheter our lesson, you are ready to start exploring the universe of tutorials available. [Phyloseq tutorial](https://joey711.github.io/phyloseq/) contain much more examples of metagenome data manipulation. More information about 4 Ciénegas is available in these podcasts [Mexican oasis in the desert](https://www.sciencemag.org/podcast/oasis-biodiversity-mexican-desert-and-making-sound-heat) and [Mexican Oasis]( https://youtu.be/xMMm_GKZsnU), in this [Cuatro Ciénegas Video](https://www.youtube.com/embed/VzImXRI9wYE?autoplay=1&rel=0) , and finally in paper [Cuatro Ciénegas](https://www.sciencemag.org/news/2020/06/pools-mexican-desert-are-window-earth-s-early-life). There are othe tutorials such as the [Computational Genomics tutorial](https://genomics.sschmeier.com/) by Sschmeir with detalied explanation about what is happening. The ISME course (in spanish) is a tutorial for 16s metagenomcis [Phyloseq Análisis de diversidad](http://www.castrolab.org/isme/biodiversity/biodiversity.html). A'nvio developer Meren has a blog entry about [History of metagenomics](http://merenlab.org/2020/07/27/history-of-metagenomics/) and several videos explaining [Metapangenomics: A nexus between pangenomes and metagenomes](https://youtu.be/C3fHlccFxJw), [The power of metagenomic read recruitment](https://youtu.be/MqD4aN1p1qA),[Genome-resolved metagenomics: key concepts in reconstructing genomes from metagenomes](https://youtu.be/RjNdHGK4ruo)  

### MG-RAST

A useful and easy to use resource is [MG-RAST](https://www.mg-rast.org/), an online metagenomic platform where you can upload your raw data with its corresponding metadata and get a full taxonomic analysis of it. MG-RAST is a great place to get started in this type of analyzes and it is also a big repository of available data for future experiments. On the downside, it is not possible to greatly modify the steps and parameters in the MG-RAST workflow, so there is not much room when it comes to implementing our preferred analysis tools when using MG-RAST.

The Cuatro Ciénegas data that we used in the workshop is already in MG-RAST! You can check it out [here](https://www.mg-rast.org/mgmain.html?mgpage=project&project=mgp96823). 

We can check the taxonomical distribution of our sample at different taxonomical level.

The most abundant phylum is Proteobacteria.  
<a href="../fig/03-11-02.png">
  <img src="../fig/03-11-02.png" alt="Phylum" />
</a>

Since we have a shotgun metagenome, we can also investigate the metabolic functions 
present in our sample. MG-RAST can find genes and annotate their function through 
an implementation of RAST, or Rapid Annotation using Subsystems Technology. 
By looking at the charts generated by this analysis, we see that most of the genes
are dedicated to metabolism.  

<a href="../fig/03-11-04.png">
  <img src="../fig/03-11-04.png" alt="Cog Metagenome" />
</a>

<a href="../fig/03-11-05.png">
  <img src="../fig/03-11-05.png" alt="Subsystems" />
</a>

MG-RAST has it's own specific pipeline, so it is a very useful tool to have a quick look of your data, and also to store it and share it!, but it does not keep you from making your own personalized analysis like we just learn!

<!-- [Evomics](http://evomics.org/learning/genomics/), [Data Carpentry in 16S Metagenomics](https://datacarpentry.org/blog/2017/11/16s-dc)  -->

> ## Discussion: Exploring more resources
>
> Explore one of the suggested resources and discuss your findings with a neighbor.
{: .discussion}


## Carpentries Philosophy
A good lesson should be as complete and clear that becomes easy to teach by any instructor. 
Carpentries lessons are developed for the community, and now you are part of us. 
This lesson is being developed and we are sure that you can colaborate and help us improve it.  
<!--## How does our results compare with the original research-->  
<!-- ## How can we improve the data analysis!--->  
                             
{% include links.md %}
